<!DOCTYPE html>
<html class="writer-html5" lang="en">
<head>
  <meta charset="utf-8" /><meta name="generator" content="Docutils 0.18.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>End to end selection &mdash; spapros docs</title>
      <link rel="stylesheet" type="text/css" href="../_static/pygments.css" />
      <link rel="stylesheet" type="text/css" href="../_static/css/theme.css" />
      <link rel="stylesheet" type="text/css" href="../_static/sg_gallery.css" />
      <link rel="stylesheet" type="text/css" href="../_static/nbsphinx-code-cells.css" />
      <link rel="stylesheet" type="text/css" href="../_static/custom_cookietemple.css" />
      <link rel="stylesheet" type="text/css" href="../_static/dark_mode_css/general.css" />
      <link rel="stylesheet" type="text/css" href="../_static/dark_mode_css/dark.css" />

  
    <link rel="shortcut icon" href="../_static/logo.png"/>
  <!--[if lt IE 9]>
    <script src="../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script data-url_root="../" id="documentation_options" src="../_static/documentation_options.js"></script>
        <script src="../_static/jquery.js"></script>
        <script src="../_static/underscore.js"></script>
        <script src="../_static/_sphinx_javascript_frameworks_compat.js"></script>
        <script src="../_static/doctools.js"></script>
        <script src="../_static/sphinx_highlight.js"></script>
        <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
        <script src="https://cdn.jsdelivr.net/npm/@jupyter-widgets/html-manager@^1.0.1/dist/embed-amd.js"></script>
        <script>window.MathJax = {"tex": {"inlineMath": [["$", "$"], ["\\(", "\\)"]], "processEscapes": true}, "options": {"ignoreHtmlClass": "tex2jax_ignore|mathjax_ignore|document", "processHtmlClass": "tex2jax_process|mathjax_process|math|output_area"}}</script>
        <script defer="defer" src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
        <script src="../_static/dark_mode_js/default_light.js"></script>
        <script src="../_static/dark_mode_js/theme_switcher.js"></script>
    <script src="../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="Credits" href="../authors.html" />
    <link rel="prev" title="End to end selection (short version)" href="spapros_tutorial_end_to_end_selection_short.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="../index.html" class="icon icon-home">
            spapros
              <img src="../_static/logo.png" class="logo" alt="Logo"/>
          </a>
              <div class="version">
                0.1.5
              </div>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">CONTENTS:</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../installation.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../usage.html">Usage</a></li>
<li class="toctree-l1"><a class="reference internal" href="../api.html">API</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="../tutorials.html">Tutorials</a><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="spapros_tutorial_basic_selection.html">Probeset selection with Spapros</a></li>
<li class="toctree-l2"><a class="reference internal" href="spapros_tutorial_basic_evaluation.html">Evaluation of probe sets</a></li>
<li class="toctree-l2"><a class="reference internal" href="spapros_tutorial_advanced_selection.html">Advanced probeset selection</a></li>
<li class="toctree-l2"><a class="reference internal" href="spapros_tutorial_advanced_evaluation.html">Advanced evaluation</a></li>
<li class="toctree-l2"><a class="reference internal" href="spapros_tutorial_end_to_end_selection_short.html">End to end selection (short version)</a></li>
<li class="toctree-l2 current"><a class="current reference internal" href="#">End to end selection</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#Import-Packages">Import Packages</a></li>
<li class="toctree-l3"><a class="reference internal" href="#Load-and-Preprocess-Data">Load and Preprocess Data</a></li>
<li class="toctree-l3"><a class="reference internal" href="#Probeset-Design">Probeset Design</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#Generate-Genomic-Region-Files">Generate Genomic Region Files</a></li>
<li class="toctree-l4"><a class="reference internal" href="#Create-Probe-Database">Create Probe Database</a></li>
<li class="toctree-l4"><a class="reference internal" href="#Refine-Database-with-Property-Filters">Refine Database with Property Filters</a></li>
<li class="toctree-l4"><a class="reference internal" href="#Refine-Database-with-Specificity-Filters">Refine Database with Specificity Filters</a></li>
<li class="toctree-l4"><a class="reference internal" href="#Create-Probe-Sets-of-Non-Overlapping-Probes">Create Probe Sets of Non-Overlapping Probes</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#Select-Gene-Panel">Select Gene Panel</a></li>
<li class="toctree-l3"><a class="reference internal" href="#Get-Final-Padlock-Probe-Sequences">Get Final Padlock Probe Sequences</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../authors.html">Credits</a></li>
<li class="toctree-l1"><a class="reference internal" href="../code_of_conduct.html">Contributor Covenant Code of Conduct</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../index.html">spapros</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../index.html" class="icon icon-home" aria-label="Home"></a></li>
          <li class="breadcrumb-item"><a href="../tutorials.html">Tutorials</a></li>
      <li class="breadcrumb-item active">End to end selection</li>
      <li class="wy-breadcrumbs-aside">
            <a href="../_sources/_tutorials/spapros_tutorial_end_to_end_selection.ipynb.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="End-to-end-selection">
<h1>End to end selection<a class="headerlink" href="#End-to-end-selection" title="Permalink to this heading"></a></h1>
<p>In this tutorial we showcase how to design probesets and select a suitable gene set with the <code class="docutils literal notranslate"><span class="pre">spapros</span></code> package. For all genes we design probes that fulfill experiment specific requirements and select only genes for which we can design sufficient probes. Spapros then selects genes that can distinguish the cell types in the data set and capture transcriptomic varation beyond cell type labels. The final probe sequences are designed in a last step for all selected genes. The figure below gives and
overview on the pipeline.</p>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[2]:
</pre></div>
</div>
<div class="output_area docutils container">
<img alt="../_images/_tutorials_spapros_tutorial_end_to_end_selection_2_0.png" class="no-scaled-link" src="../_images/_tutorials_spapros_tutorial_end_to_end_selection_2_0.png" style="width: 900px;" />
</div>
</div>
<section id="Import-Packages">
<h2>Import Packages<a class="headerlink" href="#Import-Packages" title="Permalink to this heading"></a></h2>
<p>Besides <code class="docutils literal notranslate"><span class="pre">spapros</span></code> also install <code class="docutils literal notranslate"><span class="pre">oligo_designer_toolsuite</span></code>. Therefore, first setup a conda environment (packages is tested for Python 3.9 - 3.10), e.g.:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>conda<span class="w"> </span>create<span class="w"> </span>-n<span class="w"> </span>odt<span class="w"> </span><span class="nv">python</span><span class="o">=</span><span class="m">3</span>.10
conda<span class="w"> </span>activate<span class="w"> </span>odt
</pre></div>
</div>
<p>Then, install the required dependencies, i.e. <strong>Blast</strong> (2.15 or higher), <strong>BedTools</strong> (2.30 or higher), <strong>Bowtie</strong> (1.3 or higher) and <strong>Bowtie2</strong> (2.5 or higher), that need to be installed independently. To install those tools via conda, please activate the Bioconda and conda-forge channels in your conda environment with and update conda and all packages in your environment:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>conda<span class="w"> </span>config<span class="w"> </span>--add<span class="w"> </span>channels<span class="w"> </span>conda-forge
conda<span class="w"> </span>config<span class="w"> </span>--add<span class="w"> </span>channels<span class="w"> </span>bioconda
conda<span class="w"> </span>update<span class="w"> </span>--all

conda<span class="w"> </span>install<span class="w"> </span><span class="s2">&quot;blast&gt;=2.15.0&quot;</span>
conda<span class="w"> </span>install<span class="w"> </span><span class="s2">&quot;bedtools&gt;=2.30&quot;</span>
conda<span class="w"> </span>install<span class="w"> </span><span class="s2">&quot;bowtie&gt;=1.3.1&quot;</span>
conda<span class="w"> </span>install<span class="w"> </span><span class="s2">&quot;bowtie2&gt;=2.5&quot;</span>
</pre></div>
</div>
<p>All other required packages are automatically installed during the <code class="docutils literal notranslate"><span class="pre">pip</span></code> installation:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>git<span class="w"> </span>clone<span class="w"> </span>https://github.com/HelmholtzAI-Consultants-Munich/oligo-designer-toolsuite.git
<span class="nb">cd</span><span class="w"> </span>oligo-designer-toolsuite
git<span class="w"> </span>switch<span class="w"> </span>pipelines
pip<span class="w"> </span>install<span class="w"> </span>-e<span class="w"> </span>.
</pre></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[3]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">os</span>
<span class="kn">import</span> <span class="nn">scanpy</span> <span class="k">as</span> <span class="nn">sc</span>
<span class="n">sc</span><span class="o">.</span><span class="n">settings</span><span class="o">.</span><span class="n">verbosity</span> <span class="o">=</span> <span class="mi">0</span>
<span class="n">sc</span><span class="o">.</span><span class="n">logging</span><span class="o">.</span><span class="n">print_header</span><span class="p">()</span>

<span class="kn">import</span> <span class="nn">spapros</span> <span class="k">as</span> <span class="nn">sp</span>
<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;spapros==</span><span class="si">{</span><span class="n">sp</span><span class="o">.</span><span class="n">__version__</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

<span class="kn">from</span> <span class="nn">Bio.SeqUtils</span> <span class="kn">import</span> <span class="n">MeltingTemp</span> <span class="k">as</span> <span class="n">mt</span>
<span class="kn">from</span> <span class="nn">oligo_designer_toolsuite.database</span> <span class="kn">import</span> <span class="n">OligoDatabase</span>
<span class="kn">from</span> <span class="nn">oligo_designer_toolsuite.pipelines</span> <span class="kn">import</span> <span class="n">GenomicRegionGenerator</span><span class="p">,</span> <span class="n">ScrinshotProbeDesigner</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
scanpy==1.10.1 anndata==0.10.7 umap==0.5.6 numpy==1.26.4 scipy==1.13.1 pandas==1.5.3 scikit-learn==1.5.0 statsmodels==0.14.2 igraph==0.11.5 pynndescent==0.5.12
spapros==0.1.5
</pre></div></div>
</div>
</section>
<section id="Load-and-Preprocess-Data">
<h2>Load and Preprocess Data<a class="headerlink" href="#Load-and-Preprocess-Data" title="Permalink to this heading"></a></h2>
<p>For this tutorial, we use a PBMC example scRNA-seq reference dataset. The count data should be log-normalised and genes should <strong>not</strong> be scaled to mean=0 and std=1. We can load the processed version of the data, including cell / gene filters, cell type annotations, and the umap embedding, directly with <code class="docutils literal notranslate"><span class="pre">sp.ut.get_processed_pbmc_data()</span></code> function. For a step by step processing of the PBMC dataset please refer to the <a class="reference internal" href="spapros_tutorial_basic_selection.html"><span class="doc">basic usage tutorial</span></a>. For sake
of simplicity, we pre-select the top 1000 highly variable genes for the probe and geneset selection. In real world applications we typically go for top 8000 genes.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[4]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">pbmc_data</span> <span class="o">=</span> <span class="n">sp</span><span class="o">.</span><span class="n">ut</span><span class="o">.</span><span class="n">get_processed_pbmc_data</span><span class="p">(</span><span class="n">n_hvg</span><span class="o">=</span><span class="mi">1000</span><span class="p">)</span>
<span class="n">highly_variable_genes</span> <span class="o">=</span> <span class="nb">sorted</span><span class="p">(</span><span class="n">pbmc_data</span><span class="o">.</span><span class="n">var</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">pbmc_data</span><span class="o">.</span><span class="n">var</span><span class="p">[</span><span class="s1">&#39;highly_variable&#39;</span><span class="p">]]</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">tolist</span><span class="p">())</span>
<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Number of highly variable genes: </span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">highly_variable_genes</span><span class="p">)</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Number of highly variable genes: 1000
</pre></div></div>
</div>
</section>
<section id="Probeset-Design">
<h2>Probeset Design<a class="headerlink" href="#Probeset-Design" title="Permalink to this heading"></a></h2>
<p>Before choosing a gene panel, we design probesets for our given set of 1000 highly variable genes that fulfill certain experiment-specific criteria. Therefore, we first have to create the input fasta files for the probe design pipeline. We can create those files from custom annotation (GFF and fasta) files or download those annotation files directly from the NCBI or Ensembl FTP server. For generating the probe design pipeline input fasta files, we use the <code class="docutils literal notranslate"><span class="pre">GenomicRegionGenerator</span></code>class for
which we need to set the parameters:</p>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">dir_output</span></code>: name of the directory where the output files will be written</p></li>
</ul>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[5]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1">## Define parameters</span>
<span class="n">dir_output</span> <span class="o">=</span> <span class="s2">&quot;output_genomic_region_generator_ncbi&quot;</span>
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[6]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1">## Setup pipeline</span>
<span class="n">pipeline</span> <span class="o">=</span> <span class="n">GenomicRegionGenerator</span><span class="p">(</span><span class="n">dir_output</span><span class="o">=</span><span class="n">dir_output</span><span class="p">)</span>
</pre></div>
</div>
</div>
<section id="Generate-Genomic-Region-Files">
<h3>Generate Genomic Region Files<a class="headerlink" href="#Generate-Genomic-Region-Files" title="Permalink to this heading"></a></h3>
<p>Our example dataset uses a custom annotation and we create input fasta files for the transcriptome, consisting of exons and exon-exon junctions. Hence, we define <em>custom</em> as source and define the custom-specific parameters listed below. Apart from <em>custom</em> annotation, we can also choose an <em>NCBI</em> or <em>Ensembl</em> annotation. If <em>source=”ncbi”</em> or <em>source=”ensembl”</em> is choosen, the annotation files are automatically downloaded from their servers.</p>
<div class="line-block">
<div class="line"><strong>Parameters for annotation loader</strong></div>
<div class="line">- <code class="docutils literal notranslate"><span class="pre">source</span></code>: define annotation source -&gt; currently supported: ncbi, ensembl and custom</div>
</div>
<p><em>Custom annnotation parameters:</em> - <code class="docutils literal notranslate"><span class="pre">source_params</span></code>: - <code class="docutils literal notranslate"><span class="pre">file_annotation</span></code>: required: GTF file with gene annotation - <code class="docutils literal notranslate"><span class="pre">file_sequence</span></code>: required: FASTA file with genome sequence - <code class="docutils literal notranslate"><span class="pre">files_source</span></code>: optional: original source of the genomic files - <code class="docutils literal notranslate"><span class="pre">species</span></code>: optional: species of provided annotation, leave empty if unknown - <code class="docutils literal notranslate"><span class="pre">annotation_release</span></code>: optional: release number of provided annotation, leave empty if unknown - <code class="docutils literal notranslate"><span class="pre">genome_assembly</span></code>: optional: genome assembly of provided annotation,
leave empty if unknown</p>
<p><em>For NCBI or Ensembl parameters, see examples in the code below</em></p>
<div class="line-block">
<div class="line"><strong>Parameters for sequences generation</strong></div>
<div class="line">Set the genomic regions that should be generated to True. For each region one fasta file is generated. - <code class="docutils literal notranslate"><span class="pre">genomic_regions</span></code>: - <code class="docutils literal notranslate"><span class="pre">gene</span></code>: create fasta file from gene regions in GFF annotation - <code class="docutils literal notranslate"><span class="pre">exon</span></code>: create fasta file from exon regions in GFF annotation and merge same exons coming from different transcripts into one sequence entry while preserving the transcript information - <code class="docutils literal notranslate"><span class="pre">exon_exon_junction</span></code>: create fasta file from exon exon junctions in GFF annotation and merge same junctions
coming from different transcripts into one sequence entry while preserving the transcript information - <code class="docutils literal notranslate"><span class="pre">cds</span></code>: create fasta file from cds regions in GFF annotation and merge same cds coming from different transcripts into one sequence entry while preserving the transcript information - <code class="docutils literal notranslate"><span class="pre">intron</span></code>: create fasta file from regions between exons</div>
</div>
<p><code class="docutils literal notranslate"><span class="pre">exon_exon_junction_block_size</span></code>: If exon_exon_junction is set to true, specify the block size, i.e. +/- “block_size” bp around the junction. Hint: it does not make sense to set the block size larger than the maximum probe length</p>
<p><em>Note: if an error occurs for the unzipping of files, this might be due to a faulty download of files from the ftp server. In this case, try to download the files manually from the ftp server and use those files as input for the pipeline with custom input files.</em></p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[7]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1">## Define parameters for NCBI source</span>
<span class="n">source</span> <span class="o">=</span> <span class="s2">&quot;ncbi&quot;</span>
<span class="n">source_params</span> <span class="o">=</span> <span class="p">{</span>
    <span class="s2">&quot;taxon&quot;</span><span class="p">:</span> <span class="s2">&quot;vertebrate_mammalian&quot;</span><span class="p">,</span> <span class="c1"># required: taxon of the species, valid taxa are: archaea, bacteria, fungi, invertebrate, mitochondrion, plant, plasmid, plastid, protozoa, vertebrate_mammalian, vertebrate_other, viral</span>
    <span class="s2">&quot;species&quot;</span><span class="p">:</span> <span class="s2">&quot;Homo_sapiens&quot;</span><span class="p">,</span> <span class="c1"># required: species name in NCBI download format, e.g. &#39;Homo_sapiens&#39; for human; see https://ftp.ncbi.nlm.nih.gov/genomes/refseq/ for available species name</span>
    <span class="s2">&quot;annotation_release&quot;</span><span class="p">:</span> <span class="mi">110</span> <span class="c1"># required: release number of annotation e.g. &#39;109&#39; or &#39;109.20211119&#39;  or &#39;current&#39; to use most recent annotation release. Check out release numbers for NCBI at ftp.ncbi.nlm.nih.gov/refseq/H_sapiens/annotation/annotation_releases/</span>
<span class="p">}</span>

<span class="c1">## Define parameters for ensembl source</span>
<span class="c1"># source = &quot;ensembl&quot;</span>
<span class="c1"># source_params = {</span>
<span class="c1">#     &quot;species&quot;: &quot;homo_sapiens&quot;, # required: species name in ensemble download format, e.g. &#39;homo_sapiens&#39; for human; see http://ftp.ensembl.org/pub/release-108/gtf/ for available species names</span>
<span class="c1">#     &quot;annotation_release&quot;: &quot;109&quot;, # required: release number of annotation, e.g. &#39;release-108&#39; or &#39;current&#39; to use most recent annotation release. Check out release numbers for ensemble at ftp.ensembl.org/pub/</span>
<span class="c1"># }</span>

<span class="c1">## Define parameters for custom source</span>
<span class="c1"># source = &quot;custom&quot;</span>
<span class="c1"># source_params = {</span>
<span class="c1">#     &quot;file_annotation&quot;: &quot;data/custom_GCF_000001405.40_GRCh38.p14_genomic_chr16.gtf&quot;,</span>
<span class="c1">#     &quot;file_sequence&quot;: &quot;data/custom_GCF_000001405.40_GRCh38.p14_genomic_chr16.fna&quot;,</span>
<span class="c1">#     &quot;files_source&quot;: &quot;NCBI&quot;,</span>
<span class="c1">#     &quot;species&quot;: &quot;Homo_sapiens&quot;,</span>
<span class="c1">#     &quot;annotation_release&quot;: 110,</span>
<span class="c1">#     &quot;genome_assembly&quot;: &quot;GRCh38&quot;,</span>
<span class="c1"># }</span>

<span class="n">genomic_regions</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;gene&quot;</span><span class="p">:</span> <span class="kc">False</span><span class="p">,</span> <span class="s2">&quot;exon&quot;</span><span class="p">:</span> <span class="kc">True</span><span class="p">,</span> <span class="s2">&quot;exon_exon_junction&quot;</span><span class="p">:</span> <span class="kc">True</span><span class="p">,</span> <span class="s2">&quot;cds&quot;</span><span class="p">:</span> <span class="kc">False</span><span class="p">,</span> <span class="s2">&quot;intron&quot;</span><span class="p">:</span> <span class="kc">False</span><span class="p">}</span>
<span class="n">block_size</span> <span class="o">=</span> <span class="mi">39</span> <span class="c1"># min probe size - 1</span>
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[8]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Generate the genomic regions</span>
<span class="n">region_generator</span> <span class="o">=</span> <span class="n">pipeline</span><span class="o">.</span><span class="n">load_annotations</span><span class="p">(</span>
    <span class="n">source</span><span class="o">=</span><span class="n">source</span><span class="p">,</span>
    <span class="n">source_params</span><span class="o">=</span><span class="n">source_params</span><span class="p">,</span>
<span class="p">)</span>

<span class="n">fasta_files</span> <span class="o">=</span> <span class="n">pipeline</span><span class="o">.</span><span class="n">generate_genomic_regions</span><span class="p">(</span>
    <span class="n">region_generator</span><span class="o">=</span><span class="n">region_generator</span><span class="p">,</span>
    <span class="n">genomic_regions</span><span class="o">=</span><span class="n">genomic_regions</span><span class="p">,</span>
    <span class="n">block_size</span><span class="o">=</span><span class="n">block_size</span><span class="p">,</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
2024-07-08 14:35:32,314 [INFO] Parameters Load Annotations:
2024-07-08 14:35:32,316 [INFO] source = ncbi
2024-07-08 14:35:32,317 [INFO] source_params = {&#39;taxon&#39;: &#39;vertebrate_mammalian&#39;, &#39;species&#39;: &#39;Homo_sapiens&#39;, &#39;annotation_release&#39;: 110}
2024-07-08 14:44:42,316 [INFO] The following annotation files are used for GTF annotation of regions: /Users/lisa.barros/Desktop/spapros/docs/_tutorials/output_genomic_region_generator_ncbi/annotation/GCF_000001405.40_GRCh38.p14_genomic.gtf and for fasta sequence file: /Users/lisa.barros/Desktop/spapros/docs/_tutorials/output_genomic_region_generator_ncbi/annotation/GCF_000001405.40_GRCh38.p14_genomic.fna .
2024-07-08 14:44:42,381 [INFO] The annotations are from NCBI source, for the species: Homo_sapiens, release number: 110 and genome assembly: GRCh38.p14
2024-07-08 14:48:17,407 [INFO] The genomic region &#39;exon&#39; was stored in :/Users/lisa.barros/Desktop/spapros/docs/_tutorials/output_genomic_region_generator_ncbi/annotation/exon_annotation_source-NCBI_species-Homo_sapiens_annotation_release-110_genome_assemly-GRCh38.p14.fna.
2024-07-08 15:00:11,519 [INFO] The genomic region &#39;exon_exon_junction&#39; was stored in :/Users/lisa.barros/Desktop/spapros/docs/_tutorials/output_genomic_region_generator_ncbi/annotation/exon_exon_junction_annotation_source-NCBI_species-Homo_sapiens_annotation_release-110_genome_assemly-GRCh38.p14.fna.
</pre></div></div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[9]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">fasta_files</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;output_genomic_region_generator_ncbi/annotation/exon_annotation_source-NCBI_species-Homo_sapiens_annotation_release-110_genome_assemly-GRCh38.p14.fna&quot;</span><span class="p">,</span>
               <span class="s2">&quot;output_genomic_region_generator_ncbi/annotation/exon_exon_junction_annotation_source-NCBI_species-Homo_sapiens_annotation_release-110_genome_assemly-GRCh38.p14.fna&quot;</span><span class="p">]</span>
</pre></div>
</div>
</div>
</section>
<section id="Create-Probe-Database">
<h3>Create Probe Database<a class="headerlink" href="#Create-Probe-Database" title="Permalink to this heading"></a></h3>
<p>After generating the input fasta files, we can start with the probe design process. Therefore, we first create an instance of a ProbeDesigner class, where we can choose from <code class="docutils literal notranslate"><span class="pre">ScrinshotProbeDesigner</span></code>, <code class="docutils literal notranslate"><span class="pre">MerfishProbeDesigner</span></code> an <code class="docutils literal notranslate"><span class="pre">SeqfishPlusProbeDesigner</span></code> (see our <a class="reference external" href="https://www.google.com/url?q=https://docs.google.com/spreadsheets/d/1NCDLscdmzn32U7_IKy6OKYHqfECn76x0pAD3KFhuJgQ/edit%23gid%3D0&amp;sa=D&amp;source=docs&amp;ust=1692975800552487&amp;usg=AOvVaw20CRwiObnVkWVS54CcqXMq">resource table</a> for an
overview of differences between the technologies). For each of those classes, we need to set the parameter:</p>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">gene_ids</span></code>: a list the genes used to generate the oligos sequences, leave empty list if all the genes should be used</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">write_intermediate_steps</span></code>: if true, writes the oligo sequences after each step of the pipeline into a tsv file</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">dir_output</span></code>: name of the directory where the output files will be written</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">n_jobs</span></code>: number of cores used to run the pipeline and number of regions that should be stored in cache. If memory consumption of pipeline is too high reduce this number, if a lot of RAM is available increase this number to decrease runtime</p></li>
</ul>
<p>Here, we showcase the probe design of padlock probes for 1000 highly variable genes.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[10]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1">## Define parameters</span>
<span class="n">gene_ids</span> <span class="o">=</span> <span class="n">highly_variable_genes</span> <span class="c1"># here we use the list of genes of highly variable genes we stored in step 1</span>
<span class="n">write_intermediate_steps</span> <span class="o">=</span> <span class="kc">True</span>
<span class="n">dir_output</span> <span class="o">=</span> <span class="s2">&quot;output_scrinshot_probe_designer&quot;</span>
<span class="n">n_jobs</span> <span class="o">=</span> <span class="mi">4</span>
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[11]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1">## Setup pipeline</span>
<span class="n">pipeline</span> <span class="o">=</span> <span class="n">ScrinshotProbeDesigner</span><span class="p">(</span>
    <span class="n">write_intermediate_steps</span><span class="o">=</span><span class="n">write_intermediate_steps</span><span class="p">,</span>
    <span class="n">dir_output</span><span class="o">=</span><span class="n">dir_output</span><span class="p">,</span>
    <span class="n">n_jobs</span><span class="o">=</span><span class="n">n_jobs</span><span class="p">,</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
<p>After instatiating our ScrinshotProbeDesigner class, we have to create the oligo database. Running the function below, will automatically create all possible probes from the input fasta files for each gene that is provided in the <em>file_regions</em> file.</p>
<p><strong>Parameters for Probe Database</strong> - <code class="docutils literal notranslate"><span class="pre">probe_length_min</span></code>: minimum length of probes - <code class="docutils literal notranslate"><span class="pre">probe_length_max</span></code>: maximum length of probes - <code class="docutils literal notranslate"><span class="pre">probes_per_gene_min</span></code>: minimum number of probes that a gene must have before it gets deleted - <code class="docutils literal notranslate"><span class="pre">files_fasta_oligo_database</span></code>: fasta file with sequences form which the oligos should be generated. Hint: use the GenomicRegionGenerator class to create fasta files of genomic regions of interest</p>
<div class="line-block">
<div class="line"><em>Note: Instead of creating a new probe database, we can also load an existing databases.</em></div>
<div class="line">Loading a database can be useful when starting the pipeline from a certain step, e.g. load a database which was already filtered by probe properties and continue immediately with the specificity filter step. We can load an existing database by calling <code class="docutils literal notranslate"><span class="pre">load_database()</span></code>. See example code in the cells below (commented).</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[12]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1">## Define parameters</span>
<span class="n">probe_length_min</span> <span class="o">=</span> <span class="mi">40</span>
<span class="n">probe_length_max</span> <span class="o">=</span> <span class="mi">45</span>
<span class="n">min_probes_per_gene</span> <span class="o">=</span> <span class="mi">3</span> <span class="c1"># should be at least &quot;min_probeset_size&quot; probes per gene to create one set</span>
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[13]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1">## Create initial database</span>
<span class="n">probe_database</span><span class="p">,</span> <span class="n">file_database</span> <span class="o">=</span> <span class="n">pipeline</span><span class="o">.</span><span class="n">create_probe_database</span><span class="p">(</span>
    <span class="n">gene_ids</span><span class="o">=</span><span class="n">gene_ids</span><span class="p">,</span>
    <span class="n">probe_length_min</span><span class="o">=</span><span class="n">probe_length_min</span><span class="p">,</span>
    <span class="n">probe_length_max</span><span class="o">=</span><span class="n">probe_length_max</span><span class="p">,</span>
    <span class="n">files_fasta_oligo_database</span><span class="o">=</span><span class="n">fasta_files</span><span class="p">,</span>
    <span class="n">min_probes_per_gene</span><span class="o">=</span><span class="n">min_probes_per_gene</span><span class="p">,</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
2024-07-08 15:00:47,199 [INFO] Parameters Create Database:
2024-07-08 15:00:47,204 [INFO] Function: create_probe_database
2024-07-08 15:00:47,205 [INFO] Parameter: gene_ids = [&#39;AAGAB&#39;, &#39;AATF&#39;, &#39;ABCC10&#39;, &#39;ABHD12&#39;, &#39;ABHD17B&#39;, &#39;ABHD5&#39;, &#39;ABRACL&#39;, &#39;ABT1&#39;, &#39;AC005082.12&#39;, &#39;AC074138.3&#39;, &#39;AC093323.3&#39;, &#39;ACAP1&#39;, &#39;ACBD3&#39;, &#39;ACD&#39;, &#39;ACOT13&#39;, &#39;ACP1&#39;, &#39;ACRBP&#39;, &#39;ACTL6A&#39;, &#39;ACTR6&#39;, &#39;ACVR2A&#39;, &#39;ADAL&#39;, &#39;ADAM10&#39;, &#39;ADAM28&#39;, &#39;ADD1&#39;, &#39;ADIPOR2&#39;, &#39;ADPRM&#39;, &#39;ADSL&#39;, &#39;AEBP1&#39;, &#39;AGPAT1&#39;, &#39;AHSA1&#39;, &#39;AIF1&#39;, &#39;AIM2&#39;, &#39;AKTIP&#39;, &#39;AL928768.3&#39;, &#39;ALKBH7&#39;, &#39;ANAPC13&#39;, &#39;ANKAR&#39;, &#39;ANKEF1&#39;, &#39;ANKRD27&#39;, &#39;ANKRD54&#39;, &#39;AP001462.6&#39;, &#39;AP003419.16&#39;, &#39;AP3M2&#39;, &#39;AP4B1-AS1&#39;, &#39;AP4S1&#39;, &#39;APOBEC3A&#39;, &#39;APOBEC3B&#39;, &#39;APOBEC3G&#39;, &#39;AQP3&#39;, &#39;ARHGAP11A&#39;, &#39;ARHGAP19&#39;, &#39;ARHGAP24&#39;, &#39;ARHGAP33&#39;, &#39;ARHGAP6&#39;, &#39;ARID4A&#39;, &#39;ARIH2OS&#39;, &#39;ARL2&#39;, &#39;ARL2BP&#39;, &#39;ARL4A&#39;, &#39;ARL6IP5&#39;, &#39;ARMC7&#39;, &#39;ARMCX5&#39;, &#39;ARRDC3&#39;, &#39;ARRDC4&#39;, &#39;ARSD&#39;, &#39;ARSG&#39;, &#39;ARVCF&#39;, &#39;ASB8&#39;, &#39;ASXL2&#39;, &#39;ATAD3C&#39;, &#39;ATF7IP2&#39;, &#39;ATG16L1&#39;, &#39;ATP10A&#39;, &#39;ATP5H&#39;, &#39;ATP5O&#39;, &#39;ATP5SL&#39;, &#39;ATP6V0E2&#39;, &#39;ATXN1L&#39;, &#39;ATXN3&#39;, &#39;AURKC&#39;, &#39;BABAM1&#39;, &#39;BACE2&#39;, &#39;BAZ2A&#39;, &#39;BBX&#39;, &#39;BCDIN3D&#39;, &#39;BET1&#39;, &#39;BEX4&#39;, &#39;BGLAP&#39;, &#39;BLNK&#39;, &#39;BLZF1&#39;, &#39;BMPR2&#39;, &#39;BNIP2&#39;, &#39;BOLA1&#39;, &#39;BOLA3&#39;, &#39;BRAT1&#39;, &#39;BRWD1&#39;, &#39;BTN3A1&#39;, &#39;BTN3A2&#39;, &#39;BUB3&#39;, &#39;C10orf32&#39;, &#39;C12orf45&#39;, &#39;C14orf1&#39;, &#39;C14orf166&#39;, &#39;C14orf80&#39;, &#39;C15orf57&#39;, &#39;C16orf13&#39;, &#39;C16orf52&#39;, &#39;C16orf54&#39;, &#39;C16orf58&#39;, &#39;C16orf74&#39;, &#39;C16orf80&#39;, &#39;C17orf59&#39;, &#39;C17orf62&#39;, &#39;C19orf33&#39;, &#39;C19orf52&#39;, &#39;C1QA&#39;, &#39;C1QB&#39;, &#39;C1QC&#39;, &#39;C1orf162&#39;, &#39;C1orf35&#39;, &#39;C21orf33&#39;, &#39;C2CD4D&#39;, &#39;C2orf76&#39;, &#39;C2orf88&#39;, &#39;C3orf18&#39;, &#39;C5orf15&#39;, &#39;C5orf42&#39;, &#39;C8orf44&#39;, &#39;C9orf142&#39;, &#39;C9orf16&#39;, &#39;C9orf37&#39;, &#39;CAMK1D&#39;, &#39;CAMK2G&#39;, &#39;CAMK2N1&#39;, &#39;CAPN12&#39;, &#39;CARHSP1&#39;, &#39;CARS&#39;, &#39;CASC4&#39;, &#39;CBX5&#39;, &#39;CCDC115&#39;, &#39;CCDC122&#39;, &#39;CCDC66&#39;, &#39;CCDC91&#39;, &#39;CCL3&#39;, &#39;CCL4&#39;, &#39;CCL5&#39;, &#39;CCND2&#39;, &#39;CCNG1&#39;, &#39;CCP110&#39;, &#39;CCT4&#39;, &#39;CCT7&#39;, &#39;CD160&#39;, &#39;CD19&#39;, &#39;CD2&#39;, &#39;CD247&#39;, &#39;CD274&#39;, &#39;CD2AP&#39;, &#39;CD320&#39;, &#39;CD72&#39;, &#39;CD79A&#39;, &#39;CD79B&#39;, &#39;CD82&#39;, &#39;CD9&#39;, &#39;CD96&#39;, &#39;CDC123&#39;, &#39;CDC16&#39;, &#39;CDC37&#39;, &#39;CDC40&#39;, &#39;CDK19&#39;, &#39;CDKN2A&#39;, &#39;CEACAM4&#39;, &#39;CEBPB&#39;, &#39;CECR5&#39;, &#39;CEP120&#39;, &#39;CEP68&#39;, &#39;CEP85L&#39;, &#39;CEPT1&#39;, &#39;CES4A&#39;, &#39;CGRRF1&#39;, &#39;CHD2&#39;, &#39;CHD7&#39;, &#39;CHERP&#39;, &#39;CHI3L2&#39;, &#39;CHPF2&#39;, &#39;CIAPIN1&#39;, &#39;CISD1&#39;, &#39;CISH&#39;, &#39;CITED4&#39;, &#39;CKS1B&#39;, &#39;CLDN5&#39;, &#39;CLEC2B&#39;, &#39;CLIC3&#39;, &#39;CLNS1A&#39;, &#39;CLPX&#39;, &#39;CLU&#39;, &#39;CLYBL&#39;, &#39;CMTM5&#39;, &#39;CNEP1R1&#39;, &#39;COMMD10&#39;, &#39;COQ7&#39;, &#39;CORO1B&#39;, &#39;COTL1&#39;, &#39;CPNE2&#39;, &#39;CPQ&#39;, &#39;CPSF3L&#39;, &#39;CR1&#39;, &#39;CRIP3&#39;, &#39;CRTC2&#39;, &#39;CST3&#39;, &#39;CST7&#39;, &#39;CTA-29F11.1&#39;, &#39;CTB-113I20.2&#39;, &#39;CTB-152G17.6&#39;, &#39;CTC-444N24.11&#39;, &#39;CTD-2015H6.3&#39;, &#39;CTD-2302E22.4&#39;, &#39;CTD-2368P22.1&#39;, &#39;CTD-2537I9.12&#39;, &#39;CTSS&#39;, &#39;CTSW&#39;, &#39;CWC15&#39;, &#39;CWC27&#39;, &#39;CXCL10&#39;, &#39;CXCL3&#39;, &#39;CYB5B&#39;, &#39;CYTH2&#39;, &#39;DAGLB&#39;, &#39;DCAF5&#39;, &#39;DDI2&#39;, &#39;DDT&#39;, &#39;DDX1&#39;, &#39;DDX17&#39;, &#39;DDX46&#39;, &#39;DDX56&#39;, &#39;DENND1C&#39;, &#39;DENND2D&#39;, &#39;DENND5B&#39;, &#39;DENND6A&#39;, &#39;DERL1&#39;, &#39;DEXI&#39;, &#39;DHX34&#39;, &#39;DHX9&#39;, &#39;DIDO1&#39;, &#39;DIMT1&#39;, &#39;DIS3&#39;, &#39;DISP1&#39;, &#39;DLST&#39;, &#39;DMTN&#39;, &#39;DNAJA3&#39;, &#39;DNAJB14&#39;, &#39;DNAJC10&#39;, &#39;DNAJC15&#39;, &#39;DNAJC2&#39;, &#39;DNAJC27&#39;, &#39;DNASE1L3&#39;, &#39;DNMT3A&#39;, &#39;DOK3&#39;, &#39;DPH6&#39;, &#39;DPY19L4&#39;, &#39;DRAXIN&#39;, &#39;DSCR3&#39;, &#39;DTX3&#39;, &#39;DUS3L&#39;, &#39;DUSP10&#39;, &#39;EAF2&#39;, &#39;EARS2&#39;, &#39;ECHDC1&#39;, &#39;EDC3&#39;, &#39;EID2&#39;, &#39;EIF1AY&#39;, &#39;EIF1B&#39;, &#39;EIF2B1&#39;, &#39;EIF3D&#39;, &#39;ELANE&#39;, &#39;ELOF1&#39;, &#39;ELOVL4&#39;, &#39;ELP6&#39;, &#39;EMB&#39;, &#39;EMG1&#39;, &#39;EML6&#39;, &#39;ENTPD3-AS1&#39;, &#39;EOGT&#39;, &#39;ERH&#39;, &#39;ERV3-1&#39;, &#39;EVA1B&#39;, &#39;EWSR1&#39;, &#39;EXOC6&#39;, &#39;F5&#39;, &#39;FADS1&#39;, &#39;FAM107B&#39;, &#39;FAM173A&#39;, &#39;FAM210B&#39;, &#39;FAM96A&#39;, &#39;FAM98A&#39;, &#39;FBXL14&#39;, &#39;FBXO21&#39;, &#39;FBXO33&#39;, &#39;FBXO4&#39;, &#39;FBXW4&#39;, &#39;FCER1A&#39;, &#39;FCER1G&#39;, &#39;FCGR2B&#39;, &#39;FCGR3A&#39;, &#39;FCN1&#39;, &#39;FCRLA&#39;, &#39;FEM1A&#39;, &#39;FERMT3&#39;, &#39;FGFBP2&#39;, &#39;FH&#39;, &#39;FHL1&#39;, &#39;FKBP3&#39;, &#39;FKBP5&#39;, &#39;FLOT1&#39;, &#39;FMO4&#39;, &#39;FN3KRP&#39;, &#39;FNBP4&#39;, &#39;FNTA&#39;, &#39;FOPNL&#39;, &#39;FRY-AS1&#39;, &#39;FUS&#39;, &#39;FXN&#39;, &#39;FYB&#39;, &#39;G0S2&#39;, &#39;GADD45B&#39;, &#39;GALT&#39;, &#39;GBGT1&#39;, &#39;GBP1&#39;, &#39;GDF11&#39;, &#39;GFER&#39;, &#39;GGA3&#39;, &#39;GGNBP2&#39;, &#39;GIMAP2&#39;, &#39;GIMAP4&#39;, &#39;GIMAP5&#39;, &#39;GIMAP7&#39;, &#39;GIT2&#39;, &#39;GMPPA&#39;, &#39;GNE&#39;, &#39;GNG11&#39;, &#39;GNG3&#39;, &#39;GNLY&#39;, &#39;GNPAT&#39;, &#39;GOLGB1&#39;, &#39;GP9&#39;, &#39;GPATCH4&#39;, &#39;GPKOW&#39;, &#39;GPR171&#39;, &#39;GPR183&#39;, &#39;GPR35&#39;, &#39;GPS1&#39;, &#39;GPX1&#39;, &#39;GRAP&#39;, &#39;GRN&#39;, &#39;GSTP1&#39;, &#39;GTPBP6&#39;, &#39;GUSB&#39;, &#39;GYS1&#39;, &#39;GZMA&#39;, &#39;GZMB&#39;, &#39;GZMH&#39;, &#39;GZMK&#39;, &#39;HAGH&#39;, &#39;HBA1&#39;, &#39;HBP1&#39;, &#39;HCFC2&#39;, &#39;HDAC1&#39;, &#39;HDAC5&#39;, &#39;HDAC9&#39;, &#39;HELQ&#39;, &#39;HEMK1&#39;, &#39;HERPUD2&#39;, &#39;HIST1H1B&#39;, &#39;HIST1H2AC&#39;, &#39;HIST1H2AH&#39;, &#39;HLA-DMA&#39;, &#39;HLA-DMB&#39;, &#39;HLA-DOB&#39;, &#39;HLA-DPA1&#39;, &#39;HLA-DPB1&#39;, &#39;HLA-DQA1&#39;, &#39;HLA-DQB1&#39;, &#39;HLA-DRB1&#39;, &#39;HMBOX1&#39;, &#39;HMGCL&#39;, &#39;HMGXB4&#39;, &#39;HNRNPH3&#39;, &#39;HOOK2&#39;, &#39;HOPX&#39;, &#39;HSPB11&#39;, &#39;HVCN1&#39;, &#39;ICAM2&#39;, &#39;ICOS&#39;, &#39;ICOSLG&#39;, &#39;ID2&#39;, &#39;IDUA&#39;, &#39;IFFO1&#39;, &#39;IFI27&#39;, &#39;IFIT1&#39;, &#39;IFIT2&#39;, &#39;IFITM3&#39;, &#39;IGFBP7&#39;, &#39;IGJ&#39;, &#39;IGLL5&#39;, &#39;IL1B&#39;, &#39;IL1RAP&#39;, &#39;IL23A&#39;, &#39;IL24&#39;, &#39;IL27RA&#39;, &#39;IL32&#39;, &#39;IL6&#39;, &#39;IL8&#39;, &#39;ILF3&#39;, &#39;ILF3-AS1&#39;, &#39;ING5&#39;, &#39;INSL3&#39;, &#39;INTS12&#39;, &#39;INTS2&#39;, &#39;IP6K1&#39;, &#39;IQCE&#39;, &#39;IRF8&#39;, &#39;IRF9&#39;, &#39;ISCA2&#39;, &#39;ISOC1&#39;, &#39;ITGA2B&#39;, &#39;ITGB7&#39;, &#39;ITM2A&#39;, &#39;ITSN2&#39;, &#39;JAKMIP1&#39;, &#39;JUND&#39;, &#39;KARS&#39;, &#39;KCNG1&#39;, &#39;KCNQ1OT1&#39;, &#39;KIAA0040&#39;, &#39;KIAA0125&#39;, &#39;KIAA0196&#39;, &#39;KIAA1430&#39;, &#39;KIF3A&#39;, &#39;KIF3C&#39;, &#39;KIF5B&#39;, &#39;KLHL24&#39;, &#39;KLRB1&#39;, &#39;KLRG1&#39;, &#39;KRBOX4&#39;, &#39;LAMP3&#39;, &#39;LARS&#39;, &#39;LAT2&#39;, &#39;LBR&#39;, &#39;LDLRAP1&#39;, &#39;LGALS1&#39;, &#39;LGALS2&#39;, &#39;LGALS3&#39;, &#39;LILRA4&#39;, &#39;LIN52&#39;, &#39;LINC00494&#39;, &#39;LINC00662&#39;, &#39;LINC00886&#39;, &#39;LINC00926&#39;, &#39;LINC00936&#39;, &#39;LINC01013&#39;, &#39;LIX1L&#39;, &#39;LONRF1&#39;, &#39;LPIN1&#39;, &#39;LRBA&#39;, &#39;LRRIQ3&#39;, &#39;LSM14A&#39;, &#39;LST1&#39;, &#39;LTB&#39;, &#39;LTV1&#39;, &#39;LUC7L&#39;, &#39;LUC7L3&#39;, &#39;LYAR&#39;, &#39;LYPD2&#39;, &#39;LYPLA1&#39;, &#39;LYRM4&#39;, &#39;LYSMD4&#39;, &#39;LZTS2&#39;, &#39;MADD&#39;, &#39;MAEA&#39;, &#39;MAGEH1&#39;, &#39;MAL&#39;, &#39;MALT1&#39;, &#39;MAP2K7&#39;, &#39;MARCKSL1&#39;, &#39;MCF2L&#39;, &#39;MCM3&#39;, &#39;MDS2&#39;, &#39;MED30&#39;, &#39;MED9&#39;, &#39;METTL21A&#39;, &#39;METTL3&#39;, &#39;METTL8&#39;, &#39;MFF&#39;, &#39;MFSD10&#39;, &#39;MIS18A&#39;, &#39;MKKS&#39;, &#39;MLLT11&#39;, &#39;MLLT6&#39;, &#39;MMADHC&#39;, &#39;MMP9&#39;, &#39;MNAT1&#39;, &#39;MOCS2&#39;, &#39;MORF4L2&#39;, &#39;MPHOSPH10&#39;, &#39;MRM1&#39;, &#39;MRPL1&#39;, &#39;MRPL19&#39;, &#39;MRPL42&#39;, &#39;MRPS12&#39;, &#39;MRPS33&#39;, &#39;MS4A1&#39;, &#39;MS4A6A&#39;, &#39;MTERFD2&#39;, &#39;MTIF2&#39;, &#39;MTRF1&#39;, &#39;MUM1&#39;, &#39;MYADM&#39;, &#39;MYCBP2&#39;, &#39;MYL9&#39;, &#39;MYO1E&#39;, &#39;MYOM2&#39;, &#39;MZB1&#39;, &#39;MZT1&#39;, &#39;NAA20&#39;, &#39;NAP1L4&#39;, &#39;NAPA-AS1&#39;, &#39;NARG2&#39;, &#39;NAT9&#39;, &#39;NBR1&#39;, &#39;NCOR2&#39;, &#39;NCR3&#39;, &#39;NDUFA10&#39;, &#39;NDUFA12&#39;, &#39;NECAB3&#39;, &#39;NEFH&#39;, &#39;NEK8&#39;, &#39;NELFB&#39;, &#39;NEMF&#39;, &#39;NFAT5&#39;, &#39;NFE2L2&#39;, &#39;NFIC&#39;, &#39;NFU1&#39;, &#39;NIT2&#39;, &#39;NKAP&#39;, &#39;NKG7&#39;, &#39;NKTR&#39;, &#39;NME3&#39;, &#39;NME6&#39;, &#39;NMNAT3&#39;, &#39;NNT-AS1&#39;, &#39;NOC4L&#39;, &#39;NOG&#39;, &#39;NOL11&#39;, &#39;NONO&#39;, &#39;NOP58&#39;, &#39;NPC2&#39;, &#39;NPHP3&#39;, &#39;NPRL2&#39;, &#39;NR2C1&#39;, &#39;NR3C1&#39;, &#39;NSA2&#39;, &#39;NT5C&#39;, &#39;NT5C3A&#39;, &#39;NUDCD1&#39;, &#39;NUDT16L1&#39;, &#39;NUP54&#39;, &#39;NXT2&#39;, &#39;OARD1&#39;, &#39;OAT&#39;, &#39;OBSCN&#39;, &#39;ODC1&#39;, &#39;ORAI1&#39;, &#39;ORC2&#39;, &#39;OSBPL1A&#39;, &#39;OSBPL7&#39;, &#39;OXLD1&#39;, &#39;P2RX5&#39;, &#39;P2RY10&#39;, &#39;PACS1&#39;, &#39;PACSIN2&#39;, &#39;PAICS&#39;, &#39;PARP1&#39;, &#39;PARS2&#39;, &#39;PASK&#39;, &#39;PAWR&#39;, &#39;PAXIP1-AS1&#39;, &#39;PBLD&#39;, &#39;PBRM1&#39;, &#39;PCNA&#39;, &#39;PCSK7&#39;, &#39;PDCD1&#39;, &#39;PDCD2L&#39;, &#39;PDE6B&#39;, &#39;PDIA3&#39;, &#39;PDIK1L&#39;, &#39;PDK2&#39;, &#39;PDXDC1&#39;, &#39;PDZD4&#39;, &#39;PEMT&#39;, &#39;PEX16&#39;, &#39;PEX26&#39;, &#39;PF4&#39;, &#39;PGM1&#39;, &#39;PGM2L1&#39;, &#39;PHACTR4&#39;, &#39;PHF12&#39;, &#39;PHF14&#39;, &#39;PHF3&#39;, &#39;PIGF&#39;, &#39;PIGU&#39;, &#39;PIGX&#39;, &#39;PIK3R1&#39;, &#39;PITHD1&#39;, &#39;PITPNA-AS1&#39;, &#39;PJA1&#39;, &#39;PKIG&#39;, &#39;PLA2G12A&#39;, &#39;PLCL1&#39;, &#39;PLD6&#39;, &#39;PLEKHA1&#39;, &#39;PLEKHA3&#39;, &#39;PLRG1&#39;, &#39;PMEPA1&#39;, &#39;PNOC&#39;, &#39;POLR2I&#39;, &#39;POLR2K&#39;, &#39;POLR3E&#39;, &#39;POMT1&#39;, &#39;PPA2&#39;, &#39;PPBP&#39;, &#39;PPIE&#39;, &#39;PPIG&#39;, &#39;PPIL2&#39;, &#39;PPIL4&#39;, &#39;PPP1R14A&#39;, &#39;PPP1R2&#39;, &#39;PPP2R1B&#39;, &#39;PPP6C&#39;, &#39;PPT2-EGFL8&#39;, &#39;PQBP1&#39;, &#39;PRAF2&#39;, &#39;PRDX1&#39;, &#39;PRELID2&#39;, &#39;PRF1&#39;, &#39;PRICKLE1&#39;, &#39;PRKACB&#39;, &#39;PRKCB&#39;, &#39;PRKD2&#39;, &#39;PRMT2&#39;, &#39;PRNP&#39;, &#39;PRPF31&#39;, &#39;PRPS2&#39;, &#39;PRR5&#39;, &#39;PSMD14&#39;, &#39;PTCRA&#39;, &#39;PTGDR&#39;, &#39;PTGDS&#39;, &#39;PTGES2&#39;, &#39;PTPN7&#39;, &#39;PURA&#39;, &#39;PWP1&#39;, &#39;PXMP4&#39;, &#39;PYCARD&#39;, &#39;R3HDM1&#39;, &#39;R3HDM2&#39;, &#39;RAB40C&#39;, &#39;RABEP2&#39;, &#39;RABL6&#39;, &#39;RAD51B&#39;, &#39;RALBP1&#39;, &#39;RALY&#39;, &#39;RASD1&#39;, &#39;RASGRP2&#39;, &#39;RBM25&#39;, &#39;RBM26-AS1&#39;, &#39;RBM39&#39;, &#39;RBM4&#39;, &#39;RBM48&#39;, &#39;RBM5&#39;, &#39;RBM7&#39;, &#39;RBPJ&#39;, &#39;RCE1&#39;, &#39;RCHY1&#39;, &#39;RCL1&#39;, &#39;RCN2&#39;, &#39;RDH14&#39;, &#39;RELB&#39;, &#39;REXO2&#39;, &#39;RFC1&#39;, &#39;RFC5&#39;, &#39;RFNG&#39;, &#39;RFPL2&#39;, &#39;RGS14&#39;, &#39;RIC3&#39;, &#39;RIOK1&#39;, &#39;RIOK2&#39;, &#39;RNF113A&#39;, &#39;RNF125&#39;, &#39;RNF139&#39;, &#39;RNF14&#39;, &#39;RNF168&#39;, &#39;RNF187&#39;, &#39;RNF213&#39;, &#39;RNF25&#39;, &#39;RNF26&#39;, &#39;RORA&#39;, &#39;RP1-28O10.1&#39;, &#39;RP11-1055B8.7&#39;, &#39;RP11-138A9.2&#39;, &#39;RP11-141B14.1&#39;, &#39;RP11-142C4.6&#39;, &#39;RP11-162G10.5&#39;, &#39;RP11-164H13.1&#39;, &#39;RP11-178G16.4&#39;, &#39;RP11-18H21.1&#39;, &#39;RP11-211G3.2&#39;, &#39;RP11-219B17.1&#39;, &#39;RP11-219B4.7&#39;, &#39;RP11-252A24.3&#39;, &#39;RP11-291B21.2&#39;, &#39;RP11-314N13.3&#39;, &#39;RP11-324I22.4&#39;, &#39;RP11-349A22.5&#39;, &#39;RP11-378J18.3&#39;, &#39;RP11-390B4.5&#39;, &#39;RP11-398C13.6&#39;, &#39;RP11-400F19.6&#39;, &#39;RP11-421L21.3&#39;, &#39;RP11-428G5.5&#39;, &#39;RP11-432I5.1&#39;, &#39;RP11-468E2.4&#39;, &#39;RP11-488C13.5&#39;, &#39;RP11-493L12.4&#39;, &#39;RP11-527L4.5&#39;, &#39;RP11-545I5.3&#39;, &#39;RP11-589C21.6&#39;, &#39;RP11-5C23.1&#39;, &#39;RP11-701P16.5&#39;, &#39;RP11-706O15.1&#39;, &#39;RP11-70P17.1&#39;, &#39;RP11-727F15.9&#39;, &#39;RP11-798G7.6&#39;, &#39;RP11-879F14.2&#39;, &#39;RP11-950C14.3&#39;, &#39;RP3-325F22.5&#39;, &#39;RP5-1073O3.7&#39;, &#39;RP5-827C21.4&#39;, &#39;RP5-887A10.1&#39;, &#39;RPH3A&#39;, &#39;RPL39L&#39;, &#39;RPL7L1&#39;, &#39;RPN2&#39;, &#39;RPS6KL1&#39;, &#39;RPUSD2&#39;, &#39;RRAGC&#39;, &#39;RRS1&#39;, &#39;RUNDC1&#39;, &#39;S100A11&#39;, &#39;S100A12&#39;, &#39;S100A8&#39;, &#39;S100B&#39;, &#39;SAFB2&#39;, &#39;SAMD1&#39;, &#39;SAMD3&#39;, &#39;SAMSN1&#39;, &#39;SARDH&#39;, &#39;SARS&#39;, &#39;SAT1&#39;, &#39;SCAI&#39;, &#39;SCAPER&#39;, &#39;SCGB3A1&#39;, &#39;SCPEP1&#39;, &#39;SDCCAG8&#39;, &#39;SDPR&#39;, &#39;SEC61A2&#39;, &#39;SELL&#39;, &#39;SEPT11&#39;, &#39;SERAC1&#39;, &#39;SETD1B&#39;, &#39;SF3B1&#39;, &#39;SF3B5&#39;, &#39;SH3GLB1&#39;, &#39;SH3KBP1&#39;, &#39;SHOC2&#39;, &#39;SHPK-1&#39;, &#39;SIAH2&#39;, &#39;SIRPG&#39;, &#39;SIRT1&#39;, &#39;SIVA1&#39;, &#39;SLA&#39;, &#39;SLBP&#39;, &#39;SLC22A4&#39;, &#39;SLC25A11&#39;, &#39;SLC25A12&#39;, &#39;SLC25A14&#39;, &#39;SLC27A1&#39;, &#39;SLC2A13&#39;, &#39;SLC35A2&#39;, &#39;SLC48A1&#39;, &#39;SLFN5&#39;, &#39;SMARCA4&#39;, &#39;SMARCC2&#39;, &#39;SMC2&#39;, &#39;SMCHD1&#39;, &#39;SMDT1&#39;, &#39;SMIM14&#39;, &#39;SMIM7&#39;, &#39;SNAP47&#39;, &#39;SNHG12&#39;, &#39;SNHG8&#39;, &#39;SNTA1&#39;, &#39;SNX29P2&#39;, &#39;SOX13&#39;, &#39;SPARC&#39;, &#39;SPATA7&#39;, &#39;SPG7&#39;, &#39;SPIB&#39;, &#39;SPIN1&#39;, &#39;SPOCD1&#39;, &#39;SPON2&#39;, &#39;SPSB2&#39;, &#39;SREBF1&#39;, &#39;SRM&#39;, &#39;SRP9&#39;, &#39;SRSF6&#39;, &#39;SSBP1&#39;, &#39;ST3GAL2&#39;, &#39;STAMBP&#39;, &#39;STAU2&#39;, &#39;STK17A&#39;, &#39;STK38&#39;, &#39;STMN1&#39;, &#39;STOML2&#39;, &#39;STUB1&#39;, &#39;STX16&#39;, &#39;STX18&#39;, &#39;SUCLG2&#39;, &#39;SUOX&#39;, &#39;SURF1&#39;, &#39;SURF6&#39;, &#39;SWAP70&#39;, &#39;SYCE1&#39;, &#39;SYP&#39;, &#39;SYVN1&#39;, &#39;TACR2&#39;, &#39;TADA2A&#39;, &#39;TAF10&#39;, &#39;TAF12&#39;, &#39;TAF1D&#39;, &#39;TAL1&#39;, &#39;TALDO1&#39;, &#39;TAPBP&#39;, &#39;TARSL2&#39;, &#39;TASP1&#39;, &#39;TBC1D15&#39;, &#39;TBCK&#39;, &#39;TBXA2R&#39;, &#39;TCEAL4&#39;, &#39;TCEAL8&#39;, &#39;TCL1A&#39;, &#39;TCL1B&#39;, &#39;TCP1&#39;, &#39;TDG&#39;, &#39;TERF2IP&#39;, &#39;TGFBRAP1&#39;, &#39;THAP2&#39;, &#39;THEM4&#39;, &#39;THOC7&#39;, &#39;THUMPD3&#39;, &#39;THYN1&#39;, &#39;TIGIT&#39;, &#39;TIMM10B&#39;, &#39;TMEM116&#39;, &#39;TMEM138&#39;, &#39;TMEM140&#39;, &#39;TMEM14B&#39;, &#39;TMEM165&#39;, &#39;TMEM177&#39;, &#39;TMEM194A&#39;, &#39;TMEM219&#39;, &#39;TMEM242&#39;, &#39;TMEM40&#39;, &#39;TMEM60&#39;, &#39;TMEM80&#39;, &#39;TMEM87A&#39;, &#39;TMEM87B&#39;, &#39;TMEM91&#39;, &#39;TMTC2&#39;, &#39;TMX2&#39;, &#39;TMX3&#39;, &#39;TNFRSF17&#39;, &#39;TNFRSF25&#39;, &#39;TNFRSF4&#39;, &#39;TNFRSF9&#39;, &#39;TNFSF10&#39;, &#39;TOP1MT&#39;, &#39;TOP2B&#39;, &#39;TRABD2A&#39;, &#39;TRAF3IP3&#39;, &#39;TRAPPC12-AS1&#39;, &#39;TRAPPC3&#39;, &#39;TREML1&#39;, &#39;TRIM23&#39;, &#39;TRIP12&#39;, &#39;TRIT1&#39;, &#39;TRMT61A&#39;, &#39;TRPM4&#39;, &#39;TSC22D1&#39;, &#39;TSPAN15&#39;, &#39;TSSC1&#39;, &#39;TTC1&#39;, &#39;TTC14&#39;, &#39;TTC3&#39;, &#39;TTC8&#39;, &#39;TTN-AS1&#39;, &#39;TUBB1&#39;, &#39;TUBG2&#39;, &#39;TYMP&#39;, &#39;TYROBP&#39;, &#39;U2SURP&#39;, &#39;UBA5&#39;, &#39;UBAC2&#39;, &#39;UBE2D2&#39;, &#39;UBE2D4&#39;, &#39;UBE2K&#39;, &#39;UBE2Q1&#39;, &#39;UBE3A&#39;, &#39;UBIAD1&#39;, &#39;UBLCP1&#39;, &#39;UBXN4&#39;, &#39;UCK1&#39;, &#39;UNC45A&#39;, &#39;UQCC1&#39;, &#39;URB2&#39;, &#39;URGCP&#39;, &#39;USP30&#39;, &#39;USP33&#39;, &#39;USP36&#39;, &#39;USP38&#39;, &#39;USP5&#39;, &#39;USP7&#39;, &#39;VAMP5&#39;, &#39;VDAC3&#39;, &#39;VIPR1&#39;, &#39;VPS13A&#39;, &#39;VPS13C&#39;, &#39;VPS25&#39;, &#39;VPS26B&#39;, &#39;VPS28&#39;, &#39;VTI1A&#39;, &#39;VTI1B&#39;, &#39;WARS2&#39;, &#39;WBP2NL&#39;, &#39;WDR55&#39;, &#39;WDR91&#39;, &#39;WDYHV1&#39;, &#39;WNK1&#39;, &#39;WTAP&#39;, &#39;XCL2&#39;, &#39;XPOT&#39;, &#39;XRRA1&#39;, &#39;XXbac-BPG299F13.17&#39;, &#39;YEATS2&#39;, &#39;YES1&#39;, &#39;YPEL2&#39;, &#39;YPEL3&#39;, &#39;YTHDF2&#39;, &#39;ZAP70&#39;, &#39;ZBED5-AS1&#39;, &#39;ZBP1&#39;, &#39;ZC3H15&#39;, &#39;ZCCHC11&#39;, &#39;ZCCHC9&#39;, &#39;ZFAND4&#39;, &#39;ZNF175&#39;, &#39;ZNF232&#39;, &#39;ZNF256&#39;, &#39;ZNF263&#39;, &#39;ZNF276&#39;, &#39;ZNF32&#39;, &#39;ZNF350&#39;, &#39;ZNF436&#39;, &#39;ZNF45&#39;, &#39;ZNF493&#39;, &#39;ZNF503&#39;, &#39;ZNF528&#39;, &#39;ZNF559&#39;, &#39;ZNF561&#39;, &#39;ZNF587B&#39;, &#39;ZNF594&#39;, &#39;ZNF653&#39;, &#39;ZNF682&#39;, &#39;ZNF688&#39;, &#39;ZNF718&#39;, &#39;ZNF747&#39;, &#39;ZNF799&#39;, &#39;ZNF836&#39;, &#39;ZNF92&#39;, &#39;ZRANB3&#39;, &#39;ZSWIM6&#39;, &#39;ZUFSP&#39;]
2024-07-08 15:00:47,207 [INFO] Parameter: probe_length_min = 40
2024-07-08 15:00:47,208 [INFO] Parameter: probe_length_max = 45
2024-07-08 15:00:47,209 [INFO] Parameter: files_fasta_oligo_database = [&#39;output_genomic_region_generator_ncbi/annotation/exon_annotation_source-NCBI_species-Homo_sapiens_annotation_release-110_genome_assemly-GRCh38.p14.fna&#39;, &#39;output_genomic_region_generator_ncbi/annotation/exon_exon_junction_annotation_source-NCBI_species-Homo_sapiens_annotation_release-110_genome_assemly-GRCh38.p14.fna&#39;]
2024-07-08 15:00:47,210 [INFO] Parameter: min_probes_per_gene = 3
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<script type="application/vnd.jupyter.widget-view+json">{"model_id": "456d1e2d0cc8415e90c2a28419ea6040", "version_major": 2, "version_minor": 0}</script></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area rendered_html docutils container">
<pre style="white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace"></pre></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area rendered_html docutils container">
<pre style="white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace">
</pre></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
2024-07-08 17:17:26,931 [INFO] Step - Create Database: database contains 26663811 oligos from 887 regions.
2024-07-08 17:17:27,020 [DEBUG] handle_msg[456d1e2d0cc8415e90c2a28419ea6040]({&#39;header&#39;: {&#39;date&#39;: datetime.datetime(2024, 7, 8, 15, 17, 27, 15000, tzinfo=tzutc()), &#39;msg_id&#39;: &#39;bffef3a3-1674-47de-8b8f-9180581b259b&#39;, &#39;msg_type&#39;: &#39;comm_msg&#39;, &#39;session&#39;: &#39;f15cb79e-4700-4ef8-aed8-65224aba1be6&#39;, &#39;username&#39;: &#39;5bb12fba-c395-4a0a-afbc-31e5784dd153&#39;, &#39;version&#39;: &#39;5.2&#39;}, &#39;msg_id&#39;: &#39;bffef3a3-1674-47de-8b8f-9180581b259b&#39;, &#39;msg_type&#39;: &#39;comm_msg&#39;, &#39;parent_header&#39;: {}, &#39;metadata&#39;: {}, &#39;content&#39;: {&#39;comm_id&#39;: &#39;456d1e2d0cc8415e90c2a28419ea6040&#39;, &#39;data&#39;: {&#39;method&#39;: &#39;update&#39;, &#39;state&#39;: {&#39;outputs&#39;: [{&#39;output_type&#39;: &#39;display_data&#39;, &#39;data&#39;: {&#39;text/plain&#39;: &#39;  \x1b[35m100%\x1b[0m \x1b[36mDatabase Loading\x1b[0m \x1b[38;2;114;156;31m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\x1b[0m \x1b[32m887/887\x1b[0m \x1b[33m1:12:17\x1b[0m &lt; \x1b[36m0:00:00\x1b[0m\n&#39;, &#39;text/html&#39;: &#39;&lt;pre style=&#34;white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,\&#39;DejaVu Sans Mono\&#39;,consolas,\&#39;Courier New\&#39;,monospace&#34;&gt;  &lt;span style=&#34;color: #800080; text-decoration-color: #800080&#34;&gt;100%&lt;/span&gt; &lt;span style=&#34;color: #008080; text-decoration-color: #008080&#34;&gt;Database Loading&lt;/span&gt; &lt;span style=&#34;color: #729c1f; text-decoration-color: #729c1f&#34;&gt;━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━&lt;/span&gt; &lt;span style=&#34;color: #008000; text-decoration-color: #008000&#34;&gt;887/887&lt;/span&gt; &lt;span style=&#34;color: #808000; text-decoration-color: #808000&#34;&gt;1:12:17&lt;/span&gt; &amp;lt; &lt;span style=&#34;color: #008080; text-decoration-color: #008080&#34;&gt;0:00:00&lt;/span&gt;\n&lt;/pre&gt;\n&#39;}, &#39;metadata&#39;: {}}]}, &#39;buffer_paths&#39;: []}}, &#39;buffers&#39;: []})
</pre></div></div>
</div>
</section>
<section id="Refine-Database-with-Property-Filters">
<h3>Refine Database with Property Filters<a class="headerlink" href="#Refine-Database-with-Property-Filters" title="Permalink to this heading"></a></h3>
<p>In order to create experiment-specific probes, we have to apply several filter to each probe, e.g. melting temperature or GC content filters.</p>
<p><strong>Parameters for Probe Sequences:</strong> - <code class="docutils literal notranslate"><span class="pre">probe_GC_content_min</span></code>: minimum GC content of oligos - <code class="docutils literal notranslate"><span class="pre">probe_GC_content_max</span></code>: maximum GC content of oligos - <code class="docutils literal notranslate"><span class="pre">probe_Tm_min</span></code>: minimum melting temperature of oligos - <code class="docutils literal notranslate"><span class="pre">probe_Tm_max</span></code>: maximum melting temperature of oligos - <code class="docutils literal notranslate"><span class="pre">homopolymeric_base_n</span></code>: minimum number of nucleotides to consider it a homopolymeric run per base and all probes with such a homopolymeric run are filtered out, e.g. if A: 5 is set then the sequence “CTGGTAAAAACTGGA” is filtered
out, but the sequence “CTGGTAAAACTGGA” is kept</p>
<p><strong>Parameters for Padlock Arms:</strong> - <code class="docutils literal notranslate"><span class="pre">arm_Tm_dif_max</span></code>: maximum melting temperature difference of both arms (difference shouldn’t be higher than 5! But range is not super important, the lower the better) - <code class="docutils literal notranslate"><span class="pre">arm_length_min</span></code>: minimum length of each arm - <code class="docutils literal notranslate"><span class="pre">arm_Tm_min</span></code>: minimum melting temperature of each arm - <code class="docutils literal notranslate"><span class="pre">arm_Tm_max</span></code>: maximum melting temperature of each arm</p>
<p><strong>Parameters for Detection Oligos:</strong> - <code class="docutils literal notranslate"><span class="pre">min_thymines</span></code>: minimal number of Thymines in detection oligo. - <code class="docutils literal notranslate"><span class="pre">detect_oligo_length_min</span></code>: minimum length of detection probe - <code class="docutils literal notranslate"><span class="pre">detect_oligo_length_max</span></code>: maximum length of detection probe</p>
<p><em>Note: The melting temperature is used in 2 different stages (probe and detection oligo design), where a few parameters are shared and the others differ. Parameters for melting temperature -&gt; for more information on parameters, see:</em><a class="reference external" href="https://biopython.org/docs/1.75/api/Bio.SeqUtils.MeltingTemp.html#Bio.SeqUtils.MeltingTemp.Tm_NN">here</a></p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[16]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1">## Define parameters</span>
<span class="c1"># probe sequence</span>
<span class="n">probe_GC_content_min</span> <span class="o">=</span> <span class="mi">40</span>
<span class="n">probe_GC_content_max</span> <span class="o">=</span> <span class="mi">60</span>
<span class="n">probe_Tm_min</span> <span class="o">=</span> <span class="mi">65</span>
<span class="n">probe_Tm_max</span> <span class="o">=</span> <span class="mi">75</span>
<span class="n">homopolymeric_base_n</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;A&quot;</span><span class="p">:</span> <span class="mi">5</span><span class="p">,</span> <span class="s2">&quot;T&quot;</span><span class="p">:</span> <span class="mi">5</span><span class="p">,</span> <span class="s2">&quot;C&quot;</span><span class="p">:</span> <span class="mi">5</span><span class="p">,</span> <span class="s2">&quot;G&quot;</span><span class="p">:</span> <span class="mi">5</span><span class="p">}</span>

<span class="c1"># padlock arms</span>
<span class="n">arm_Tm_dif_max</span> <span class="o">=</span> <span class="mi">2</span>
<span class="n">arm_length_min</span> <span class="o">=</span> <span class="mi">10</span>
<span class="n">arm_Tm_min</span> <span class="o">=</span> <span class="mi">50</span>
<span class="n">arm_Tm_max</span> <span class="o">=</span> <span class="mi">60</span>

<span class="c1"># detection oligos</span>
<span class="n">min_thymines</span> <span class="o">=</span> <span class="mi">2</span>
<span class="n">U_distance</span> <span class="o">=</span> <span class="mi">5</span>
<span class="n">detect_oligo_length_min</span> <span class="o">=</span> <span class="mi">15</span>
<span class="n">detect_oligo_length_max</span> <span class="o">=</span> <span class="mi">40</span>
<span class="n">detect_oligo_Tm_opt</span> <span class="o">=</span> <span class="mi">56</span>


<span class="n">Tm_parameters_probe</span> <span class="o">=</span> <span class="p">{</span>
    <span class="s2">&quot;check&quot;</span><span class="p">:</span> <span class="kc">True</span><span class="p">,</span>  <span class="c1"># default</span>
    <span class="s2">&quot;strict&quot;</span><span class="p">:</span> <span class="kc">True</span><span class="p">,</span>  <span class="c1"># default</span>
    <span class="s2">&quot;c_seq&quot;</span><span class="p">:</span> <span class="kc">None</span><span class="p">,</span>  <span class="c1"># default</span>
    <span class="s2">&quot;shift&quot;</span><span class="p">:</span> <span class="mi">0</span><span class="p">,</span>  <span class="c1"># default</span>
    <span class="s2">&quot;nn_table&quot;</span><span class="p">:</span> <span class="nb">getattr</span><span class="p">(</span><span class="n">mt</span><span class="p">,</span> <span class="s2">&quot;DNA_NN3&quot;</span><span class="p">),</span> <span class="c1"># Allawi &amp; SantaLucia (1997)</span>
    <span class="s2">&quot;tmm_table&quot;</span><span class="p">:</span> <span class="nb">getattr</span><span class="p">(</span><span class="n">mt</span><span class="p">,</span> <span class="s2">&quot;DNA_TMM1&quot;</span><span class="p">),</span> <span class="c1">#default</span>
    <span class="s2">&quot;imm_table&quot;</span><span class="p">:</span> <span class="nb">getattr</span><span class="p">(</span><span class="n">mt</span><span class="p">,</span> <span class="s2">&quot;DNA_IMM1&quot;</span><span class="p">),</span> <span class="c1">#default</span>
    <span class="s2">&quot;de_table&quot;</span><span class="p">:</span> <span class="nb">getattr</span><span class="p">(</span><span class="n">mt</span><span class="p">,</span> <span class="s2">&quot;DNA_DE1&quot;</span><span class="p">),</span> <span class="c1">#default</span>
    <span class="s2">&quot;dnac1&quot;</span><span class="p">:</span> <span class="mi">50</span><span class="p">,</span>  <span class="c1"># [nM]</span>
    <span class="s2">&quot;dnac2&quot;</span><span class="p">:</span> <span class="mi">0</span><span class="p">,</span>  <span class="c1"># [nM]</span>
    <span class="s2">&quot;selfcomp&quot;</span><span class="p">:</span> <span class="kc">False</span><span class="p">,</span>  <span class="c1"># default</span>
    <span class="s2">&quot;saltcorr&quot;</span><span class="p">:</span> <span class="mi">7</span><span class="p">,</span>  <span class="c1"># Owczarzy et al. (2008)</span>
    <span class="s2">&quot;Na&quot;</span><span class="p">:</span> <span class="mi">39</span><span class="p">,</span>  <span class="c1"># [mM]</span>
    <span class="s2">&quot;K&quot;</span><span class="p">:</span> <span class="mi">75</span><span class="p">,</span>  <span class="c1"># [mM]</span>
    <span class="s2">&quot;Tris&quot;</span><span class="p">:</span> <span class="mi">20</span><span class="p">,</span>  <span class="c1"># [mM]</span>
    <span class="s2">&quot;Mg&quot;</span><span class="p">:</span> <span class="mi">10</span><span class="p">,</span>  <span class="c1"># [mM]</span>
    <span class="s2">&quot;dNTPs&quot;</span><span class="p">:</span> <span class="mi">0</span><span class="p">,</span>  <span class="c1"># [mM] default</span>
<span class="p">}</span>

<span class="n">Tm_chem_correction_param_probe</span> <span class="o">=</span> <span class="p">{</span>
    <span class="s2">&quot;DMSO&quot;</span><span class="p">:</span> <span class="mi">0</span><span class="p">,</span>  <span class="c1"># default</span>
    <span class="s2">&quot;fmd&quot;</span><span class="p">:</span> <span class="mi">20</span><span class="p">,</span>
    <span class="s2">&quot;DMSOfactor&quot;</span><span class="p">:</span> <span class="mf">0.75</span><span class="p">,</span>  <span class="c1"># default</span>
    <span class="s2">&quot;fmdfactor&quot;</span><span class="p">:</span> <span class="mf">0.65</span><span class="p">,</span>  <span class="c1"># default</span>
    <span class="s2">&quot;fmdmethod&quot;</span><span class="p">:</span> <span class="mi">1</span><span class="p">,</span>  <span class="c1"># default</span>
    <span class="s2">&quot;GC&quot;</span><span class="p">:</span> <span class="kc">None</span><span class="p">,</span>  <span class="c1"># default</span>
<span class="p">}</span>

<span class="n">Tm_salt_correction_param_probe</span> <span class="o">=</span> <span class="kc">None</span>
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[17]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1">####### Load existing database #######</span>
<span class="c1"># dir_database = os.path.join(dir_output, &quot;db_probes/1_db_probes_initial&quot;)</span>
<span class="c1"># probe_database = OligoDatabase(min_oligos_per_region=min_probes_per_gene, write_regions_with_insufficient_oligos=True, lru_db_max_in_memory=n_jobs*2+2, database_name=&quot;db_probes&quot;, dir_output=dir_output, n_jobs=n_jobs)</span>
<span class="c1"># probe_database.load_database(dir_database=dir_database, region_ids=gene_ids, database_overwrite=True)</span>

<span class="c1">## Apply property filters</span>
<span class="n">probe_database</span><span class="p">,</span> <span class="n">file_database</span> <span class="o">=</span> <span class="n">pipeline</span><span class="o">.</span><span class="n">filter_by_property</span><span class="p">(</span>
    <span class="n">oligo_database</span><span class="o">=</span><span class="n">probe_database</span><span class="p">,</span>
    <span class="n">probe_GC_content_min</span><span class="o">=</span><span class="n">probe_GC_content_min</span><span class="p">,</span>
    <span class="n">probe_GC_content_max</span><span class="o">=</span><span class="n">probe_GC_content_max</span><span class="p">,</span>
    <span class="n">probe_Tm_min</span><span class="o">=</span><span class="n">probe_Tm_min</span><span class="p">,</span>
    <span class="n">probe_Tm_max</span><span class="o">=</span><span class="n">probe_Tm_max</span><span class="p">,</span>
    <span class="n">detect_oligo_length_min</span><span class="o">=</span><span class="n">detect_oligo_length_min</span><span class="p">,</span>
    <span class="n">detect_oligo_length_max</span><span class="o">=</span><span class="n">detect_oligo_length_max</span><span class="p">,</span>
    <span class="n">min_thymines</span><span class="o">=</span><span class="n">min_thymines</span><span class="p">,</span>
    <span class="n">arm_Tm_dif_max</span><span class="o">=</span><span class="n">arm_Tm_dif_max</span><span class="p">,</span>
    <span class="n">arm_length_min</span><span class="o">=</span><span class="n">arm_length_min</span><span class="p">,</span>
    <span class="n">arm_Tm_min</span><span class="o">=</span><span class="n">arm_Tm_min</span><span class="p">,</span>
    <span class="n">arm_Tm_max</span><span class="o">=</span><span class="n">arm_Tm_max</span><span class="p">,</span>
    <span class="n">homopolymeric_base_n</span><span class="o">=</span><span class="n">homopolymeric_base_n</span><span class="p">,</span>
    <span class="n">Tm_parameters_probe</span><span class="o">=</span><span class="n">Tm_parameters_probe</span><span class="p">,</span>
    <span class="n">Tm_chem_correction_param_probe</span><span class="o">=</span><span class="n">Tm_chem_correction_param_probe</span><span class="p">,</span>
    <span class="n">Tm_salt_correction_param_probe</span><span class="o">=</span><span class="n">Tm_salt_correction_param_probe</span><span class="p">,</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
2024-07-08 17:41:57,263 [INFO] Parameters Property Filters:
2024-07-08 17:41:57,266 [INFO] Function: filter_by_property
2024-07-08 17:41:57,268 [INFO] Parameter: oligo_database = &lt;oligo_designer_toolsuite.database._oligo_database.OligoDatabase object at 0x10d917550&gt;
2024-07-08 17:41:57,270 [INFO] Parameter: probe_GC_content_min = 40
2024-07-08 17:41:57,272 [INFO] Parameter: probe_GC_content_max = 60
2024-07-08 17:41:57,273 [INFO] Parameter: probe_Tm_min = 65
2024-07-08 17:41:57,274 [INFO] Parameter: probe_Tm_max = 75
2024-07-08 17:41:57,275 [INFO] Parameter: detect_oligo_length_min = 15
2024-07-08 17:41:57,276 [INFO] Parameter: detect_oligo_length_max = 40
2024-07-08 17:41:57,278 [INFO] Parameter: min_thymines = 2
2024-07-08 17:41:57,279 [INFO] Parameter: arm_Tm_dif_max = 2
2024-07-08 17:41:57,279 [INFO] Parameter: arm_length_min = 10
2024-07-08 17:41:57,280 [INFO] Parameter: arm_Tm_min = 50
2024-07-08 17:41:57,282 [INFO] Parameter: arm_Tm_max = 60
2024-07-08 17:41:57,284 [INFO] Parameter: homopolymeric_base_n = {&#39;A&#39;: 5, &#39;T&#39;: 5, &#39;C&#39;: 5, &#39;G&#39;: 5}
2024-07-08 17:41:57,286 [INFO] Parameter: Tm_parameters_probe = {&#39;check&#39;: True, &#39;strict&#39;: True, &#39;c_seq&#39;: None, &#39;shift&#39;: 0, &#39;nn_table&#39;: {&#39;init&#39;: (0, 0), &#39;init_A/T&#39;: (2.3, 4.1), &#39;init_G/C&#39;: (0.1, -2.8), &#39;init_oneG/C&#39;: (0, 0), &#39;init_allA/T&#39;: (0, 0), &#39;init_5T/A&#39;: (0, 0), &#39;sym&#39;: (0, -1.4), &#39;AA/TT&#39;: (-7.9, -22.2), &#39;AT/TA&#39;: (-7.2, -20.4), &#39;TA/AT&#39;: (-7.2, -21.3), &#39;CA/GT&#39;: (-8.5, -22.7), &#39;GT/CA&#39;: (-8.4, -22.4), &#39;CT/GA&#39;: (-7.8, -21.0), &#39;GA/CT&#39;: (-8.2, -22.2), &#39;CG/GC&#39;: (-10.6, -27.2), &#39;GC/CG&#39;: (-9.8, -24.4), &#39;GG/CC&#39;: (-8.0, -19.9)}, &#39;tmm_table&#39;: {&#39;AA/TA&#39;: (-3.1, -7.8), &#39;TA/AA&#39;: (-2.5, -6.3), &#39;CA/GA&#39;: (-4.3, -10.7), &#39;GA/CA&#39;: (-8.0, -22.5), &#39;AC/TC&#39;: (-0.1, 0.5), &#39;TC/AC&#39;: (-0.7, -1.3), &#39;CC/GC&#39;: (-2.1, -5.1), &#39;GC/CC&#39;: (-3.9, -10.6), &#39;AG/TG&#39;: (-1.1, -2.1), &#39;TG/AG&#39;: (-1.1, -2.7), &#39;CG/GG&#39;: (-3.8, -9.5), &#39;GG/CG&#39;: (-0.7, -19.2), &#39;AT/TT&#39;: (-2.4, -6.5), &#39;TT/AT&#39;: (-3.2, -8.9), &#39;CT/GT&#39;: (-6.1, -16.9), &#39;GT/CT&#39;: (-7.4, -21.2), &#39;AA/TC&#39;: (-1.6, -4.0), &#39;AC/TA&#39;: (-1.8, -3.8), &#39;CA/GC&#39;: (-2.6, -5.9), &#39;CC/GA&#39;: (-2.7, -6.0), &#39;GA/CC&#39;: (-5.0, -13.8), &#39;GC/CA&#39;: (-3.2, -7.1), &#39;TA/AC&#39;: (-2.3, -5.9), &#39;TC/AA&#39;: (-2.7, -7.0), &#39;AC/TT&#39;: (-0.9, -1.7), &#39;AT/TC&#39;: (-2.3, -6.3), &#39;CC/GT&#39;: (-3.2, -8.0), &#39;CT/GC&#39;: (-3.9, -10.6), &#39;GC/CT&#39;: (-4.9, -13.5), &#39;GT/CC&#39;: (-3.0, -7.8), &#39;TC/AT&#39;: (-2.5, -6.3), &#39;TT/AC&#39;: (-0.7, -1.2), &#39;AA/TG&#39;: (-1.9, -4.4), &#39;AG/TA&#39;: (-2.5, -5.9), &#39;CA/GG&#39;: (-3.9, -9.6), &#39;CG/GA&#39;: (-6.0, -15.5), &#39;GA/CG&#39;: (-4.3, -11.1), &#39;GG/CA&#39;: (-4.6, -11.4), &#39;TA/AG&#39;: (-2.0, -4.7), &#39;TG/AA&#39;: (-2.4, -5.8), &#39;AG/TT&#39;: (-3.2, -8.7), &#39;AT/TG&#39;: (-3.5, -9.4), &#39;CG/GT&#39;: (-3.8, -9.0), &#39;CT/GG&#39;: (-6.6, -18.7), &#39;GG/CT&#39;: (-5.7, -15.9), &#39;GT/CG&#39;: (-5.9, -16.1), &#39;TG/AT&#39;: (-3.9, -10.5), &#39;TT/AG&#39;: (-3.6, -9.8)}, &#39;imm_table&#39;: {&#39;AG/TT&#39;: (1.0, 0.9), &#39;AT/TG&#39;: (-2.5, -8.3), &#39;CG/GT&#39;: (-4.1, -11.7), &#39;CT/GG&#39;: (-2.8, -8.0), &#39;GG/CT&#39;: (3.3, 10.4), &#39;GG/TT&#39;: (5.8, 16.3), &#39;GT/CG&#39;: (-4.4, -12.3), &#39;GT/TG&#39;: (4.1, 9.5), &#39;TG/AT&#39;: (-0.1, -1.7), &#39;TG/GT&#39;: (-1.4, -6.2), &#39;TT/AG&#39;: (-1.3, -5.3), &#39;AA/TG&#39;: (-0.6, -2.3), &#39;AG/TA&#39;: (-0.7, -2.3), &#39;CA/GG&#39;: (-0.7, -2.3), &#39;CG/GA&#39;: (-4.0, -13.2), &#39;GA/CG&#39;: (-0.6, -1.0), &#39;GG/CA&#39;: (0.5, 3.2), &#39;TA/AG&#39;: (0.7, 0.7), &#39;TG/AA&#39;: (3.0, 7.4), &#39;AC/TT&#39;: (0.7, 0.2), &#39;AT/TC&#39;: (-1.2, -6.2), &#39;CC/GT&#39;: (-0.8, -4.5), &#39;CT/GC&#39;: (-1.5, -6.1), &#39;GC/CT&#39;: (2.3, 5.4), &#39;GT/CC&#39;: (5.2, 13.5), &#39;TC/AT&#39;: (1.2, 0.7), &#39;TT/AC&#39;: (1.0, 0.7), &#39;AA/TC&#39;: (2.3, 4.6), &#39;AC/TA&#39;: (5.3, 14.6), &#39;CA/GC&#39;: (1.9, 3.7), &#39;CC/GA&#39;: (0.6, -0.6), &#39;GA/CC&#39;: (5.2, 14.2), &#39;GC/CA&#39;: (-0.7, -3.8), &#39;TA/AC&#39;: (3.4, 8.0), &#39;TC/AA&#39;: (7.6, 20.2), &#39;AA/TA&#39;: (1.2, 1.7), &#39;CA/GA&#39;: (-0.9, -4.2), &#39;GA/CA&#39;: (-2.9, -9.8), &#39;TA/AA&#39;: (4.7, 12.9), &#39;AC/TC&#39;: (0.0, -4.4), &#39;CC/GC&#39;: (-1.5, -7.2), &#39;GC/CC&#39;: (3.6, 8.9), &#39;TC/AC&#39;: (6.1, 16.4), &#39;AG/TG&#39;: (-3.1, -9.5), &#39;CG/GG&#39;: (-4.9, -15.3), &#39;GG/CG&#39;: (-6.0, -15.8), &#39;TG/AG&#39;: (1.6, 3.6), &#39;AT/TT&#39;: (-2.7, -10.8), &#39;CT/GT&#39;: (-5.0, -15.8), &#39;GT/CT&#39;: (-2.2, -8.4), &#39;TT/AT&#39;: (0.2, -1.5), &#39;AI/TC&#39;: (-8.9, -25.5), &#39;TI/AC&#39;: (-5.9, -17.4), &#39;AC/TI&#39;: (-8.8, -25.4), &#39;TC/AI&#39;: (-4.9, -13.9), &#39;CI/GC&#39;: (-5.4, -13.7), &#39;GI/CC&#39;: (-6.8, -19.1), &#39;CC/GI&#39;: (-8.3, -23.8), &#39;GC/CI&#39;: (-5.0, -12.6), &#39;AI/TA&#39;: (-8.3, -25.0), &#39;TI/AA&#39;: (-3.4, -11.2), &#39;AA/TI&#39;: (-0.7, -2.6), &#39;TA/AI&#39;: (-1.3, -4.6), &#39;CI/GA&#39;: (2.6, 8.9), &#39;GI/CA&#39;: (-7.8, -21.1), &#39;CA/GI&#39;: (-7.0, -20.0), &#39;GA/CI&#39;: (-7.6, -20.2), &#39;AI/TT&#39;: (0.49, -0.7), &#39;TI/AT&#39;: (-6.5, -22.0), &#39;AT/TI&#39;: (-5.6, -18.7), &#39;TT/AI&#39;: (-0.8, -4.3), &#39;CI/GT&#39;: (-1.0, -2.4), &#39;GI/CT&#39;: (-3.5, -10.6), &#39;CT/GI&#39;: (0.1, -1.0), &#39;GT/CI&#39;: (-4.3, -12.1), &#39;AI/TG&#39;: (-4.9, -15.8), &#39;TI/AG&#39;: (-1.9, -8.5), &#39;AG/TI&#39;: (0.1, -1.8), &#39;TG/AI&#39;: (1.0, 1.0), &#39;CI/GG&#39;: (7.1, 21.3), &#39;GI/CG&#39;: (-1.1, -3.2), &#39;CG/GI&#39;: (5.8, 16.9), &#39;GG/CI&#39;: (-7.6, -22.0), &#39;AI/TI&#39;: (-3.3, -11.9), &#39;TI/AI&#39;: (0.1, -2.3), &#39;CI/GI&#39;: (1.3, 3.0), &#39;GI/CI&#39;: (-0.5, -1.3)}, &#39;de_table&#39;: {&#39;AA/.T&#39;: (0.2, 2.3), &#39;AC/.G&#39;: (-6.3, -17.1), &#39;AG/.C&#39;: (-3.7, -10.0), &#39;AT/.A&#39;: (-2.9, -7.6), &#39;CA/.T&#39;: (0.6, 3.3), &#39;CC/.G&#39;: (-4.4, -12.6), &#39;CG/.C&#39;: (-4.0, -11.9), &#39;CT/.A&#39;: (-4.1, -13.0), &#39;GA/.T&#39;: (-1.1, -1.6), &#39;GC/.G&#39;: (-5.1, -14.0), &#39;GG/.C&#39;: (-3.9, -10.9), &#39;GT/.A&#39;: (-4.2, -15.0), &#39;TA/.T&#39;: (-6.9, -20.0), &#39;TC/.G&#39;: (-4.0, -10.9), &#39;TG/.C&#39;: (-4.9, -13.8), &#39;TT/.A&#39;: (-0.2, -0.5), &#39;.A/AT&#39;: (-0.7, -0.8), &#39;.C/AG&#39;: (-2.1, -3.9), &#39;.G/AC&#39;: (-5.9, -16.5), &#39;.T/AA&#39;: (-0.5, -1.1), &#39;.A/CT&#39;: (4.4, 14.9), &#39;.C/CG&#39;: (-0.2, -0.1), &#39;.G/CC&#39;: (-2.6, -7.4), &#39;.T/CA&#39;: (4.7, 14.2), &#39;.A/GT&#39;: (-1.6, -3.6), &#39;.C/GG&#39;: (-3.9, -11.2), &#39;.G/GC&#39;: (-3.2, -10.4), &#39;.T/GA&#39;: (-4.1, -13.1), &#39;.A/TT&#39;: (2.9, 10.4), &#39;.C/TG&#39;: (-4.4, -13.1), &#39;.G/TC&#39;: (-5.2, -15.0), &#39;.T/TA&#39;: (-3.8, -12.6)}, &#39;dnac1&#39;: 50, &#39;dnac2&#39;: 0, &#39;selfcomp&#39;: False, &#39;saltcorr&#39;: 7, &#39;Na&#39;: 39, &#39;K&#39;: 75, &#39;Tris&#39;: 20, &#39;Mg&#39;: 10, &#39;dNTPs&#39;: 0}
2024-07-08 17:41:57,288 [INFO] Parameter: Tm_chem_correction_param_probe = {&#39;DMSO&#39;: 0, &#39;fmd&#39;: 20, &#39;DMSOfactor&#39;: 0.75, &#39;fmdfactor&#39;: 0.65, &#39;fmdmethod&#39;: 1, &#39;GC&#39;: None}
2024-07-08 17:41:57,288 [INFO] Parameter: Tm_salt_correction_param_probe = None
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<script type="application/vnd.jupyter.widget-view+json">{"model_id": "af295b15415d4d37bd0c05502a7e2966", "version_major": 2, "version_minor": 0}</script></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area rendered_html docutils container">
<pre style="white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace"></pre></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area rendered_html docutils container">
<pre style="white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace">
</pre></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
2024-07-08 18:52:48,260 [INFO] Step - Property Filters: database contains 3468382 oligos from 887 regions.
2024-07-08 18:52:48,357 [DEBUG] handle_msg[af295b15415d4d37bd0c05502a7e2966]({&#39;header&#39;: {&#39;date&#39;: datetime.datetime(2024, 7, 8, 16, 52, 48, 323000, tzinfo=tzutc()), &#39;msg_id&#39;: &#39;94375a4a-c46b-4ea8-aa8f-b1e5092c97ae&#39;, &#39;msg_type&#39;: &#39;comm_msg&#39;, &#39;session&#39;: &#39;f15cb79e-4700-4ef8-aed8-65224aba1be6&#39;, &#39;username&#39;: &#39;5bb12fba-c395-4a0a-afbc-31e5784dd153&#39;, &#39;version&#39;: &#39;5.2&#39;}, &#39;msg_id&#39;: &#39;94375a4a-c46b-4ea8-aa8f-b1e5092c97ae&#39;, &#39;msg_type&#39;: &#39;comm_msg&#39;, &#39;parent_header&#39;: {}, &#39;metadata&#39;: {}, &#39;content&#39;: {&#39;comm_id&#39;: &#39;af295b15415d4d37bd0c05502a7e2966&#39;, &#39;data&#39;: {&#39;method&#39;: &#39;update&#39;, &#39;state&#39;: {&#39;outputs&#39;: [{&#39;output_type&#39;: &#39;display_data&#39;, &#39;data&#39;: {&#39;text/plain&#39;: &#39;  \x1b[35m100%\x1b[0m \x1b[36mProperty Filter\x1b[0m \x1b[38;2;114;156;31m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\x1b[0m \x1b[32m887/887\x1b[0m \x1b[33m0:59:00\x1b[0m &lt; \x1b[36m0:00:00\x1b[0m\n&#39;, &#39;text/html&#39;: &#39;&lt;pre style=&#34;white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,\&#39;DejaVu Sans Mono\&#39;,consolas,\&#39;Courier New\&#39;,monospace&#34;&gt;  &lt;span style=&#34;color: #800080; text-decoration-color: #800080&#34;&gt;100%&lt;/span&gt; &lt;span style=&#34;color: #008080; text-decoration-color: #008080&#34;&gt;Property Filter&lt;/span&gt; &lt;span style=&#34;color: #729c1f; text-decoration-color: #729c1f&#34;&gt;━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━&lt;/span&gt; &lt;span style=&#34;color: #008000; text-decoration-color: #008000&#34;&gt;887/887&lt;/span&gt; &lt;span style=&#34;color: #808000; text-decoration-color: #808000&#34;&gt;0:59:00&lt;/span&gt; &amp;lt; &lt;span style=&#34;color: #008080; text-decoration-color: #008080&#34;&gt;0:00:00&lt;/span&gt;\n&lt;/pre&gt;\n&#39;}, &#39;metadata&#39;: {}}]}, &#39;buffer_paths&#39;: []}}, &#39;buffers&#39;: []})
</pre></div></div>
</div>
</section>
<section id="Refine-Database-with-Specificity-Filters">
<h3>Refine Database with Specificity Filters<a class="headerlink" href="#Refine-Database-with-Specificity-Filters" title="Permalink to this heading"></a></h3>
<p>In order to create target specific probes, we have to apply several filter to each probe to lower the probability of off-target effects and cross-hybridization.</p>
<div class="line-block">
<div class="line"><strong>Parameters for BlastN Crosshybridization filter</strong></div>
<div class="line">In this filter, the probes are aligned against the probe database, to detect possible cross-hybridization events. As filter policy, we remove the probe coming from the region with more probes, to avoid deleting too many probes in this step. - <code class="docutils literal notranslate"><span class="pre">cross_hybridization_blastn_search_parameters</span></code>: parameters for the BlastN search - <code class="docutils literal notranslate"><span class="pre">cross_hybridization_blastn_hit_parameters</span></code>: parameters for the evaluation of BlastN hits</div>
</div>
<div class="line-block">
<div class="line"><strong>Parameters for BlastN Specificity filter</strong></div>
<div class="line">In this filter, the probes are aligned against a reference fasta file, to detect possible off-target events. In our case, we use the transcriptome that we generated before with the Genomic Region Generator, but this can be chaned upon needs. - <code class="docutils literal notranslate"><span class="pre">files_fasta_reference_database</span></code>: fasta file with sequences used as reference for the specificity filters. Multiple fasta files can be used as reference. Hint: use the genomic_region_generator pipeline to create fasta files of genomic regions of
interest - <code class="docutils literal notranslate"><span class="pre">ligation_region_size</span></code>: size of the seed region around the ligation site for blast seed region filter; set to 0 if ligation region should not be considered for blast search - <code class="docutils literal notranslate"><span class="pre">specificity_blastn_search_parameters</span></code>: parameters for the BlastN search - <code class="docutils literal notranslate"><span class="pre">specificity_blastn_hit_parameters</span></code>: parameters for the evaluation of BlastN hits</div>
</div>
<p><em>Note: Depending on the number of genes, this step might be time and memory consuming. For high number of genes, you might want to run this step on a bigger machine!</em></p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[18]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1">## Define parameters</span>
<span class="c1"># Cross-hybridization filter</span>
<span class="n">cross_hybridization_blastn_search_parameters</span> <span class="o">=</span> <span class="p">{</span>
  <span class="s2">&quot;perc_identity&quot;</span><span class="p">:</span> <span class="mi">80</span><span class="p">,</span>
  <span class="s2">&quot;strand&quot;</span><span class="p">:</span> <span class="s2">&quot;minus&quot;</span><span class="p">,</span> <span class="c1"># this parameter is fixed</span>
  <span class="s2">&quot;word_size&quot;</span><span class="p">:</span> <span class="mi">10</span><span class="p">,</span>
  <span class="s2">&quot;dust&quot;</span><span class="p">:</span> <span class="s2">&quot;no&quot;</span><span class="p">,</span>
  <span class="s2">&quot;soft_masking&quot;</span><span class="p">:</span> <span class="s2">&quot;false&quot;</span><span class="p">,</span>
  <span class="s2">&quot;max_target_seqs&quot;</span><span class="p">:</span> <span class="mi">10</span><span class="p">,</span>
<span class="p">}</span>

<span class="n">cross_hybridization_blastn_hit_parameters</span> <span class="o">=</span> <span class="p">{</span>
  <span class="s2">&quot;coverage&quot;</span><span class="p">:</span> <span class="mi">80</span> <span class="c1"># can be turned into min_alignment_length</span>
<span class="p">}</span>


<span class="c1"># Specificity Filter</span>
<span class="n">files_fasta_reference_database</span> <span class="o">=</span> <span class="n">fasta_files</span>
<span class="n">ligation_region_size</span> <span class="o">=</span> <span class="mi">5</span>
<span class="n">specificity_blastn_search_parameters</span> <span class="o">=</span> <span class="p">{</span>
  <span class="s2">&quot;perc_identity&quot;</span><span class="p">:</span> <span class="mi">80</span><span class="p">,</span>
  <span class="s2">&quot;strand&quot;</span><span class="p">:</span> <span class="s2">&quot;minus&quot;</span><span class="p">,</span> <span class="c1"># this parameter is fixed, however if reference is whole genome, consider using &quot;both&quot;</span>
  <span class="s2">&quot;word_size&quot;</span><span class="p">:</span> <span class="mi">10</span><span class="p">,</span>
  <span class="s2">&quot;dust&quot;</span><span class="p">:</span> <span class="s2">&quot;no&quot;</span><span class="p">,</span>
  <span class="s2">&quot;soft_masking&quot;</span><span class="p">:</span> <span class="s2">&quot;false&quot;</span><span class="p">,</span>
  <span class="s2">&quot;max_target_seqs&quot;</span><span class="p">:</span> <span class="mi">10</span><span class="p">,</span>
  <span class="s2">&quot;max_hsps&quot;</span><span class="p">:</span> <span class="mi">1000</span><span class="p">,</span>
<span class="p">}</span>
<span class="n">specificity_blastn_hit_parameters</span> <span class="o">=</span> <span class="p">{</span>
  <span class="s2">&quot;coverage&quot;</span><span class="p">:</span> <span class="mi">50</span> <span class="c1"># can be turned into min_alignment_length if more meaningful</span>
<span class="p">}</span>
<br/></pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[20]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1">####### Load existing database #######</span>
<span class="c1"># dir_database = os.path.join(dir_output, &quot;db_probes/2_db_probes_property_filter&quot;)</span>
<span class="c1"># probe_database = OligoDatabase(min_oligos_per_region=min_probes_per_gene, write_regions_with_insufficient_oligos=True, lru_db_max_in_memory=n_jobs*2+2, database_name=&quot;db_probes&quot;, dir_output=dir_output, n_jobs=n_jobs)</span>
<span class="c1"># probe_database.load_database(dir_database=dir_database, region_ids=gene_ids, database_overwrite=True)</span>

<span class="c1">## Apply specificity filters</span>
<span class="n">probe_database</span><span class="p">,</span> <span class="n">file_database</span> <span class="o">=</span> <span class="n">pipeline</span><span class="o">.</span><span class="n">filter_by_specificity</span><span class="p">(</span>
    <span class="n">oligo_database</span><span class="o">=</span><span class="n">probe_database</span><span class="p">,</span>
    <span class="n">files_fasta_reference_database</span><span class="o">=</span><span class="n">files_fasta_reference_database</span><span class="p">,</span>
    <span class="n">specificity_blastn_search_parameters</span><span class="o">=</span><span class="n">specificity_blastn_search_parameters</span><span class="p">,</span>
    <span class="n">specificity_blastn_hit_parameters</span><span class="o">=</span><span class="n">specificity_blastn_hit_parameters</span><span class="p">,</span>
    <span class="n">cross_hybridization_blastn_search_parameters</span><span class="o">=</span><span class="n">cross_hybridization_blastn_search_parameters</span><span class="p">,</span>
    <span class="n">cross_hybridization_blastn_hit_parameters</span><span class="o">=</span><span class="n">cross_hybridization_blastn_hit_parameters</span><span class="p">,</span>
    <span class="n">ligation_region_size</span><span class="o">=</span><span class="n">ligation_region_size</span><span class="p">,</span>
    <span class="n">arm_Tm_dif_max</span><span class="o">=</span><span class="n">arm_Tm_dif_max</span><span class="p">,</span>
    <span class="n">arm_length_min</span><span class="o">=</span><span class="n">arm_length_min</span><span class="p">,</span>
    <span class="n">arm_Tm_min</span><span class="o">=</span><span class="n">arm_Tm_min</span><span class="p">,</span>
    <span class="n">arm_Tm_max</span><span class="o">=</span><span class="n">arm_Tm_max</span><span class="p">,</span>
    <span class="n">Tm_parameters_probe</span><span class="o">=</span><span class="n">Tm_parameters_probe</span><span class="p">,</span>
    <span class="n">Tm_chem_correction_param_probe</span><span class="o">=</span><span class="n">Tm_chem_correction_param_probe</span><span class="p">,</span>
    <span class="n">Tm_salt_correction_param_probe</span><span class="o">=</span><span class="n">Tm_salt_correction_param_probe</span><span class="p">,</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
2024-07-08 19:46:51,968 [INFO] Parameters Specificity Filters:
2024-07-08 19:46:51,969 [INFO] Function: filter_by_specificity
2024-07-08 19:46:51,972 [INFO] Parameter: oligo_database = &lt;oligo_designer_toolsuite.database._oligo_database.OligoDatabase object at 0x10d917550&gt;
2024-07-08 19:46:51,973 [INFO] Parameter: files_fasta_reference_database = [&#39;output_genomic_region_generator_ncbi/annotation/exon_annotation_source-NCBI_species-Homo_sapiens_annotation_release-110_genome_assemly-GRCh38.p14.fna&#39;, &#39;output_genomic_region_generator_ncbi/annotation/exon_exon_junction_annotation_source-NCBI_species-Homo_sapiens_annotation_release-110_genome_assemly-GRCh38.p14.fna&#39;]
2024-07-08 19:46:51,975 [INFO] Parameter: specificity_blastn_search_parameters = {&#39;perc_identity&#39;: 80, &#39;strand&#39;: &#39;minus&#39;, &#39;word_size&#39;: 10, &#39;dust&#39;: &#39;no&#39;, &#39;soft_masking&#39;: &#39;false&#39;, &#39;max_target_seqs&#39;: 10, &#39;max_hsps&#39;: 1000}
2024-07-08 19:46:51,977 [INFO] Parameter: specificity_blastn_hit_parameters = {&#39;coverage&#39;: 50}
2024-07-08 19:46:51,978 [INFO] Parameter: cross_hybridization_blastn_search_parameters = {&#39;perc_identity&#39;: 80, &#39;strand&#39;: &#39;minus&#39;, &#39;word_size&#39;: 10, &#39;dust&#39;: &#39;no&#39;, &#39;soft_masking&#39;: &#39;false&#39;, &#39;max_target_seqs&#39;: 10}
2024-07-08 19:46:51,979 [INFO] Parameter: cross_hybridization_blastn_hit_parameters = {&#39;coverage&#39;: 80}
2024-07-08 19:46:51,980 [INFO] Parameter: ligation_region_size = 5
2024-07-08 19:46:51,984 [INFO] Parameter: arm_Tm_dif_max = 2
2024-07-08 19:46:51,986 [INFO] Parameter: arm_length_min = 10
2024-07-08 19:46:51,988 [INFO] Parameter: arm_Tm_min = 50
2024-07-08 19:46:51,990 [INFO] Parameter: arm_Tm_max = 60
2024-07-08 19:46:51,992 [INFO] Parameter: Tm_parameters_probe = {&#39;check&#39;: True, &#39;strict&#39;: True, &#39;c_seq&#39;: None, &#39;shift&#39;: 0, &#39;nn_table&#39;: {&#39;init&#39;: (0, 0), &#39;init_A/T&#39;: (2.3, 4.1), &#39;init_G/C&#39;: (0.1, -2.8), &#39;init_oneG/C&#39;: (0, 0), &#39;init_allA/T&#39;: (0, 0), &#39;init_5T/A&#39;: (0, 0), &#39;sym&#39;: (0, -1.4), &#39;AA/TT&#39;: (-7.9, -22.2), &#39;AT/TA&#39;: (-7.2, -20.4), &#39;TA/AT&#39;: (-7.2, -21.3), &#39;CA/GT&#39;: (-8.5, -22.7), &#39;GT/CA&#39;: (-8.4, -22.4), &#39;CT/GA&#39;: (-7.8, -21.0), &#39;GA/CT&#39;: (-8.2, -22.2), &#39;CG/GC&#39;: (-10.6, -27.2), &#39;GC/CG&#39;: (-9.8, -24.4), &#39;GG/CC&#39;: (-8.0, -19.9)}, &#39;tmm_table&#39;: {&#39;AA/TA&#39;: (-3.1, -7.8), &#39;TA/AA&#39;: (-2.5, -6.3), &#39;CA/GA&#39;: (-4.3, -10.7), &#39;GA/CA&#39;: (-8.0, -22.5), &#39;AC/TC&#39;: (-0.1, 0.5), &#39;TC/AC&#39;: (-0.7, -1.3), &#39;CC/GC&#39;: (-2.1, -5.1), &#39;GC/CC&#39;: (-3.9, -10.6), &#39;AG/TG&#39;: (-1.1, -2.1), &#39;TG/AG&#39;: (-1.1, -2.7), &#39;CG/GG&#39;: (-3.8, -9.5), &#39;GG/CG&#39;: (-0.7, -19.2), &#39;AT/TT&#39;: (-2.4, -6.5), &#39;TT/AT&#39;: (-3.2, -8.9), &#39;CT/GT&#39;: (-6.1, -16.9), &#39;GT/CT&#39;: (-7.4, -21.2), &#39;AA/TC&#39;: (-1.6, -4.0), &#39;AC/TA&#39;: (-1.8, -3.8), &#39;CA/GC&#39;: (-2.6, -5.9), &#39;CC/GA&#39;: (-2.7, -6.0), &#39;GA/CC&#39;: (-5.0, -13.8), &#39;GC/CA&#39;: (-3.2, -7.1), &#39;TA/AC&#39;: (-2.3, -5.9), &#39;TC/AA&#39;: (-2.7, -7.0), &#39;AC/TT&#39;: (-0.9, -1.7), &#39;AT/TC&#39;: (-2.3, -6.3), &#39;CC/GT&#39;: (-3.2, -8.0), &#39;CT/GC&#39;: (-3.9, -10.6), &#39;GC/CT&#39;: (-4.9, -13.5), &#39;GT/CC&#39;: (-3.0, -7.8), &#39;TC/AT&#39;: (-2.5, -6.3), &#39;TT/AC&#39;: (-0.7, -1.2), &#39;AA/TG&#39;: (-1.9, -4.4), &#39;AG/TA&#39;: (-2.5, -5.9), &#39;CA/GG&#39;: (-3.9, -9.6), &#39;CG/GA&#39;: (-6.0, -15.5), &#39;GA/CG&#39;: (-4.3, -11.1), &#39;GG/CA&#39;: (-4.6, -11.4), &#39;TA/AG&#39;: (-2.0, -4.7), &#39;TG/AA&#39;: (-2.4, -5.8), &#39;AG/TT&#39;: (-3.2, -8.7), &#39;AT/TG&#39;: (-3.5, -9.4), &#39;CG/GT&#39;: (-3.8, -9.0), &#39;CT/GG&#39;: (-6.6, -18.7), &#39;GG/CT&#39;: (-5.7, -15.9), &#39;GT/CG&#39;: (-5.9, -16.1), &#39;TG/AT&#39;: (-3.9, -10.5), &#39;TT/AG&#39;: (-3.6, -9.8)}, &#39;imm_table&#39;: {&#39;AG/TT&#39;: (1.0, 0.9), &#39;AT/TG&#39;: (-2.5, -8.3), &#39;CG/GT&#39;: (-4.1, -11.7), &#39;CT/GG&#39;: (-2.8, -8.0), &#39;GG/CT&#39;: (3.3, 10.4), &#39;GG/TT&#39;: (5.8, 16.3), &#39;GT/CG&#39;: (-4.4, -12.3), &#39;GT/TG&#39;: (4.1, 9.5), &#39;TG/AT&#39;: (-0.1, -1.7), &#39;TG/GT&#39;: (-1.4, -6.2), &#39;TT/AG&#39;: (-1.3, -5.3), &#39;AA/TG&#39;: (-0.6, -2.3), &#39;AG/TA&#39;: (-0.7, -2.3), &#39;CA/GG&#39;: (-0.7, -2.3), &#39;CG/GA&#39;: (-4.0, -13.2), &#39;GA/CG&#39;: (-0.6, -1.0), &#39;GG/CA&#39;: (0.5, 3.2), &#39;TA/AG&#39;: (0.7, 0.7), &#39;TG/AA&#39;: (3.0, 7.4), &#39;AC/TT&#39;: (0.7, 0.2), &#39;AT/TC&#39;: (-1.2, -6.2), &#39;CC/GT&#39;: (-0.8, -4.5), &#39;CT/GC&#39;: (-1.5, -6.1), &#39;GC/CT&#39;: (2.3, 5.4), &#39;GT/CC&#39;: (5.2, 13.5), &#39;TC/AT&#39;: (1.2, 0.7), &#39;TT/AC&#39;: (1.0, 0.7), &#39;AA/TC&#39;: (2.3, 4.6), &#39;AC/TA&#39;: (5.3, 14.6), &#39;CA/GC&#39;: (1.9, 3.7), &#39;CC/GA&#39;: (0.6, -0.6), &#39;GA/CC&#39;: (5.2, 14.2), &#39;GC/CA&#39;: (-0.7, -3.8), &#39;TA/AC&#39;: (3.4, 8.0), &#39;TC/AA&#39;: (7.6, 20.2), &#39;AA/TA&#39;: (1.2, 1.7), &#39;CA/GA&#39;: (-0.9, -4.2), &#39;GA/CA&#39;: (-2.9, -9.8), &#39;TA/AA&#39;: (4.7, 12.9), &#39;AC/TC&#39;: (0.0, -4.4), &#39;CC/GC&#39;: (-1.5, -7.2), &#39;GC/CC&#39;: (3.6, 8.9), &#39;TC/AC&#39;: (6.1, 16.4), &#39;AG/TG&#39;: (-3.1, -9.5), &#39;CG/GG&#39;: (-4.9, -15.3), &#39;GG/CG&#39;: (-6.0, -15.8), &#39;TG/AG&#39;: (1.6, 3.6), &#39;AT/TT&#39;: (-2.7, -10.8), &#39;CT/GT&#39;: (-5.0, -15.8), &#39;GT/CT&#39;: (-2.2, -8.4), &#39;TT/AT&#39;: (0.2, -1.5), &#39;AI/TC&#39;: (-8.9, -25.5), &#39;TI/AC&#39;: (-5.9, -17.4), &#39;AC/TI&#39;: (-8.8, -25.4), &#39;TC/AI&#39;: (-4.9, -13.9), &#39;CI/GC&#39;: (-5.4, -13.7), &#39;GI/CC&#39;: (-6.8, -19.1), &#39;CC/GI&#39;: (-8.3, -23.8), &#39;GC/CI&#39;: (-5.0, -12.6), &#39;AI/TA&#39;: (-8.3, -25.0), &#39;TI/AA&#39;: (-3.4, -11.2), &#39;AA/TI&#39;: (-0.7, -2.6), &#39;TA/AI&#39;: (-1.3, -4.6), &#39;CI/GA&#39;: (2.6, 8.9), &#39;GI/CA&#39;: (-7.8, -21.1), &#39;CA/GI&#39;: (-7.0, -20.0), &#39;GA/CI&#39;: (-7.6, -20.2), &#39;AI/TT&#39;: (0.49, -0.7), &#39;TI/AT&#39;: (-6.5, -22.0), &#39;AT/TI&#39;: (-5.6, -18.7), &#39;TT/AI&#39;: (-0.8, -4.3), &#39;CI/GT&#39;: (-1.0, -2.4), &#39;GI/CT&#39;: (-3.5, -10.6), &#39;CT/GI&#39;: (0.1, -1.0), &#39;GT/CI&#39;: (-4.3, -12.1), &#39;AI/TG&#39;: (-4.9, -15.8), &#39;TI/AG&#39;: (-1.9, -8.5), &#39;AG/TI&#39;: (0.1, -1.8), &#39;TG/AI&#39;: (1.0, 1.0), &#39;CI/GG&#39;: (7.1, 21.3), &#39;GI/CG&#39;: (-1.1, -3.2), &#39;CG/GI&#39;: (5.8, 16.9), &#39;GG/CI&#39;: (-7.6, -22.0), &#39;AI/TI&#39;: (-3.3, -11.9), &#39;TI/AI&#39;: (0.1, -2.3), &#39;CI/GI&#39;: (1.3, 3.0), &#39;GI/CI&#39;: (-0.5, -1.3)}, &#39;de_table&#39;: {&#39;AA/.T&#39;: (0.2, 2.3), &#39;AC/.G&#39;: (-6.3, -17.1), &#39;AG/.C&#39;: (-3.7, -10.0), &#39;AT/.A&#39;: (-2.9, -7.6), &#39;CA/.T&#39;: (0.6, 3.3), &#39;CC/.G&#39;: (-4.4, -12.6), &#39;CG/.C&#39;: (-4.0, -11.9), &#39;CT/.A&#39;: (-4.1, -13.0), &#39;GA/.T&#39;: (-1.1, -1.6), &#39;GC/.G&#39;: (-5.1, -14.0), &#39;GG/.C&#39;: (-3.9, -10.9), &#39;GT/.A&#39;: (-4.2, -15.0), &#39;TA/.T&#39;: (-6.9, -20.0), &#39;TC/.G&#39;: (-4.0, -10.9), &#39;TG/.C&#39;: (-4.9, -13.8), &#39;TT/.A&#39;: (-0.2, -0.5), &#39;.A/AT&#39;: (-0.7, -0.8), &#39;.C/AG&#39;: (-2.1, -3.9), &#39;.G/AC&#39;: (-5.9, -16.5), &#39;.T/AA&#39;: (-0.5, -1.1), &#39;.A/CT&#39;: (4.4, 14.9), &#39;.C/CG&#39;: (-0.2, -0.1), &#39;.G/CC&#39;: (-2.6, -7.4), &#39;.T/CA&#39;: (4.7, 14.2), &#39;.A/GT&#39;: (-1.6, -3.6), &#39;.C/GG&#39;: (-3.9, -11.2), &#39;.G/GC&#39;: (-3.2, -10.4), &#39;.T/GA&#39;: (-4.1, -13.1), &#39;.A/TT&#39;: (2.9, 10.4), &#39;.C/TG&#39;: (-4.4, -13.1), &#39;.G/TC&#39;: (-5.2, -15.0), &#39;.T/TA&#39;: (-3.8, -12.6)}, &#39;dnac1&#39;: 50, &#39;dnac2&#39;: 0, &#39;selfcomp&#39;: False, &#39;saltcorr&#39;: 7, &#39;Na&#39;: 39, &#39;K&#39;: 75, &#39;Tris&#39;: 20, &#39;Mg&#39;: 10, &#39;dNTPs&#39;: 0}
2024-07-08 19:46:51,996 [INFO] Parameter: Tm_chem_correction_param_probe = {&#39;DMSO&#39;: 0, &#39;fmd&#39;: 20, &#39;DMSOfactor&#39;: 0.75, &#39;fmdfactor&#39;: 0.65, &#39;fmdmethod&#39;: 1, &#39;GC&#39;: None}
2024-07-08 19:46:51,997 [INFO] Parameter: Tm_salt_correction_param_probe = None
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<script type="application/vnd.jupyter.widget-view+json">{"model_id": "ecdf5993189c4082b846e0bd4d424bc1", "version_major": 2, "version_minor": 0}</script></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area rendered_html docutils container">
<pre style="white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace"></pre></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area rendered_html docutils container">
<pre style="white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace">
</pre></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<script type="application/vnd.jupyter.widget-view+json">{"model_id": "f78d8fb19a4845f5bb559e02ee379443", "version_major": 2, "version_minor": 0}</script></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area rendered_html docutils container">
<pre style="white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace"></pre></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area rendered_html docutils container">
<pre style="white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace">
</pre></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<script type="application/vnd.jupyter.widget-view+json">{"model_id": "086807ee363b4bbea301fac1c93c5444", "version_major": 2, "version_minor": 0}</script></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area rendered_html docutils container">
<pre style="white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace"></pre></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area rendered_html docutils container">
<pre style="white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace">
</pre></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
2024-07-08 21:42:02,456 [INFO] Step - Specificity Filters: database contains 408735 oligos from 806 regions.
2024-07-08 21:42:02,542 [DEBUG] handle_msg[ecdf5993189c4082b846e0bd4d424bc1]({&#39;header&#39;: {&#39;date&#39;: datetime.datetime(2024, 7, 8, 19, 42, 2, 520000, tzinfo=tzutc()), &#39;msg_id&#39;: &#39;3d0f9311-1134-4051-a004-bce437d0daed&#39;, &#39;msg_type&#39;: &#39;comm_msg&#39;, &#39;session&#39;: &#39;f15cb79e-4700-4ef8-aed8-65224aba1be6&#39;, &#39;username&#39;: &#39;5bb12fba-c395-4a0a-afbc-31e5784dd153&#39;, &#39;version&#39;: &#39;5.2&#39;}, &#39;msg_id&#39;: &#39;3d0f9311-1134-4051-a004-bce437d0daed&#39;, &#39;msg_type&#39;: &#39;comm_msg&#39;, &#39;parent_header&#39;: {}, &#39;metadata&#39;: {}, &#39;content&#39;: {&#39;comm_id&#39;: &#39;ecdf5993189c4082b846e0bd4d424bc1&#39;, &#39;data&#39;: {&#39;method&#39;: &#39;update&#39;, &#39;state&#39;: {&#39;outputs&#39;: [{&#39;output_type&#39;: &#39;display_data&#39;, &#39;data&#39;: {&#39;text/plain&#39;: &#39;  \x1b[35m100%\x1b[0m \x1b[36mSpecificity Filter: Exact Match\x1b[0m \x1b[38;2;114;156;31m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\x1b[0m \x1b[32m887/887\x1b[0m \x1b[33m0:04:30\x1b[0m &lt; \x1b[36m0:00:00\x1b[0m\n&#39;, &#39;text/html&#39;: &#39;&lt;pre style=&#34;white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,\&#39;DejaVu Sans Mono\&#39;,consolas,\&#39;Courier New\&#39;,monospace&#34;&gt;  &lt;span style=&#34;color: #800080; text-decoration-color: #800080&#34;&gt;100%&lt;/span&gt; &lt;span style=&#34;color: #008080; text-decoration-color: #008080&#34;&gt;Specificity Filter: Exact Match&lt;/span&gt; &lt;span style=&#34;color: #729c1f; text-decoration-color: #729c1f&#34;&gt;━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━&lt;/span&gt; &lt;span style=&#34;color: #008000; text-decoration-color: #008000&#34;&gt;887/887&lt;/span&gt; &lt;span style=&#34;color: #808000; text-decoration-color: #808000&#34;&gt;0:04:30&lt;/span&gt; &amp;lt; &lt;span style=&#34;color: #008080; text-decoration-color: #008080&#34;&gt;0:00:00&lt;/span&gt;\n&lt;/pre&gt;\n&#39;}, &#39;metadata&#39;: {}}]}, &#39;buffer_paths&#39;: []}}, &#39;buffers&#39;: []})
2024-07-08 21:42:02,547 [DEBUG] handle_msg[f78d8fb19a4845f5bb559e02ee379443]({&#39;header&#39;: {&#39;date&#39;: datetime.datetime(2024, 7, 8, 19, 42, 2, 521000, tzinfo=tzutc()), &#39;msg_id&#39;: &#39;00a4381b-1e14-42e5-a94a-246c14c0bee4&#39;, &#39;msg_type&#39;: &#39;comm_msg&#39;, &#39;session&#39;: &#39;f15cb79e-4700-4ef8-aed8-65224aba1be6&#39;, &#39;username&#39;: &#39;5bb12fba-c395-4a0a-afbc-31e5784dd153&#39;, &#39;version&#39;: &#39;5.2&#39;}, &#39;msg_id&#39;: &#39;00a4381b-1e14-42e5-a94a-246c14c0bee4&#39;, &#39;msg_type&#39;: &#39;comm_msg&#39;, &#39;parent_header&#39;: {}, &#39;metadata&#39;: {}, &#39;content&#39;: {&#39;comm_id&#39;: &#39;f78d8fb19a4845f5bb559e02ee379443&#39;, &#39;data&#39;: {&#39;method&#39;: &#39;update&#39;, &#39;state&#39;: {&#39;outputs&#39;: [{&#39;output_type&#39;: &#39;display_data&#39;, &#39;data&#39;: {&#39;text/plain&#39;: &#39;  \x1b[35m100%\x1b[0m \x1b[36mSpecificity Filter: Blastn Specificity\x1b[0m \x1b[38;2;114;156;31m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\x1b[0m \x1b[32m887/887\x1b[0m \x1b[33m1:08:54\x1b[0m &lt; \x1b[36m0:00:00\x1b[0m\n&#39;, &#39;text/html&#39;: &#39;&lt;pre style=&#34;white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,\&#39;DejaVu Sans Mono\&#39;,consolas,\&#39;Courier New\&#39;,monospace&#34;&gt;  &lt;span style=&#34;color: #800080; text-decoration-color: #800080&#34;&gt;100%&lt;/span&gt; &lt;span style=&#34;color: #008080; text-decoration-color: #008080&#34;&gt;Specificity Filter: Blastn Specificity&lt;/span&gt; &lt;span style=&#34;color: #729c1f; text-decoration-color: #729c1f&#34;&gt;━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━&lt;/span&gt; &lt;span style=&#34;color: #008000; text-decoration-color: #008000&#34;&gt;887/887&lt;/span&gt; &lt;span style=&#34;color: #808000; text-decoration-color: #808000&#34;&gt;1:08:54&lt;/span&gt; &amp;lt; &lt;span style=&#34;color: #008080; text-decoration-color: #008080&#34;&gt;0:00:00&lt;/span&gt;\n&lt;/pre&gt;\n&#39;}, &#39;metadata&#39;: {}}]}, &#39;buffer_paths&#39;: []}}, &#39;buffers&#39;: []})
2024-07-08 21:42:02,552 [DEBUG] handle_msg[086807ee363b4bbea301fac1c93c5444]({&#39;header&#39;: {&#39;date&#39;: datetime.datetime(2024, 7, 8, 19, 42, 2, 521000, tzinfo=tzutc()), &#39;msg_id&#39;: &#39;4d438cb6-e1cd-4b07-9718-cb0949ab2096&#39;, &#39;msg_type&#39;: &#39;comm_msg&#39;, &#39;session&#39;: &#39;f15cb79e-4700-4ef8-aed8-65224aba1be6&#39;, &#39;username&#39;: &#39;5bb12fba-c395-4a0a-afbc-31e5784dd153&#39;, &#39;version&#39;: &#39;5.2&#39;}, &#39;msg_id&#39;: &#39;4d438cb6-e1cd-4b07-9718-cb0949ab2096&#39;, &#39;msg_type&#39;: &#39;comm_msg&#39;, &#39;parent_header&#39;: {}, &#39;metadata&#39;: {}, &#39;content&#39;: {&#39;comm_id&#39;: &#39;086807ee363b4bbea301fac1c93c5444&#39;, &#39;data&#39;: {&#39;method&#39;: &#39;update&#39;, &#39;state&#39;: {&#39;outputs&#39;: [{&#39;output_type&#39;: &#39;display_data&#39;, &#39;data&#39;: {&#39;text/plain&#39;: &#39;  \x1b[35m100%\x1b[0m \x1b[36mSpecificity Filter: Blastn Crosshybridization\x1b[0m \x1b[38;2;114;156;31m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\x1b[0m \x1b[32m806/806\x1b[0m \x1b[33m0:02:35\x1b[0m &lt; \x1b[36m0:00:00\x1b[0m\n&#39;, &#39;text/html&#39;: &#39;&lt;pre style=&#34;white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,\&#39;DejaVu Sans Mono\&#39;,consolas,\&#39;Courier New\&#39;,monospace&#34;&gt;  &lt;span style=&#34;color: #800080; text-decoration-color: #800080&#34;&gt;100%&lt;/span&gt; &lt;span style=&#34;color: #008080; text-decoration-color: #008080&#34;&gt;Specificity Filter: Blastn Crosshybridization&lt;/span&gt; &lt;span style=&#34;color: #729c1f; text-decoration-color: #729c1f&#34;&gt;━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━&lt;/span&gt; &lt;span style=&#34;color: #008000; text-decoration-color: #008000&#34;&gt;806/806&lt;/span&gt; &lt;span style=&#34;color: #808000; text-decoration-color: #808000&#34;&gt;0:02:35&lt;/span&gt; &amp;lt; &lt;span style=&#34;color: #008080; text-decoration-color: #008080&#34;&gt;0:00:00&lt;/span&gt;\n&lt;/pre&gt;\n&#39;}, &#39;metadata&#39;: {}}]}, &#39;buffer_paths&#39;: []}}, &#39;buffers&#39;: []})
</pre></div></div>
</div>
</section>
<section id="Create-Probe-Sets-of-Non-Overlapping-Probes">
<h3>Create Probe Sets of Non-Overlapping Probes<a class="headerlink" href="#Create-Probe-Sets-of-Non-Overlapping-Probes" title="Permalink to this heading"></a></h3>
<p>After applying different sets of filters to the probe database, we will create probesets for each gene, which are sets of probes that do not overlap and have a high efficiency score (calculated from melting temperature, GC content and isoform consensus).</p>
<p><strong>Parameters for the Oligo Set Selection</strong> - <code class="docutils literal notranslate"><span class="pre">probe_isoform_weight</span></code>: weight of the isoform consensus of the probe in the efficiency score - <code class="docutils literal notranslate"><span class="pre">probe_GC_content_opt</span></code>: max and min values are defiend above - <code class="docutils literal notranslate"><span class="pre">probe_GC_weight</span></code>: weight of the GC content of the probe in the efficiency score - <code class="docutils literal notranslate"><span class="pre">probe_Tm_opt</span></code>: max and min values are defiend above - <code class="docutils literal notranslate"><span class="pre">probe_Tm_weight</span></code>: weight of the Tm of the probe in the efficiency score - <code class="docutils literal notranslate"><span class="pre">probeset_size_min</span></code>: minimum size of probe sets (in case there exist no
set of the optimal size) -&gt; genes with less oligos will be filtered out and stored in regions_with_insufficient_oligos_for_db_probes - <code class="docutils literal notranslate"><span class="pre">probeset_size_opt</span></code>: optimal size of probe sets - <code class="docutils literal notranslate"><span class="pre">distance_between_probes</span></code>: how much overlap should be allowed between oligos, e.g. if oligos can overlpap x bases choose -x, if oligos can be next to one another choose 0, if oligos should be x bases apart choose x - <code class="docutils literal notranslate"><span class="pre">n_sets</span></code>: maximum number of sets to generate - <code class="docutils literal notranslate"><span class="pre">max_graph_size</span></code>: maximum number of oligos
that are taken into consisderation in the last step (5000 -&gt; ~5GB, 2500 -&gt; ~1GB)</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[21]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1">## Define parameters</span>
<span class="n">probe_isoform_weight</span> <span class="o">=</span> <span class="mi">2</span>
<span class="n">probe_GC_weight</span> <span class="o">=</span> <span class="mi">1</span>
<span class="n">probe_GC_content_opt</span> <span class="o">=</span> <span class="mi">50</span>
<span class="n">probe_Tm_weight</span> <span class="o">=</span> <span class="mi">1</span>
<span class="n">probe_Tm_opt</span> <span class="o">=</span> <span class="mi">70</span>

<span class="n">probeset_size_min</span> <span class="o">=</span> <span class="mi">3</span>
<span class="n">probeset_size_opt</span> <span class="o">=</span> <span class="mi">5</span>
<span class="n">distance_between_probes</span> <span class="o">=</span> <span class="mi">0</span>
<span class="n">n_sets</span> <span class="o">=</span> <span class="mi">100</span>
<span class="n">max_graph_size</span> <span class="o">=</span> <span class="mi">5000</span>
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[23]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1">####### Load existing database #######</span>
<span class="c1"># dir_database = os.path.join(dir_output, &quot;db_probes/3_db_probes_specificity_filter&quot;)</span>
<span class="c1"># probe_database = OligoDatabase(min_oligos_per_region=min_probes_per_gene, write_regions_with_insufficient_oligos=True, lru_db_max_in_memory=n_jobs*2+2, database_name=&quot;db_probes&quot;, dir_output=dir_output, n_jobs=n_jobs)</span>
<span class="c1"># probe_database.load_database(dir_database=dir_database, region_ids=gene_ids, database_overwrite=True)</span>

<span class="c1">## Apply probe selection</span>
<span class="n">probe_database</span><span class="p">,</span> <span class="n">file_database</span><span class="p">,</span> <span class="n">dir_probesets</span> <span class="o">=</span> <span class="n">pipeline</span><span class="o">.</span><span class="n">create_probe_sets</span><span class="p">(</span>
    <span class="n">oligo_database</span><span class="o">=</span><span class="n">probe_database</span><span class="p">,</span>
    <span class="n">probe_isoform_weight</span><span class="o">=</span><span class="n">probe_isoform_weight</span><span class="p">,</span>
    <span class="n">probe_Tm_weight</span><span class="o">=</span><span class="n">probe_Tm_weight</span><span class="p">,</span>
    <span class="n">probe_Tm_min</span><span class="o">=</span><span class="n">probe_Tm_min</span><span class="p">,</span>
    <span class="n">probe_Tm_opt</span><span class="o">=</span><span class="n">probe_Tm_opt</span><span class="p">,</span>
    <span class="n">probe_Tm_max</span><span class="o">=</span><span class="n">probe_Tm_max</span><span class="p">,</span>
    <span class="n">Tm_parameters_probe</span><span class="o">=</span><span class="n">Tm_parameters_probe</span><span class="p">,</span>
    <span class="n">Tm_chem_correction_param_probe</span><span class="o">=</span><span class="n">Tm_chem_correction_param_probe</span><span class="p">,</span>
    <span class="n">Tm_salt_correction_param_probe</span><span class="o">=</span><span class="n">Tm_salt_correction_param_probe</span><span class="p">,</span>
    <span class="n">probe_GC_weight</span><span class="o">=</span><span class="n">probe_GC_weight</span><span class="p">,</span>
    <span class="n">probe_GC_content_min</span><span class="o">=</span><span class="n">probe_GC_content_min</span><span class="p">,</span>
    <span class="n">probe_GC_content_opt</span><span class="o">=</span><span class="n">probe_GC_content_opt</span><span class="p">,</span>
    <span class="n">probe_GC_content_max</span><span class="o">=</span><span class="n">probe_GC_content_max</span><span class="p">,</span>
    <span class="n">probeset_size_opt</span><span class="o">=</span><span class="n">probeset_size_opt</span><span class="p">,</span>
    <span class="n">probeset_size_min</span><span class="o">=</span><span class="n">probeset_size_min</span><span class="p">,</span>
    <span class="n">max_graph_size</span><span class="o">=</span><span class="n">max_graph_size</span><span class="p">,</span>
    <span class="n">n_sets</span><span class="o">=</span><span class="n">n_sets</span><span class="p">,</span>
    <span class="n">distance_between_probes</span><span class="o">=</span><span class="n">distance_between_probes</span><span class="p">,</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
2024-07-09 00:36:23,067 [INFO] Parameters Set Selection:
2024-07-09 00:36:23,070 [INFO] Function: create_probe_sets
2024-07-09 00:36:23,072 [INFO] Parameter: oligo_database = &lt;oligo_designer_toolsuite.database._oligo_database.OligoDatabase object at 0x10d917550&gt;
2024-07-09 00:36:23,075 [INFO] Parameter: probe_isoform_weight = 2
2024-07-09 00:36:23,077 [INFO] Parameter: probe_Tm_weight = 1
2024-07-09 00:36:23,078 [INFO] Parameter: probe_Tm_min = 65
2024-07-09 00:36:23,081 [INFO] Parameter: probe_Tm_opt = 70
2024-07-09 00:36:23,084 [INFO] Parameter: probe_Tm_max = 75
2024-07-09 00:36:23,086 [INFO] Parameter: Tm_parameters_probe = {&#39;check&#39;: True, &#39;strict&#39;: True, &#39;c_seq&#39;: None, &#39;shift&#39;: 0, &#39;nn_table&#39;: {&#39;init&#39;: (0, 0), &#39;init_A/T&#39;: (2.3, 4.1), &#39;init_G/C&#39;: (0.1, -2.8), &#39;init_oneG/C&#39;: (0, 0), &#39;init_allA/T&#39;: (0, 0), &#39;init_5T/A&#39;: (0, 0), &#39;sym&#39;: (0, -1.4), &#39;AA/TT&#39;: (-7.9, -22.2), &#39;AT/TA&#39;: (-7.2, -20.4), &#39;TA/AT&#39;: (-7.2, -21.3), &#39;CA/GT&#39;: (-8.5, -22.7), &#39;GT/CA&#39;: (-8.4, -22.4), &#39;CT/GA&#39;: (-7.8, -21.0), &#39;GA/CT&#39;: (-8.2, -22.2), &#39;CG/GC&#39;: (-10.6, -27.2), &#39;GC/CG&#39;: (-9.8, -24.4), &#39;GG/CC&#39;: (-8.0, -19.9)}, &#39;tmm_table&#39;: {&#39;AA/TA&#39;: (-3.1, -7.8), &#39;TA/AA&#39;: (-2.5, -6.3), &#39;CA/GA&#39;: (-4.3, -10.7), &#39;GA/CA&#39;: (-8.0, -22.5), &#39;AC/TC&#39;: (-0.1, 0.5), &#39;TC/AC&#39;: (-0.7, -1.3), &#39;CC/GC&#39;: (-2.1, -5.1), &#39;GC/CC&#39;: (-3.9, -10.6), &#39;AG/TG&#39;: (-1.1, -2.1), &#39;TG/AG&#39;: (-1.1, -2.7), &#39;CG/GG&#39;: (-3.8, -9.5), &#39;GG/CG&#39;: (-0.7, -19.2), &#39;AT/TT&#39;: (-2.4, -6.5), &#39;TT/AT&#39;: (-3.2, -8.9), &#39;CT/GT&#39;: (-6.1, -16.9), &#39;GT/CT&#39;: (-7.4, -21.2), &#39;AA/TC&#39;: (-1.6, -4.0), &#39;AC/TA&#39;: (-1.8, -3.8), &#39;CA/GC&#39;: (-2.6, -5.9), &#39;CC/GA&#39;: (-2.7, -6.0), &#39;GA/CC&#39;: (-5.0, -13.8), &#39;GC/CA&#39;: (-3.2, -7.1), &#39;TA/AC&#39;: (-2.3, -5.9), &#39;TC/AA&#39;: (-2.7, -7.0), &#39;AC/TT&#39;: (-0.9, -1.7), &#39;AT/TC&#39;: (-2.3, -6.3), &#39;CC/GT&#39;: (-3.2, -8.0), &#39;CT/GC&#39;: (-3.9, -10.6), &#39;GC/CT&#39;: (-4.9, -13.5), &#39;GT/CC&#39;: (-3.0, -7.8), &#39;TC/AT&#39;: (-2.5, -6.3), &#39;TT/AC&#39;: (-0.7, -1.2), &#39;AA/TG&#39;: (-1.9, -4.4), &#39;AG/TA&#39;: (-2.5, -5.9), &#39;CA/GG&#39;: (-3.9, -9.6), &#39;CG/GA&#39;: (-6.0, -15.5), &#39;GA/CG&#39;: (-4.3, -11.1), &#39;GG/CA&#39;: (-4.6, -11.4), &#39;TA/AG&#39;: (-2.0, -4.7), &#39;TG/AA&#39;: (-2.4, -5.8), &#39;AG/TT&#39;: (-3.2, -8.7), &#39;AT/TG&#39;: (-3.5, -9.4), &#39;CG/GT&#39;: (-3.8, -9.0), &#39;CT/GG&#39;: (-6.6, -18.7), &#39;GG/CT&#39;: (-5.7, -15.9), &#39;GT/CG&#39;: (-5.9, -16.1), &#39;TG/AT&#39;: (-3.9, -10.5), &#39;TT/AG&#39;: (-3.6, -9.8)}, &#39;imm_table&#39;: {&#39;AG/TT&#39;: (1.0, 0.9), &#39;AT/TG&#39;: (-2.5, -8.3), &#39;CG/GT&#39;: (-4.1, -11.7), &#39;CT/GG&#39;: (-2.8, -8.0), &#39;GG/CT&#39;: (3.3, 10.4), &#39;GG/TT&#39;: (5.8, 16.3), &#39;GT/CG&#39;: (-4.4, -12.3), &#39;GT/TG&#39;: (4.1, 9.5), &#39;TG/AT&#39;: (-0.1, -1.7), &#39;TG/GT&#39;: (-1.4, -6.2), &#39;TT/AG&#39;: (-1.3, -5.3), &#39;AA/TG&#39;: (-0.6, -2.3), &#39;AG/TA&#39;: (-0.7, -2.3), &#39;CA/GG&#39;: (-0.7, -2.3), &#39;CG/GA&#39;: (-4.0, -13.2), &#39;GA/CG&#39;: (-0.6, -1.0), &#39;GG/CA&#39;: (0.5, 3.2), &#39;TA/AG&#39;: (0.7, 0.7), &#39;TG/AA&#39;: (3.0, 7.4), &#39;AC/TT&#39;: (0.7, 0.2), &#39;AT/TC&#39;: (-1.2, -6.2), &#39;CC/GT&#39;: (-0.8, -4.5), &#39;CT/GC&#39;: (-1.5, -6.1), &#39;GC/CT&#39;: (2.3, 5.4), &#39;GT/CC&#39;: (5.2, 13.5), &#39;TC/AT&#39;: (1.2, 0.7), &#39;TT/AC&#39;: (1.0, 0.7), &#39;AA/TC&#39;: (2.3, 4.6), &#39;AC/TA&#39;: (5.3, 14.6), &#39;CA/GC&#39;: (1.9, 3.7), &#39;CC/GA&#39;: (0.6, -0.6), &#39;GA/CC&#39;: (5.2, 14.2), &#39;GC/CA&#39;: (-0.7, -3.8), &#39;TA/AC&#39;: (3.4, 8.0), &#39;TC/AA&#39;: (7.6, 20.2), &#39;AA/TA&#39;: (1.2, 1.7), &#39;CA/GA&#39;: (-0.9, -4.2), &#39;GA/CA&#39;: (-2.9, -9.8), &#39;TA/AA&#39;: (4.7, 12.9), &#39;AC/TC&#39;: (0.0, -4.4), &#39;CC/GC&#39;: (-1.5, -7.2), &#39;GC/CC&#39;: (3.6, 8.9), &#39;TC/AC&#39;: (6.1, 16.4), &#39;AG/TG&#39;: (-3.1, -9.5), &#39;CG/GG&#39;: (-4.9, -15.3), &#39;GG/CG&#39;: (-6.0, -15.8), &#39;TG/AG&#39;: (1.6, 3.6), &#39;AT/TT&#39;: (-2.7, -10.8), &#39;CT/GT&#39;: (-5.0, -15.8), &#39;GT/CT&#39;: (-2.2, -8.4), &#39;TT/AT&#39;: (0.2, -1.5), &#39;AI/TC&#39;: (-8.9, -25.5), &#39;TI/AC&#39;: (-5.9, -17.4), &#39;AC/TI&#39;: (-8.8, -25.4), &#39;TC/AI&#39;: (-4.9, -13.9), &#39;CI/GC&#39;: (-5.4, -13.7), &#39;GI/CC&#39;: (-6.8, -19.1), &#39;CC/GI&#39;: (-8.3, -23.8), &#39;GC/CI&#39;: (-5.0, -12.6), &#39;AI/TA&#39;: (-8.3, -25.0), &#39;TI/AA&#39;: (-3.4, -11.2), &#39;AA/TI&#39;: (-0.7, -2.6), &#39;TA/AI&#39;: (-1.3, -4.6), &#39;CI/GA&#39;: (2.6, 8.9), &#39;GI/CA&#39;: (-7.8, -21.1), &#39;CA/GI&#39;: (-7.0, -20.0), &#39;GA/CI&#39;: (-7.6, -20.2), &#39;AI/TT&#39;: (0.49, -0.7), &#39;TI/AT&#39;: (-6.5, -22.0), &#39;AT/TI&#39;: (-5.6, -18.7), &#39;TT/AI&#39;: (-0.8, -4.3), &#39;CI/GT&#39;: (-1.0, -2.4), &#39;GI/CT&#39;: (-3.5, -10.6), &#39;CT/GI&#39;: (0.1, -1.0), &#39;GT/CI&#39;: (-4.3, -12.1), &#39;AI/TG&#39;: (-4.9, -15.8), &#39;TI/AG&#39;: (-1.9, -8.5), &#39;AG/TI&#39;: (0.1, -1.8), &#39;TG/AI&#39;: (1.0, 1.0), &#39;CI/GG&#39;: (7.1, 21.3), &#39;GI/CG&#39;: (-1.1, -3.2), &#39;CG/GI&#39;: (5.8, 16.9), &#39;GG/CI&#39;: (-7.6, -22.0), &#39;AI/TI&#39;: (-3.3, -11.9), &#39;TI/AI&#39;: (0.1, -2.3), &#39;CI/GI&#39;: (1.3, 3.0), &#39;GI/CI&#39;: (-0.5, -1.3)}, &#39;de_table&#39;: {&#39;AA/.T&#39;: (0.2, 2.3), &#39;AC/.G&#39;: (-6.3, -17.1), &#39;AG/.C&#39;: (-3.7, -10.0), &#39;AT/.A&#39;: (-2.9, -7.6), &#39;CA/.T&#39;: (0.6, 3.3), &#39;CC/.G&#39;: (-4.4, -12.6), &#39;CG/.C&#39;: (-4.0, -11.9), &#39;CT/.A&#39;: (-4.1, -13.0), &#39;GA/.T&#39;: (-1.1, -1.6), &#39;GC/.G&#39;: (-5.1, -14.0), &#39;GG/.C&#39;: (-3.9, -10.9), &#39;GT/.A&#39;: (-4.2, -15.0), &#39;TA/.T&#39;: (-6.9, -20.0), &#39;TC/.G&#39;: (-4.0, -10.9), &#39;TG/.C&#39;: (-4.9, -13.8), &#39;TT/.A&#39;: (-0.2, -0.5), &#39;.A/AT&#39;: (-0.7, -0.8), &#39;.C/AG&#39;: (-2.1, -3.9), &#39;.G/AC&#39;: (-5.9, -16.5), &#39;.T/AA&#39;: (-0.5, -1.1), &#39;.A/CT&#39;: (4.4, 14.9), &#39;.C/CG&#39;: (-0.2, -0.1), &#39;.G/CC&#39;: (-2.6, -7.4), &#39;.T/CA&#39;: (4.7, 14.2), &#39;.A/GT&#39;: (-1.6, -3.6), &#39;.C/GG&#39;: (-3.9, -11.2), &#39;.G/GC&#39;: (-3.2, -10.4), &#39;.T/GA&#39;: (-4.1, -13.1), &#39;.A/TT&#39;: (2.9, 10.4), &#39;.C/TG&#39;: (-4.4, -13.1), &#39;.G/TC&#39;: (-5.2, -15.0), &#39;.T/TA&#39;: (-3.8, -12.6)}, &#39;dnac1&#39;: 50, &#39;dnac2&#39;: 0, &#39;selfcomp&#39;: False, &#39;saltcorr&#39;: 7, &#39;Na&#39;: 39, &#39;K&#39;: 75, &#39;Tris&#39;: 20, &#39;Mg&#39;: 10, &#39;dNTPs&#39;: 0}
2024-07-09 00:36:23,089 [INFO] Parameter: Tm_chem_correction_param_probe = {&#39;DMSO&#39;: 0, &#39;fmd&#39;: 20, &#39;DMSOfactor&#39;: 0.75, &#39;fmdfactor&#39;: 0.65, &#39;fmdmethod&#39;: 1, &#39;GC&#39;: None}
2024-07-09 00:36:23,090 [INFO] Parameter: Tm_salt_correction_param_probe = None
2024-07-09 00:36:23,091 [INFO] Parameter: probe_GC_weight = 1
2024-07-09 00:36:23,094 [INFO] Parameter: probe_GC_content_min = 40
2024-07-09 00:36:23,096 [INFO] Parameter: probe_GC_content_opt = 50
2024-07-09 00:36:23,097 [INFO] Parameter: probe_GC_content_max = 60
2024-07-09 00:36:23,098 [INFO] Parameter: probeset_size_opt = 5
2024-07-09 00:36:23,099 [INFO] Parameter: probeset_size_min = 3
2024-07-09 00:36:23,100 [INFO] Parameter: max_graph_size = 5000
2024-07-09 00:36:23,101 [INFO] Parameter: n_sets = 100
2024-07-09 00:36:23,103 [INFO] Parameter: distance_between_probes = 0
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<script type="application/vnd.jupyter.widget-view+json">{"model_id": "b011dd7910384d038ec20102cf492c08", "version_major": 2, "version_minor": 0}</script></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area rendered_html docutils container">
<pre style="white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace"></pre></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area rendered_html docutils container">
<pre style="white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace">
</pre></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
2024-07-09 03:04:22,408 [INFO] Step - Set Selection: database contains 16129 oligos from 735 regions.
2024-07-09 03:04:22,449 [DEBUG] handle_msg[b011dd7910384d038ec20102cf492c08]({&#39;header&#39;: {&#39;date&#39;: datetime.datetime(2024, 7, 9, 1, 4, 22, 445000, tzinfo=tzutc()), &#39;msg_id&#39;: &#39;75f170fe-5fbb-46a3-9620-3f145fbff019&#39;, &#39;msg_type&#39;: &#39;comm_msg&#39;, &#39;session&#39;: &#39;f15cb79e-4700-4ef8-aed8-65224aba1be6&#39;, &#39;username&#39;: &#39;5bb12fba-c395-4a0a-afbc-31e5784dd153&#39;, &#39;version&#39;: &#39;5.2&#39;}, &#39;msg_id&#39;: &#39;75f170fe-5fbb-46a3-9620-3f145fbff019&#39;, &#39;msg_type&#39;: &#39;comm_msg&#39;, &#39;parent_header&#39;: {}, &#39;metadata&#39;: {}, &#39;content&#39;: {&#39;comm_id&#39;: &#39;b011dd7910384d038ec20102cf492c08&#39;, &#39;data&#39;: {&#39;method&#39;: &#39;update&#39;, &#39;state&#39;: {&#39;outputs&#39;: [{&#39;output_type&#39;: &#39;display_data&#39;, &#39;data&#39;: {&#39;text/plain&#39;: &#39;  \x1b[35m100%\x1b[0m \x1b[36mFind Oligosets\x1b[0m \x1b[38;2;114;156;31m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\x1b[0m \x1b[32m806/806\x1b[0m \x1b[33m2:27:50\x1b[0m &lt; \x1b[36m0:00:00\x1b[0m\n&#39;, &#39;text/html&#39;: &#39;&lt;pre style=&#34;white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,\&#39;DejaVu Sans Mono\&#39;,consolas,\&#39;Courier New\&#39;,monospace&#34;&gt;  &lt;span style=&#34;color: #800080; text-decoration-color: #800080&#34;&gt;100%&lt;/span&gt; &lt;span style=&#34;color: #008080; text-decoration-color: #008080&#34;&gt;Find Oligosets&lt;/span&gt; &lt;span style=&#34;color: #729c1f; text-decoration-color: #729c1f&#34;&gt;━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━&lt;/span&gt; &lt;span style=&#34;color: #008000; text-decoration-color: #008000&#34;&gt;806/806&lt;/span&gt; &lt;span style=&#34;color: #808000; text-decoration-color: #808000&#34;&gt;2:27:50&lt;/span&gt; &amp;lt; &lt;span style=&#34;color: #008080; text-decoration-color: #008080&#34;&gt;0:00:00&lt;/span&gt;\n&lt;/pre&gt;\n&#39;}, &#39;metadata&#39;: {}}]}, &#39;buffer_paths&#39;: []}}, &#39;buffers&#39;: []})
</pre></div></div>
</div>
</section>
</section>
<section id="Select-Gene-Panel">
<h2>Select Gene Panel<a class="headerlink" href="#Select-Gene-Panel" title="Permalink to this heading"></a></h2>
<p>In the probe database, the gene names are the keys of the database. All genes that do not have sufficient probes were removed from the database. Once we hve all genes with sufficient probes, we can run the gene set selection step. Therefore, we include additional metadata information to our adata object, i.e. the genes that have sufficient probes and the genes that fulfill both constraint (<em>highly variable</em> and <em>sufficient probes</em>).</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[24]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1">####### Load existing database #######</span>
<span class="c1"># dir_database = os.path.join(dir_output, &quot;db_probes/4_db_probes_probesets&quot;)</span>
<span class="c1"># probe_database = OligoDatabase(min_oligos_per_region=min_probes_per_gene, write_regions_with_insufficient_oligos=True, lru_db_max_in_memory=n_jobs*2+2, database_name=&quot;db_probes&quot;, dir_output=dir_output, n_jobs=n_jobs)</span>
<span class="c1"># probe_database.load_database(dir_database=dir_database, region_ids=gene_ids, database_overwrite=True)</span>

<span class="c1"># get gene names of genes with sufficient number of probes to proceed with next step</span>
<span class="n">genes_with_sufficient_probes</span> <span class="o">=</span> <span class="n">probe_database</span><span class="o">.</span><span class="n">database</span><span class="o">.</span><span class="n">keys</span><span class="p">()</span>

<span class="c1"># add extra clolumn to anndata to mark genes with sufficient probes</span>
<span class="n">pbmc_data</span><span class="o">.</span><span class="n">var</span><span class="p">[</span><span class="s2">&quot;sufficient_probes&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="kc">False</span>
<span class="k">for</span> <span class="n">gene</span> <span class="ow">in</span> <span class="n">highly_variable_genes</span><span class="p">:</span>
    <span class="k">if</span> <span class="n">gene</span> <span class="ow">in</span> <span class="n">genes_with_sufficient_probes</span><span class="p">:</span>
        <span class="n">pbmc_data</span><span class="o">.</span><span class="n">var</span><span class="p">[</span><span class="s2">&quot;sufficient_probes&quot;</span><span class="p">][</span><span class="n">gene</span><span class="p">]</span> <span class="o">=</span> <span class="kc">True</span>

<span class="c1"># create a new variable that indicates if the gene passes the first constraint filter</span>
<span class="n">pbmc_data</span><span class="o">.</span><span class="n">var</span><span class="p">[</span><span class="s2">&quot;pass_constraints&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="n">su_p</span> <span class="ow">and</span> <span class="n">hi_v</span> <span class="k">for</span> <span class="n">su_p</span><span class="p">,</span> <span class="n">hi_v</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span><span class="n">pbmc_data</span><span class="o">.</span><span class="n">var</span><span class="p">[</span><span class="s2">&quot;sufficient_probes&quot;</span><span class="p">],</span> <span class="n">pbmc_data</span><span class="o">.</span><span class="n">var</span><span class="p">[</span><span class="s2">&quot;highly_variable&quot;</span><span class="p">])]</span>
</pre></div>
</div>
</div>
<div class="line-block">
<div class="line">To run the geneset selection step, we create an instance of the <code class="docutils literal notranslate"><span class="pre">ProbesetSelector</span></code> class. We specify the number of genes <code class="docutils literal notranslate"><span class="pre">n</span></code> (20) and the keys in <code class="docutils literal notranslate"><span class="pre">adata.obs</span></code> and <code class="docutils literal notranslate"><span class="pre">adata.var</span></code> where we find the cell type annotations (<code class="docutils literal notranslate"><span class="pre">celltype_key=&quot;celltype&quot;</span></code>) and selected genes (<code class="docutils literal notranslate"><span class="pre">celltype_key=&quot;celltype&quot;</span></code>), respectively.</div>
<div class="line"><em>Note: that you can specify a ``save_dir`` to save results during the selection step and reload them next time a ``ProbesetSelector`` with the given ``save_dir`` is instantiated.</em></div>
</div>
<p>Executing the cell below will give us a warning that the cell type clusters for dendritic cells and megakaroycytes are quite small and therefore the genes that are selected to identify these cell types potentially don’t generalize very well. The method will not exclude these cell types automatically, but it can be done manually by setting the parameter <code class="docutils literal notranslate"><span class="pre">celltypes</span></code> to a subset of cell types instead of <code class="docutils literal notranslate"><span class="pre">celltypes=&quot;all&quot;</span></code>.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[25]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1">## Select genes for gene panel</span>
<span class="n">selector</span> <span class="o">=</span> <span class="n">sp</span><span class="o">.</span><span class="n">se</span><span class="o">.</span><span class="n">ProbesetSelector</span><span class="p">(</span><span class="n">pbmc_data</span><span class="p">,</span> <span class="n">n</span><span class="o">=</span><span class="mi">20</span><span class="p">,</span> <span class="n">genes_key</span><span class="o">=</span><span class="s2">&quot;pass_constraints&quot;</span><span class="p">,</span> <span class="n">celltype_key</span><span class="o">=</span><span class="s2">&quot;celltype&quot;</span><span class="p">,</span> <span class="n">verbosity</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">save_dir</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span>
<span class="n">selector</span><span class="o">.</span><span class="n">select_probeset</span><span class="p">()</span>
<span class="n">selected_genes</span> <span class="o">=</span> <span class="n">selector</span><span class="o">.</span><span class="n">probeset</span><span class="o">.</span><span class="n">index</span><span class="p">[</span><span class="n">selector</span><span class="o">.</span><span class="n">probeset</span><span class="o">.</span><span class="n">selection</span><span class="p">]</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<script type="application/vnd.jupyter.widget-view+json">{"model_id": "74aab958e5b54cb698e62612fa9e03f8", "version_major": 2, "version_minor": 0}</script></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Note: The following celltypes&#39; test set sizes for forest training are below min_test_n (=20):
         Dendritic cells : 9
         Megakaryocytes  : 3
The genes selected for those cell types potentially don&#39;t generalize well. Find the genes for each of those cell types in self.genes_of_primary_trees after running self.select_probeset().
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
/Users/lisa.barros/anaconda3/envs/odt_test/lib/python3.10/site-packages/threadpoolctl.py:1214: RuntimeWarning:
Found Intel OpenMP (&#39;libiomp&#39;) and LLVM OpenMP (&#39;libomp&#39;) loaded at
the same time. Both libraries are known to be incompatible and this
can cause random crashes or deadlocks on Linux when loaded in the
same Python program.
Using threadpoolctl may cause crashes or deadlocks. For more
information and possible workarounds, please see
    https://github.com/joblib/threadpoolctl/blob/master/multiple_openmp.md

  warnings.warn(msg, RuntimeWarning)
/Users/lisa.barros/anaconda3/envs/odt_test/lib/python3.10/site-packages/threadpoolctl.py:1214: RuntimeWarning:
Found Intel OpenMP (&#39;libiomp&#39;) and LLVM OpenMP (&#39;libomp&#39;) loaded at
the same time. Both libraries are known to be incompatible and this
can cause random crashes or deadlocks on Linux when loaded in the
same Python program.
Using threadpoolctl may cause crashes or deadlocks. For more
information and possible workarounds, please see
    https://github.com/joblib/threadpoolctl/blob/master/multiple_openmp.md

  warnings.warn(msg, RuntimeWarning)
/Users/lisa.barros/anaconda3/envs/odt_test/lib/python3.10/site-packages/threadpoolctl.py:1214: RuntimeWarning:
Found Intel OpenMP (&#39;libiomp&#39;) and LLVM OpenMP (&#39;libomp&#39;) loaded at
the same time. Both libraries are known to be incompatible and this
can cause random crashes or deadlocks on Linux when loaded in the
same Python program.
Using threadpoolctl may cause crashes or deadlocks. For more
information and possible workarounds, please see
    https://github.com/joblib/threadpoolctl/blob/master/multiple_openmp.md

  warnings.warn(msg, RuntimeWarning)
/Users/lisa.barros/anaconda3/envs/odt_test/lib/python3.10/site-packages/threadpoolctl.py:1214: RuntimeWarning:
Found Intel OpenMP (&#39;libiomp&#39;) and LLVM OpenMP (&#39;libomp&#39;) loaded at
the same time. Both libraries are known to be incompatible and this
can cause random crashes or deadlocks on Linux when loaded in the
same Python program.
Using threadpoolctl may cause crashes or deadlocks. For more
information and possible workarounds, please see
    https://github.com/joblib/threadpoolctl/blob/master/multiple_openmp.md

  warnings.warn(msg, RuntimeWarning)
/Users/lisa.barros/anaconda3/envs/odt_test/lib/python3.10/site-packages/threadpoolctl.py:1214: RuntimeWarning:
Found Intel OpenMP (&#39;libiomp&#39;) and LLVM OpenMP (&#39;libomp&#39;) loaded at
the same time. Both libraries are known to be incompatible and this
can cause random crashes or deadlocks on Linux when loaded in the
same Python program.
Using threadpoolctl may cause crashes or deadlocks. For more
information and possible workarounds, please see
    https://github.com/joblib/threadpoolctl/blob/master/multiple_openmp.md

  warnings.warn(msg, RuntimeWarning)
/Users/lisa.barros/anaconda3/envs/odt_test/lib/python3.10/site-packages/threadpoolctl.py:1214: RuntimeWarning:
Found Intel OpenMP (&#39;libiomp&#39;) and LLVM OpenMP (&#39;libomp&#39;) loaded at
the same time. Both libraries are known to be incompatible and this
can cause random crashes or deadlocks on Linux when loaded in the
same Python program.
Using threadpoolctl may cause crashes or deadlocks. For more
information and possible workarounds, please see
    https://github.com/joblib/threadpoolctl/blob/master/multiple_openmp.md

  warnings.warn(msg, RuntimeWarning)
/Users/lisa.barros/anaconda3/envs/odt_test/lib/python3.10/site-packages/threadpoolctl.py:1214: RuntimeWarning:
Found Intel OpenMP (&#39;libiomp&#39;) and LLVM OpenMP (&#39;libomp&#39;) loaded at
the same time. Both libraries are known to be incompatible and this
can cause random crashes or deadlocks on Linux when loaded in the
same Python program.
Using threadpoolctl may cause crashes or deadlocks. For more
information and possible workarounds, please see
    https://github.com/joblib/threadpoolctl/blob/master/multiple_openmp.md

  warnings.warn(msg, RuntimeWarning)
/Users/lisa.barros/anaconda3/envs/odt_test/lib/python3.10/site-packages/threadpoolctl.py:1214: RuntimeWarning:
Found Intel OpenMP (&#39;libiomp&#39;) and LLVM OpenMP (&#39;libomp&#39;) loaded at
the same time. Both libraries are known to be incompatible and this
can cause random crashes or deadlocks on Linux when loaded in the
same Python program.
Using threadpoolctl may cause crashes or deadlocks. For more
information and possible workarounds, please see
    https://github.com/joblib/threadpoolctl/blob/master/multiple_openmp.md

  warnings.warn(msg, RuntimeWarning)
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area rendered_html docutils container">
<pre style="white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace"></pre></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area rendered_html docutils container">
<pre style="white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace">
</pre></div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[26]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1">## Remove all genes from the database that are not selected for the gene panel</span>
<span class="n">probe_database</span><span class="o">.</span><span class="n">database</span> <span class="o">=</span> <span class="p">{</span><span class="n">key</span><span class="p">:</span> <span class="n">value</span> <span class="k">for</span> <span class="n">key</span><span class="p">,</span> <span class="n">value</span> <span class="ow">in</span> <span class="n">probe_database</span><span class="o">.</span><span class="n">database</span><span class="o">.</span><span class="n">items</span><span class="p">()</span> <span class="k">if</span> <span class="n">key</span> <span class="ow">in</span> <span class="n">selected_genes</span><span class="p">}</span>
<span class="n">probe_database</span><span class="o">.</span><span class="n">oligosets</span> <span class="o">=</span> <span class="p">{</span><span class="n">key</span><span class="p">:</span> <span class="n">value</span> <span class="k">for</span> <span class="n">key</span><span class="p">,</span> <span class="n">value</span> <span class="ow">in</span> <span class="n">probe_database</span><span class="o">.</span><span class="n">oligosets</span><span class="o">.</span><span class="n">items</span><span class="p">()</span> <span class="k">if</span> <span class="n">key</span> <span class="ow">in</span> <span class="n">selected_genes</span><span class="p">}</span>
</pre></div>
</div>
</div>
</section>
<section id="Get-Final-Padlock-Probe-Sequences">
<h2>Get Final Padlock Probe Sequences<a class="headerlink" href="#Get-Final-Padlock-Probe-Sequences" title="Permalink to this heading"></a></h2>
<p>Once we have all selected genes, we create the final “ready to order” probe sequences. Calling the fuction below will produce two files, *[padlock, merfish, seqfish]_probes.yaml* and *[padlock, merfish, seqfish]_probes_order.yaml*. The latter file contains the ready to order probe sequences for each gene.</p>
<p><strong>Parameters for Padlock Sequence Design</strong> - <code class="docutils literal notranslate"><span class="pre">U_distance</span></code>: preferred minimal distance between U(racils) - <code class="docutils literal notranslate"><span class="pre">detect_oligo_Tm_opt</span></code>: optimal melting temperature of detection probe - <code class="docutils literal notranslate"><span class="pre">top_n_sets</span></code>: maximum number of sets to report in *[padlock, merfish, seqfish]_probes.yaml* and *[padlock, merfish, seqfish]_probes_order.yaml*</p>
<p><em>Note: The melting temperature is used in 2 different stages (probe and detection oligo design), where a few parameters are shared and the others differ. Parameters for melting temperature -&gt; for more information on parameters, see:</em><a class="reference external" href="https://biopython.org/docs/1.75/api/Bio.SeqUtils.MeltingTemp.html#Bio.SeqUtils.MeltingTemp.Tm_NN">here</a></p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[30]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1">## Define parameters</span>
<span class="n">U_distance</span> <span class="o">=</span> <span class="mi">5</span>
<span class="n">detect_oligo_Tm_opt</span> <span class="o">=</span> <span class="mi">56</span>
<span class="n">top_n_sets</span> <span class="o">=</span> <span class="mi">3</span>

<span class="n">Tm_parameters_detection_oligo</span> <span class="o">=</span> <span class="p">{</span>
    <span class="s2">&quot;check&quot;</span><span class="p">:</span> <span class="kc">True</span><span class="p">,</span>  <span class="c1"># default</span>
    <span class="s2">&quot;strict&quot;</span><span class="p">:</span> <span class="kc">True</span><span class="p">,</span>  <span class="c1"># default</span>
    <span class="s2">&quot;c_seq&quot;</span><span class="p">:</span> <span class="kc">None</span><span class="p">,</span>  <span class="c1"># default</span>
    <span class="s2">&quot;shift&quot;</span><span class="p">:</span> <span class="mi">0</span><span class="p">,</span>  <span class="c1"># default</span>
    <span class="s2">&quot;nn_table&quot;</span><span class="p">:</span> <span class="nb">getattr</span><span class="p">(</span><span class="n">mt</span><span class="p">,</span> <span class="s2">&quot;DNA_NN3&quot;</span><span class="p">),</span> <span class="c1"># Allawi &amp; SantaLucia (1997)</span>
    <span class="s2">&quot;tmm_table&quot;</span><span class="p">:</span> <span class="nb">getattr</span><span class="p">(</span><span class="n">mt</span><span class="p">,</span> <span class="s2">&quot;DNA_TMM1&quot;</span><span class="p">),</span> <span class="c1">#default</span>
    <span class="s2">&quot;imm_table&quot;</span><span class="p">:</span> <span class="nb">getattr</span><span class="p">(</span><span class="n">mt</span><span class="p">,</span> <span class="s2">&quot;DNA_IMM1&quot;</span><span class="p">),</span> <span class="c1">#default</span>
    <span class="s2">&quot;de_table&quot;</span><span class="p">:</span> <span class="nb">getattr</span><span class="p">(</span><span class="n">mt</span><span class="p">,</span> <span class="s2">&quot;DNA_DE1&quot;</span><span class="p">),</span> <span class="c1">#default</span>
    <span class="s2">&quot;dnac1&quot;</span><span class="p">:</span> <span class="mi">50</span><span class="p">,</span>  <span class="c1"># [nM]</span>
    <span class="s2">&quot;dnac2&quot;</span><span class="p">:</span> <span class="mi">0</span><span class="p">,</span>  <span class="c1"># [nM]</span>
    <span class="s2">&quot;selfcomp&quot;</span><span class="p">:</span> <span class="kc">False</span><span class="p">,</span>  <span class="c1"># default</span>
    <span class="s2">&quot;saltcorr&quot;</span><span class="p">:</span> <span class="mi">7</span><span class="p">,</span>  <span class="c1"># Owczarzy et al. (2008)</span>
    <span class="s2">&quot;Na&quot;</span><span class="p">:</span> <span class="mi">39</span><span class="p">,</span>  <span class="c1"># [mM]</span>
    <span class="s2">&quot;K&quot;</span><span class="p">:</span> <span class="mi">0</span><span class="p">,</span>  <span class="c1"># [mM] default</span>
    <span class="s2">&quot;Tris&quot;</span><span class="p">:</span> <span class="mi">0</span><span class="p">,</span>  <span class="c1"># [mM] default</span>
    <span class="s2">&quot;Mg&quot;</span><span class="p">:</span> <span class="mi">0</span><span class="p">,</span>  <span class="c1"># [mM] default</span>
    <span class="s2">&quot;dNTPs&quot;</span><span class="p">:</span> <span class="mi">0</span><span class="p">,</span>  <span class="c1"># [mM] default</span>
<span class="p">}</span>

<span class="n">Tm_chem_correction_param_detection_oligo</span> <span class="o">=</span> <span class="p">{</span>
    <span class="s2">&quot;DMSO&quot;</span><span class="p">:</span> <span class="mi">0</span><span class="p">,</span>  <span class="c1"># default</span>
    <span class="s2">&quot;fmd&quot;</span><span class="p">:</span> <span class="mi">30</span><span class="p">,</span>
    <span class="s2">&quot;DMSOfactor&quot;</span><span class="p">:</span> <span class="mf">0.75</span><span class="p">,</span>  <span class="c1"># default</span>
    <span class="s2">&quot;fmdfactor&quot;</span><span class="p">:</span> <span class="mf">0.65</span><span class="p">,</span>  <span class="c1"># default</span>
    <span class="s2">&quot;fmdmethod&quot;</span><span class="p">:</span> <span class="mi">1</span><span class="p">,</span>  <span class="c1"># default</span>
    <span class="s2">&quot;GC&quot;</span><span class="p">:</span> <span class="kc">None</span><span class="p">,</span>  <span class="c1"># default</span>
<span class="p">}</span>

<span class="n">Tm_salt_correction_param_detection_oligo</span> <span class="o">=</span> <span class="kc">None</span>
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[32]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1">## Design final sequences</span>
<span class="n">probe_database</span> <span class="o">=</span> <span class="n">pipeline</span><span class="o">.</span><span class="n">design_final_probe_sequence</span><span class="p">(</span>
    <span class="n">oligo_database</span><span class="o">=</span><span class="n">probe_database</span><span class="p">,</span>
    <span class="n">min_thymines</span><span class="o">=</span><span class="n">min_thymines</span><span class="p">,</span>
    <span class="n">U_distance</span><span class="o">=</span><span class="n">U_distance</span><span class="p">,</span>
    <span class="n">detect_oligo_length_min</span><span class="o">=</span><span class="n">detect_oligo_length_min</span><span class="p">,</span>
    <span class="n">detect_oligo_length_max</span><span class="o">=</span><span class="n">detect_oligo_length_max</span><span class="p">,</span>
    <span class="n">detect_oligo_Tm_opt</span><span class="o">=</span><span class="n">detect_oligo_Tm_opt</span><span class="p">,</span>
    <span class="n">Tm_parameters_detection_oligo</span><span class="o">=</span><span class="n">Tm_parameters_detection_oligo</span><span class="p">,</span>
    <span class="n">Tm_chem_correction_param_detection_oligo</span><span class="o">=</span><span class="n">Tm_chem_correction_param_detection_oligo</span><span class="p">,</span>
    <span class="n">Tm_salt_correction_param_detection_oligo</span><span class="o">=</span><span class="n">Tm_salt_correction_param_detection_oligo</span><span class="p">,</span>
    <span class="n">Tm_parameters_probe</span><span class="o">=</span><span class="n">Tm_parameters_probe</span><span class="p">,</span>
    <span class="n">Tm_chem_correction_param_probe</span><span class="o">=</span><span class="n">Tm_chem_correction_param_probe</span><span class="p">,</span>
    <span class="n">Tm_salt_correction_param_probe</span><span class="o">=</span><span class="n">Tm_salt_correction_param_probe</span><span class="p">,</span>
<span class="p">)</span>

<span class="c1">## Compute all required probe attributes for output</span>
<span class="n">probe_database</span> <span class="o">=</span> <span class="n">pipeline</span><span class="o">.</span><span class="n">compute_probe_attributes</span><span class="p">(</span>
    <span class="n">oligo_database</span><span class="o">=</span><span class="n">probe_database</span><span class="p">,</span>
    <span class="n">Tm_parameters_probe</span><span class="o">=</span><span class="n">Tm_parameters_probe</span><span class="p">,</span>
    <span class="n">Tm_chem_correction_param_probe</span><span class="o">=</span><span class="n">Tm_chem_correction_param_probe</span><span class="p">,</span>
    <span class="n">Tm_salt_correction_param_probe</span><span class="o">=</span><span class="n">Tm_salt_correction_param_probe</span><span class="p">,</span>
<span class="p">)</span>

<span class="c1">## Write output to files</span>
<span class="n">pipeline</span><span class="o">.</span><span class="n">generate_output</span><span class="p">(</span><span class="n">oligo_database</span><span class="o">=</span><span class="n">probe_database</span><span class="p">,</span> <span class="n">top_n_sets</span><span class="o">=</span><span class="n">top_n_sets</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<script type="application/vnd.jupyter.widget-view+json">{"model_id": "1a931c4c89ec48e6be4917b9a4ac2f8f", "version_major": 2, "version_minor": 0}</script></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area rendered_html docutils container">
<pre style="white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace"></pre></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area rendered_html docutils container">
<pre style="white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace">
</pre></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
2024-07-09 10:12:10,657 [DEBUG] handle_msg[1a931c4c89ec48e6be4917b9a4ac2f8f]({&#39;header&#39;: {&#39;date&#39;: datetime.datetime(2024, 7, 9, 8, 12, 10, 653000, tzinfo=tzutc()), &#39;msg_id&#39;: &#39;394388fb-c687-478d-b79b-49e5fecd638c&#39;, &#39;msg_type&#39;: &#39;comm_msg&#39;, &#39;session&#39;: &#39;f15cb79e-4700-4ef8-aed8-65224aba1be6&#39;, &#39;username&#39;: &#39;5bb12fba-c395-4a0a-afbc-31e5784dd153&#39;, &#39;version&#39;: &#39;5.2&#39;}, &#39;msg_id&#39;: &#39;394388fb-c687-478d-b79b-49e5fecd638c&#39;, &#39;msg_type&#39;: &#39;comm_msg&#39;, &#39;parent_header&#39;: {}, &#39;metadata&#39;: {}, &#39;content&#39;: {&#39;comm_id&#39;: &#39;1a931c4c89ec48e6be4917b9a4ac2f8f&#39;, &#39;data&#39;: {&#39;method&#39;: &#39;update&#39;, &#39;state&#39;: {&#39;outputs&#39;: [{&#39;output_type&#39;: &#39;display_data&#39;, &#39;data&#39;: {&#39;text/plain&#39;: &#39;  \x1b[35m100%\x1b[0m \x1b[36mDesign Final Padlock Sequence\x1b[0m \x1b[38;2;114;156;31m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\x1b[0m \x1b[32m20/20\x1b[0m \x1b[33m0:00:22\x1b[0m &lt; \x1b[36m0:00:00\x1b[0m\n&#39;, &#39;text/html&#39;: &#39;&lt;pre style=&#34;white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,\&#39;DejaVu Sans Mono\&#39;,consolas,\&#39;Courier New\&#39;,monospace&#34;&gt;  &lt;span style=&#34;color: #800080; text-decoration-color: #800080&#34;&gt;100%&lt;/span&gt; &lt;span style=&#34;color: #008080; text-decoration-color: #008080&#34;&gt;Design Final Padlock Sequence&lt;/span&gt; &lt;span style=&#34;color: #729c1f; text-decoration-color: #729c1f&#34;&gt;━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━&lt;/span&gt; &lt;span style=&#34;color: #008000; text-decoration-color: #008000&#34;&gt;20/20&lt;/span&gt; &lt;span style=&#34;color: #808000; text-decoration-color: #808000&#34;&gt;0:00:22&lt;/span&gt; &amp;lt; &lt;span style=&#34;color: #008080; text-decoration-color: #008080&#34;&gt;0:00:00&lt;/span&gt;\n&lt;/pre&gt;\n&#39;}, &#39;metadata&#39;: {}}]}, &#39;buffer_paths&#39;: []}}, &#39;buffers&#39;: []})
</pre></div></div>
</div>
</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="spapros_tutorial_end_to_end_selection_short.html" class="btn btn-neutral float-left" title="End to end selection (short version)" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="../authors.html" class="btn btn-neutral float-right" title="Credits" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2022, Lukas Heumos.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>