<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" />
  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>End to end selection &mdash; spapros docs</title>
      <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="../_static/css/theme.css" type="text/css" />
      <link rel="stylesheet" href="../_static/sg_gallery.css" type="text/css" />
      <link rel="stylesheet" href="../_static/nbsphinx-code-cells.css" type="text/css" />
      <link rel="stylesheet" href="../_static/custom_cookietemple.css" type="text/css" />
      <link rel="stylesheet" href="../_static/dark_mode_css/general.css" type="text/css" />
      <link rel="stylesheet" href="../_static/dark_mode_css/dark.css" type="text/css" />
    <link rel="shortcut icon" href="../_static/logo.png"/>
  <!--[if lt IE 9]>
    <script src="../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script data-url_root="../" id="documentation_options" src="../_static/documentation_options.js"></script>
        <script src="../_static/jquery.js"></script>
        <script src="../_static/underscore.js"></script>
        <script src="../_static/_sphinx_javascript_frameworks_compat.js"></script>
        <script src="../_static/doctools.js"></script>
        <script src="../_static/sphinx_highlight.js"></script>
        <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
        <script src="https://cdn.jsdelivr.net/npm/@jupyter-widgets/html-manager@^1.0.1/dist/embed-amd.js"></script>
        <script>window.MathJax = {"tex": {"inlineMath": [["$", "$"], ["\\(", "\\)"]], "processEscapes": true}, "options": {"ignoreHtmlClass": "tex2jax_ignore|mathjax_ignore|document", "processHtmlClass": "tex2jax_process|mathjax_process|math|output_area"}}</script>
        <script defer="defer" src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
        <script src="../_static/dark_mode_js/default_light.js"></script>
        <script src="../_static/dark_mode_js/theme_switcher.js"></script>
    <script src="../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="Credits" href="../authors.html" />
    <link rel="prev" title="End to end selection (short version)" href="spapros_tutorial_end_to_end_selection_short.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="../index.html" class="icon icon-home">
            spapros
              <img src="../_static/logo.png" class="logo" alt="Logo"/>
          </a>
              <div class="version">
                0.1.4
              </div>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">CONTENTS:</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../installation.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../usage.html">Usage</a></li>
<li class="toctree-l1"><a class="reference internal" href="../api.html">API</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="../tutorials.html">Tutorials</a><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="spapros_tutorial_basic_selection.html">Probeset selection with Spapros</a></li>
<li class="toctree-l2"><a class="reference internal" href="spapros_tutorial_basic_evaluation.html">Evaluation of probe sets</a></li>
<li class="toctree-l2"><a class="reference internal" href="spapros_tutorial_advanced_selection.html">Advanced probeset selection</a></li>
<li class="toctree-l2"><a class="reference internal" href="spapros_tutorial_advanced_evaluation.html">Advanced evaluation</a></li>
<li class="toctree-l2"><a class="reference internal" href="spapros_tutorial_end_to_end_selection_short.html">End to end selection (short version)</a></li>
<li class="toctree-l2 current"><a class="current reference internal" href="#">End to end selection</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#Import-Packages">Import Packages</a></li>
<li class="toctree-l3"><a class="reference internal" href="#Load-and-Preprocess-Data">Load and Preprocess Data</a></li>
<li class="toctree-l3"><a class="reference internal" href="#Probeset-Design">Probeset Design</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#Parameters-for-annotation-loader">Parameters for annotation loader</a></li>
<li class="toctree-l4"><a class="reference internal" href="#Parameters-for-Probe-Sequences-Database">Parameters for Probe Sequences Database</a></li>
<li class="toctree-l4"><a class="reference internal" href="#Parameters-for-Property-Filters">Parameters for Property Filters</a></li>
<li class="toctree-l4"><a class="reference internal" href="#Parameters-for-Specificity-Filters">Parameters for Specificity Filters</a></li>
<li class="toctree-l4"><a class="reference internal" href="#Parameters-for-Oligo-Efficiency-Score">Parameters for Oligo Efficiency Score</a></li>
<li class="toctree-l4"><a class="reference internal" href="#Parameters-for-Oligosets-Generation">Parameters for Oligosets Generation</a></li>
<li class="toctree-l4"><a class="reference internal" href="#Parameters-for-Padlock-Final-Sequence-Design">Parameters for Padlock Final Sequence Design</a></li>
</ul>
</li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../authors.html">Credits</a></li>
<li class="toctree-l1"><a class="reference internal" href="../code_of_conduct.html">Contributor Covenant Code of Conduct</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../index.html">spapros</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../index.html" class="icon icon-home" aria-label="Home"></a></li>
          <li class="breadcrumb-item"><a href="../tutorials.html">Tutorials</a></li>
      <li class="breadcrumb-item active">End to end selection</li>
      <li class="wy-breadcrumbs-aside">
            <a href="../_sources/_tutorials/spapros_tutorial_end_to_end_selection.ipynb.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <div class="section" id="End-to-end-selection">
<h1>End to end selection<a class="headerlink" href="#End-to-end-selection" title="Permalink to this heading"></a></h1>
<p>In this tutorial we showcase how to design probesets and select a suitable gene set with the <code class="docutils literal notranslate"><span class="pre">spapros</span></code> package. For all genes we design probes that fulfill experiment specific requirements and select only genes for which we can design sufficient probes. Spapros then selects genes that can distinguish the cell types in the data set and capture transcriptomic varation beyond cell type labels. The final probe sequences are designed in a last step for all selected genes. The figure below gives and
overview on the pipeline.</p>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[2]:
</pre></div>
</div>
<div class="output_area docutils container">
<img alt="../_images/_tutorials_spapros_tutorial_end_to_end_selection_2_0.png" class="no-scaled-link" src="../_images/_tutorials_spapros_tutorial_end_to_end_selection_2_0.png" style="width: 900px;" />
</div>
</div>
<div class="section" id="Import-Packages">
<h2>Import Packages<a class="headerlink" href="#Import-Packages" title="Permalink to this heading"></a></h2>
<p>Besides <code class="docutils literal notranslate"><span class="pre">spapros</span></code> also install <code class="docutils literal notranslate"><span class="pre">oligo_designer_toolsuite</span></code> if not done already. First we need to install some dependencies:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>conda<span class="w"> </span>config<span class="w"> </span>--add<span class="w"> </span>channels<span class="w"> </span>bioconda
conda<span class="w"> </span>config<span class="w"> </span>--add<span class="w"> </span>channels<span class="w"> </span>conda-forge
conda<span class="w"> </span>update<span class="w"> </span>conda
conda<span class="w"> </span>update<span class="w"> </span>--all

conda<span class="w"> </span>install<span class="w"> </span><span class="s2">&quot;blast&gt;=2.12&quot;</span>
conda<span class="w"> </span>install<span class="w"> </span><span class="s2">&quot;bedtools&gt;=2.30&quot;</span>
conda<span class="w"> </span>install<span class="w"> </span><span class="s2">&quot;bowtie&gt;=1.3.1&quot;</span>
conda<span class="w"> </span>install<span class="w"> </span><span class="s2">&quot;bowtie2&gt;=2.5&quot;</span>
</pre></div>
</div>
<p>To run the code below we need to install the current dev version of the oligo designer:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>git<span class="w"> </span>clone<span class="w"> </span>https://github.com/HelmholtzAI-Consultants-Munich/oligo-designer-toolsuite.git
<span class="nb">cd</span><span class="w"> </span>oligo-designer-toolsuite
git<span class="w"> </span>switch<span class="w"> </span>pipelines
pip<span class="w"> </span>install<span class="w"> </span>-e<span class="w"> </span>.
</pre></div>
</div>
<p>Otherwise, if that didn’t work, try:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>pip<span class="w"> </span>install<span class="w"> </span>oligo_designer_toolsuite
</pre></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[3]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">scanpy</span> <span class="k">as</span> <span class="nn">sc</span>
<span class="n">sc</span><span class="o">.</span><span class="n">settings</span><span class="o">.</span><span class="n">verbosity</span> <span class="o">=</span> <span class="mi">0</span>
<span class="n">sc</span><span class="o">.</span><span class="n">logging</span><span class="o">.</span><span class="n">print_header</span><span class="p">()</span>

<span class="kn">import</span> <span class="nn">spapros</span> <span class="k">as</span> <span class="nn">sp</span>
<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;spapros==</span><span class="si">{</span><span class="n">sp</span><span class="o">.</span><span class="n">__version__</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

<span class="kn">from</span> <span class="nn">oligo_designer_toolsuite.pipelines</span> <span class="kn">import</span> <span class="n">ScrinshotProbeDesigner</span> <span class="c1">#, MerfishProbeDesigner, SeqfishPlusProbeDesigner</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
scanpy==1.9.3 anndata==0.9.2 umap==0.5.3 numpy==1.24.4 scipy==1.11.1 pandas==1.5.3 scikit-learn==1.3.0 statsmodels==0.14.0 python-igraph==0.9.11 pynndescent==0.5.10
spapros==0.1.3
</pre></div></div>
</div>
</div>
<div class="section" id="Load-and-Preprocess-Data">
<h2>Load and Preprocess Data<a class="headerlink" href="#Load-and-Preprocess-Data" title="Permalink to this heading"></a></h2>
<p>For this tutorial, we use a PBMC example scRNA-seq reference dataset. The count data should be log-normalised and genes should <strong>not</strong> be scaled to mean=0 and std=1. We can load the processed version of the data, including cell / gene filters, cell type annotations, and the umap embedding, directly with <code class="docutils literal notranslate"><span class="pre">sp.ut.get_processed_pbmc_data()</span></code> function. For a step by step processing of the PBMC dataset please refer to the <a class="reference internal" href="spapros_tutorial_basic_selection.html"><span class="doc">basic usage tutorial</span></a>. For sake
of simplicity, we pre-select the top 1000 highly variable genes for the probe and geneset selection. In real world applications we typically go for top 8000 genes.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[5]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">pbmc_data</span> <span class="o">=</span> <span class="n">sp</span><span class="o">.</span><span class="n">ut</span><span class="o">.</span><span class="n">get_processed_pbmc_data</span><span class="p">(</span><span class="n">n_hvg</span><span class="o">=</span><span class="mi">1000</span><span class="p">)</span>
<span class="n">highly_variable_genes</span> <span class="o">=</span> <span class="nb">sorted</span><span class="p">(</span><span class="n">pbmc_data</span><span class="o">.</span><span class="n">var</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">pbmc_data</span><span class="o">.</span><span class="n">var</span><span class="p">[</span><span class="s1">&#39;highly_variable&#39;</span><span class="p">]]</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">tolist</span><span class="p">())</span>
<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Number of highly variable genes: </span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">highly_variable_genes</span><span class="p">)</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Number of highly variable genes: 1000
</pre></div></div>
</div>
</div>
<div class="section" id="Probeset-Design">
<h2>Probeset Design<a class="headerlink" href="#Probeset-Design" title="Permalink to this heading"></a></h2>
<p>Before choosing a gene panel, we design probesets for our given set of 1000 highly variable genes that fulfill certain experiment-specific criteria. Therefore, we first create an instance of a ProbeDesigner class, where we can choose from <code class="docutils literal notranslate"><span class="pre">ScrinshotProbeDesigner</span></code>, <code class="docutils literal notranslate"><span class="pre">MerfishProbeDesigner</span></code> an <code class="docutils literal notranslate"><span class="pre">SeqfishPlusProbeDesigner</span></code> (see our <a class="reference external" href="https://www.google.com/url?q=https://docs.google.com/spreadsheets/d/1NCDLscdmzn32U7_IKy6OKYHqfECn76x0pAD3KFhuJgQ/edit%23gid%3D0&amp;sa=D&amp;source=docs&amp;ust=1692975800552487&amp;usg=AOvVaw20CRwiObnVkWVS54CcqXMq">resource
table</a> for an overview of differences between the technologies). For each of those classes, we need to define an output directory and set the parameters <code class="docutils literal notranslate"><span class="pre">write_removed_genes</span></code> (if true, save gene with insufficient probes in a file) and <code class="docutils literal notranslate"><span class="pre">write_intermediate_steps</span></code> (if true, save the probe
database after each processing step, such that the pipline can resume from a certain step onwards).</p>
<p>Here, we showcase the probe design of padlock probes.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[6]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">probe_designer</span> <span class="o">=</span> <span class="n">ScrinshotProbeDesigner</span><span class="p">(</span><span class="n">dir_output</span><span class="o">=</span><span class="s2">&quot;./output&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
2023-08-22 17:25:41,843 [INFO] Parameters Init:
2023-08-22 17:25:41,844 [INFO] dir_output = ./output
2023-08-22 17:25:41,845 [INFO] write_removed_genes = True
2023-08-22 17:25:41,846 [INFO] write_intermediate_steps = True
</pre></div></div>
</div>
<p>After instatiating a ProbeDesigner class, we need to load the annotation we are using. Our example dataset uses the NCBI gene annotation. Hence, we define <em>ncbi</em> as source and define the NCBI-specific parameters <em>taxon</em>, <em>species</em> and <em>annotation_release</em>. Apart from <em>NCBI</em> annotation, we can also choose an <em>Ensembl</em> annotation. If <code class="docutils literal notranslate"><span class="pre">source=&quot;ncbi&quot;</span></code> or <code class="docutils literal notranslate"><span class="pre">source=&quot;ensembl&quot;</span></code> is choosen, the annotation files are automatically downloaded from their servers. In addition, we can provide a custom
annotation when specifying <code class="docutils literal notranslate"><span class="pre">source=&quot;custom&quot;</span></code>.</p>
<div class="section" id="Parameters-for-annotation-loader">
<h3>Parameters for annotation loader<a class="headerlink" href="#Parameters-for-annotation-loader" title="Permalink to this heading"></a></h3>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">source</span></code>: define annotation source -&gt; currently supported: ncbi, ensembl and custom</p></li>
</ul>
<p><em>NCBI annnotation parameters:</em> - <code class="docutils literal notranslate"><span class="pre">taxon</span></code>: taxon of the species, valid taxa are: archaea, bacteria, fungi, invertebrate, mitochondrion, plant, plasmid, plastid, protozoa, vertebrate_mammalian, vertebrate_other, viral - <code class="docutils literal notranslate"><span class="pre">species</span></code>: species name in NCBI download format, e.g. ‘Homo_sapiens’ for human; see <a class="reference external" href="https://ftp.ncbi.nlm.nih.gov/genomes/refseq/">here</a> for available species name - <code class="docutils literal notranslate"><span class="pre">annotation_release</span></code>: release number (e.g. 109 or 109.20211119 for ncbi) of annotation or ‘current’ to use
most recent annotation release. Check out release numbers for NCBI at ftp.ncbi.nlm.nih.gov/refseq/H_sapiens/annotation/annotation_releases/</p>
<p><em>Ensembl annotation parameters:</em> - <code class="docutils literal notranslate"><span class="pre">species</span></code>: species name in ensembl download format, e.g. ‘homo_sapiens’ for human; see <a class="reference external" href="http://ftp.ensembl.org/pub/release-108/gtf/">http://ftp.ensembl.org/pub/release-108/gtf/</a> for available species names - <code class="docutils literal notranslate"><span class="pre">annotation_release</span></code>: release number of annotation, e.g. ‘release-108’ or ‘current’ to use most recent annotation release. Check out release numbers for ensemble at ftp.ensembl.org/pub/</p>
<p><em>Custom annotation parameters:</em> - <code class="docutils literal notranslate"><span class="pre">file_annotation</span></code>: GTF file with gene annotation - <code class="docutils literal notranslate"><span class="pre">file_sequence</span></code>: FASTA file with genome sequence - <code class="docutils literal notranslate"><span class="pre">files_source</span></code>: original source of the genomic files -&gt; optional - <code class="docutils literal notranslate"><span class="pre">species</span></code>: species of provided annotation, leave empty if unknown -&gt; optional - <code class="docutils literal notranslate"><span class="pre">annotation_release</span></code>: release number of provided annotation, leave empty if unknown -&gt; optional - <code class="docutils literal notranslate"><span class="pre">genome_assembly</span></code>: genome assembly of provided annotation, leave empty if unknown -&gt; optional</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[6]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># example for ncbi annotation loader</span>
<span class="n">source</span> <span class="o">=</span> <span class="s2">&quot;ncbi&quot;</span>
<span class="n">params</span> <span class="o">=</span> <span class="p">{</span>
    <span class="s2">&quot;taxon&quot;</span><span class="p">:</span> <span class="s2">&quot;vertebrate_mammalian&quot;</span><span class="p">,</span>
    <span class="s2">&quot;species&quot;</span><span class="p">:</span> <span class="s2">&quot;Homo_sapiens&quot;</span><span class="p">,</span>
    <span class="s2">&quot;annotation_release&quot;</span><span class="p">:</span> <span class="s2">&quot;110&quot;</span><span class="p">,</span>
<span class="p">}</span>

<span class="c1"># example for ensembl annotation loader</span>
<span class="c1"># source = &quot;ensembl&quot;</span>
<span class="c1"># params = {</span>
<span class="c1">#     &quot;species&quot;: &quot;homo_sapiens&quot;,</span>
<span class="c1">#     &quot;annotation_release&quot;: &quot;109&quot;,</span>
<span class="c1"># }</span>

<span class="c1"># example for custom annotation loader</span>
<span class="c1"># source = &quot;custom&quot;</span>
<span class="c1"># params = {</span>
<span class="c1">#     &quot;file_annotation&quot;: &quot;./output/annotation/GCF_000001405.40_GRCh38.p14_genomic.gtf&quot;,</span>
<span class="c1">#     &quot;file_sequence&quot;: &quot;./output/annotation/GCF_000001405.40_GRCh38.p14_genomic.fna&quot;,</span>
<span class="c1">#     &quot;files_source&quot;: &quot;NCBI&quot;,</span>
<span class="c1">#     &quot;species&quot;: &quot;Homo_sapiens&quot;,</span>
<span class="c1">#     &quot;annotation_release&quot;: &quot;110&quot;,</span>
<span class="c1">#     &quot;genome_assembly&quot;: &quot;GRCh38.p14&quot;,</span>
<span class="c1"># }</span>

<span class="n">probe_designer</span><span class="o">.</span><span class="n">load_annotations</span><span class="p">(</span><span class="n">source</span><span class="o">=</span><span class="n">source</span><span class="p">,</span> <span class="n">source_params</span><span class="o">=</span><span class="n">params</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
2023-05-26 17:13:37,570 [INFO] Parameters Load Annotations:
2023-05-26 17:13:37,572 [INFO] source = custom
2023-05-26 17:13:37,573 [INFO] source_params = {&#39;file_annotation&#39;: &#39;./output/annotation/GCF_000001405.40_GRCh38.p14_genomic.gtf&#39;, &#39;file_sequence&#39;: &#39;./output/annotation/GCF_000001405.40_GRCh38.p14_genomic.fna&#39;, &#39;files_source&#39;: &#39;NCBI&#39;, &#39;species&#39;: &#39;Homo_sapiens&#39;, &#39;annotation_release&#39;: &#39;110&#39;, &#39;genome_assembly&#39;: &#39;GRCh38.p14&#39;}
2023-05-26 17:16:31,649 [INFO] The following annotation files are used for GTF annotation of regions: ./output/annotation/GCF_000001405.40_GRCh38.p14_genomic.gtf and for fasta sequence file: ./output/annotation/GCF_000001405.40_GRCh38.p14_genomic.fna .
2023-05-26 17:16:31,653 [INFO] The annotations are from NCBI source, for the species: Homo_sapiens, release number: 110 and genome assembly: GRCh38.p14
</pre></div></div>
</div>
<p>After downloading the annotations, we have to create the oligo database. Running the function below, will automatically create a transcriptome from the given annotation (therefore, the provided GTF file must contain transcript and exon information) and use this transcriptome to create all possible probes for each gene, that is provided in the <em>gene</em> list.</p>
</div>
<div class="section" id="Parameters-for-Probe-Sequences-Database">
<h3>Parameters for Probe Sequences Database<a class="headerlink" href="#Parameters-for-Probe-Sequences-Database" title="Permalink to this heading"></a></h3>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">probe_length_min</span></code>: minimum length of probes</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">probe_length_max</span></code>: maximum length of probes</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">min_probes_per_gene</span></code>: minimum number of probes that a gene must have before it gets deleted</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">region</span></code>: Target sequence type for which probes are designed (choose from: “transcript”, “genome”, “cds”)</p></li>
</ul>
<div class="line-block">
<div class="line"><em>Note: Instead of creating a new probe database, we can also load an existing databases.</em></div>
<div class="line">Loading a database can be useful when starting the pipeline from a certain step, e.g. load a database which was already filtered by probe properties and continue immediately with the specificity filter step. We can load an existing database by calling <code class="docutils literal notranslate"><span class="pre">load_probe_database()</span></code>. See example code in the cells below (commented).</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[7]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">probe_length_min</span> <span class="o">=</span> <span class="mi">38</span>
<span class="n">probe_length_max</span> <span class="o">=</span> <span class="mi">45</span>
<span class="n">min_probes_per_gene</span> <span class="o">=</span> <span class="mi">3</span>
<span class="n">region</span> <span class="o">=</span> <span class="s2">&quot;transcript&quot;</span>

<span class="c1"># highly_variable_genes = highly_variable_genes[100:]</span>
<span class="n">probe_database</span><span class="p">,</span> <span class="n">file_database</span> <span class="o">=</span> <span class="n">probe_designer</span><span class="o">.</span><span class="n">create_probe_database</span><span class="p">(</span><span class="n">genes</span><span class="o">=</span><span class="n">highly_variable_genes</span><span class="p">,</span> <span class="n">probe_length_min</span><span class="o">=</span><span class="n">probe_length_min</span><span class="p">,</span> <span class="n">probe_length_max</span><span class="o">=</span><span class="n">probe_length_max</span><span class="p">,</span> <span class="n">region</span><span class="o">=</span><span class="n">region</span><span class="p">,</span> <span class="n">min_probes_per_gene</span><span class="o">=</span><span class="n">min_probes_per_gene</span><span class="p">,</span> <span class="n">n_jobs</span><span class="o">=</span><span class="mi">4</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
2023-05-26 17:16:31,766 [INFO] Parameters Create Database:
2023-05-26 17:16:31,767 [INFO] genes = [&#39;AAGAB&#39;, &#39;AATF&#39;, &#39;ABCC10&#39;, &#39;ABHD12&#39;, &#39;ABHD17B&#39;, &#39;ABHD5&#39;, &#39;ABRACL&#39;, &#39;ABT1&#39;, &#39;AC005082.12&#39;, &#39;AC074138.3&#39;, &#39;AC093323.3&#39;, &#39;ACAP1&#39;, &#39;ACBD3&#39;, &#39;ACD&#39;, &#39;ACOT13&#39;, &#39;ACP1&#39;, &#39;ACRBP&#39;, &#39;ACTL6A&#39;, &#39;ACTR6&#39;, &#39;ACVR2A&#39;, &#39;ADAL&#39;, &#39;ADAM10&#39;, &#39;ADAM28&#39;, &#39;ADD1&#39;, &#39;ADIPOR2&#39;, &#39;ADPRM&#39;, &#39;ADSL&#39;, &#39;AEBP1&#39;, &#39;AGPAT1&#39;, &#39;AHSA1&#39;, &#39;AIF1&#39;, &#39;AIM2&#39;, &#39;AKTIP&#39;, &#39;AL928768.3&#39;, &#39;ALKBH7&#39;, &#39;ANAPC13&#39;, &#39;ANKAR&#39;, &#39;ANKEF1&#39;, &#39;ANKRD27&#39;, &#39;ANKRD54&#39;, &#39;AP001462.6&#39;, &#39;AP003419.16&#39;, &#39;AP3M2&#39;, &#39;AP4B1-AS1&#39;, &#39;AP4S1&#39;, &#39;APOBEC3A&#39;, &#39;APOBEC3B&#39;, &#39;APOBEC3G&#39;, &#39;AQP3&#39;, &#39;ARHGAP11A&#39;, &#39;ARHGAP19&#39;, &#39;ARHGAP24&#39;, &#39;ARHGAP33&#39;, &#39;ARHGAP6&#39;, &#39;ARID4A&#39;, &#39;ARIH2OS&#39;, &#39;ARL2&#39;, &#39;ARL2BP&#39;, &#39;ARL4A&#39;, &#39;ARL6IP5&#39;, &#39;ARMC7&#39;, &#39;ARMCX5&#39;, &#39;ARRDC3&#39;, &#39;ARRDC4&#39;, &#39;ARSD&#39;, &#39;ARSG&#39;, &#39;ARVCF&#39;, &#39;ASB8&#39;, &#39;ASXL2&#39;, &#39;ATAD3C&#39;, &#39;ATF7IP2&#39;, &#39;ATG16L1&#39;, &#39;ATP10A&#39;, &#39;ATP5H&#39;, &#39;ATP5O&#39;, &#39;ATP5SL&#39;, &#39;ATP6V0E2&#39;, &#39;ATXN1L&#39;, &#39;ATXN3&#39;, &#39;AURKC&#39;, &#39;BABAM1&#39;, &#39;BACE2&#39;, &#39;BAZ2A&#39;, &#39;BBX&#39;, &#39;BCDIN3D&#39;, &#39;BET1&#39;, &#39;BEX4&#39;, &#39;BGLAP&#39;, &#39;BLNK&#39;, &#39;BLZF1&#39;, &#39;BMPR2&#39;, &#39;BNIP2&#39;, &#39;BOLA1&#39;, &#39;BOLA3&#39;, &#39;BRAT1&#39;, &#39;BRWD1&#39;, &#39;BTN3A1&#39;, &#39;BTN3A2&#39;, &#39;BUB3&#39;, &#39;C10orf32&#39;, &#39;C12orf45&#39;, &#39;C14orf1&#39;, &#39;C14orf166&#39;, &#39;C14orf80&#39;, &#39;C15orf57&#39;, &#39;C16orf13&#39;, &#39;C16orf52&#39;, &#39;C16orf54&#39;, &#39;C16orf58&#39;, &#39;C16orf74&#39;, &#39;C16orf80&#39;, &#39;C17orf59&#39;, &#39;C17orf62&#39;, &#39;C19orf33&#39;, &#39;C19orf52&#39;, &#39;C1QA&#39;, &#39;C1QB&#39;, &#39;C1QC&#39;, &#39;C1orf162&#39;, &#39;C1orf35&#39;, &#39;C21orf33&#39;, &#39;C2CD4D&#39;, &#39;C2orf76&#39;, &#39;C2orf88&#39;, &#39;C3orf18&#39;, &#39;C5orf15&#39;, &#39;C5orf42&#39;, &#39;C8orf44&#39;, &#39;C9orf142&#39;, &#39;C9orf16&#39;, &#39;C9orf37&#39;, &#39;CAMK1D&#39;, &#39;CAMK2G&#39;, &#39;CAMK2N1&#39;, &#39;CAPN12&#39;, &#39;CARHSP1&#39;, &#39;CARS&#39;, &#39;CASC4&#39;, &#39;CBX5&#39;, &#39;CCDC115&#39;, &#39;CCDC122&#39;, &#39;CCDC66&#39;, &#39;CCDC91&#39;, &#39;CCL3&#39;, &#39;CCL4&#39;, &#39;CCL5&#39;, &#39;CCND2&#39;, &#39;CCNG1&#39;, &#39;CCP110&#39;, &#39;CCT4&#39;, &#39;CCT7&#39;, &#39;CD160&#39;, &#39;CD19&#39;, &#39;CD2&#39;, &#39;CD247&#39;, &#39;CD274&#39;, &#39;CD2AP&#39;, &#39;CD320&#39;, &#39;CD72&#39;, &#39;CD79A&#39;, &#39;CD79B&#39;, &#39;CD82&#39;, &#39;CD9&#39;, &#39;CD96&#39;, &#39;CDC123&#39;, &#39;CDC16&#39;, &#39;CDC37&#39;, &#39;CDC40&#39;, &#39;CDK19&#39;, &#39;CDKN2A&#39;, &#39;CEACAM4&#39;, &#39;CEBPB&#39;, &#39;CECR5&#39;, &#39;CEP120&#39;, &#39;CEP68&#39;, &#39;CEP85L&#39;, &#39;CEPT1&#39;, &#39;CES4A&#39;, &#39;CGRRF1&#39;, &#39;CHD2&#39;, &#39;CHD7&#39;, &#39;CHERP&#39;, &#39;CHI3L2&#39;, &#39;CHPF2&#39;, &#39;CIAPIN1&#39;, &#39;CISD1&#39;, &#39;CISH&#39;, &#39;CITED4&#39;, &#39;CKS1B&#39;, &#39;CLDN5&#39;, &#39;CLEC2B&#39;, &#39;CLIC3&#39;, &#39;CLNS1A&#39;, &#39;CLPX&#39;, &#39;CLU&#39;, &#39;CLYBL&#39;, &#39;CMTM5&#39;, &#39;CNEP1R1&#39;, &#39;COMMD10&#39;, &#39;COQ7&#39;, &#39;CORO1B&#39;, &#39;COTL1&#39;, &#39;CPNE2&#39;, &#39;CPQ&#39;, &#39;CPSF3L&#39;, &#39;CR1&#39;, &#39;CRIP3&#39;, &#39;CRTC2&#39;, &#39;CST3&#39;, &#39;CST7&#39;, &#39;CTA-29F11.1&#39;, &#39;CTB-113I20.2&#39;, &#39;CTB-152G17.6&#39;, &#39;CTC-444N24.11&#39;, &#39;CTD-2015H6.3&#39;, &#39;CTD-2302E22.4&#39;, &#39;CTD-2368P22.1&#39;, &#39;CTD-2537I9.12&#39;, &#39;CTSS&#39;, &#39;CTSW&#39;, &#39;CWC15&#39;, &#39;CWC27&#39;, &#39;CXCL10&#39;, &#39;CXCL3&#39;, &#39;CYB5B&#39;, &#39;CYTH2&#39;, &#39;DAGLB&#39;, &#39;DCAF5&#39;, &#39;DDI2&#39;, &#39;DDT&#39;, &#39;DDX1&#39;, &#39;DDX17&#39;, &#39;DDX46&#39;, &#39;DDX56&#39;, &#39;DENND1C&#39;, &#39;DENND2D&#39;, &#39;DENND5B&#39;, &#39;DENND6A&#39;, &#39;DERL1&#39;, &#39;DEXI&#39;, &#39;DHX34&#39;, &#39;DHX9&#39;, &#39;DIDO1&#39;, &#39;DIMT1&#39;, &#39;DIS3&#39;, &#39;DISP1&#39;, &#39;DLST&#39;, &#39;DMTN&#39;, &#39;DNAJA3&#39;, &#39;DNAJB14&#39;, &#39;DNAJC10&#39;, &#39;DNAJC15&#39;, &#39;DNAJC2&#39;, &#39;DNAJC27&#39;, &#39;DNASE1L3&#39;, &#39;DNMT3A&#39;, &#39;DOK3&#39;, &#39;DPH6&#39;, &#39;DPY19L4&#39;, &#39;DRAXIN&#39;, &#39;DSCR3&#39;, &#39;DTX3&#39;, &#39;DUS3L&#39;, &#39;DUSP10&#39;, &#39;EAF2&#39;, &#39;EARS2&#39;, &#39;ECHDC1&#39;, &#39;EDC3&#39;, &#39;EID2&#39;, &#39;EIF1AY&#39;, &#39;EIF1B&#39;, &#39;EIF2B1&#39;, &#39;EIF3D&#39;, &#39;ELANE&#39;, &#39;ELOF1&#39;, &#39;ELOVL4&#39;, &#39;ELP6&#39;, &#39;EMB&#39;, &#39;EMG1&#39;, &#39;EML6&#39;, &#39;ENTPD3-AS1&#39;, &#39;EOGT&#39;, &#39;ERH&#39;, &#39;ERV3-1&#39;, &#39;EVA1B&#39;, &#39;EWSR1&#39;, &#39;EXOC6&#39;, &#39;F5&#39;, &#39;FADS1&#39;, &#39;FAM107B&#39;, &#39;FAM173A&#39;, &#39;FAM210B&#39;, &#39;FAM96A&#39;, &#39;FAM98A&#39;, &#39;FBXL14&#39;, &#39;FBXO21&#39;, &#39;FBXO33&#39;, &#39;FBXO4&#39;, &#39;FBXW4&#39;, &#39;FCER1A&#39;, &#39;FCER1G&#39;, &#39;FCGR2B&#39;, &#39;FCGR3A&#39;, &#39;FCN1&#39;, &#39;FCRLA&#39;, &#39;FEM1A&#39;, &#39;FERMT3&#39;, &#39;FGFBP2&#39;, &#39;FH&#39;, &#39;FHL1&#39;, &#39;FKBP3&#39;, &#39;FKBP5&#39;, &#39;FLOT1&#39;, &#39;FMO4&#39;, &#39;FN3KRP&#39;, &#39;FNBP4&#39;, &#39;FNTA&#39;, &#39;FOPNL&#39;, &#39;FRY-AS1&#39;, &#39;FUS&#39;, &#39;FXN&#39;, &#39;FYB&#39;, &#39;G0S2&#39;, &#39;GADD45B&#39;, &#39;GALT&#39;, &#39;GBGT1&#39;, &#39;GBP1&#39;, &#39;GDF11&#39;, &#39;GFER&#39;, &#39;GGA3&#39;, &#39;GGNBP2&#39;, &#39;GIMAP2&#39;, &#39;GIMAP4&#39;, &#39;GIMAP5&#39;, &#39;GIMAP7&#39;, &#39;GIT2&#39;, &#39;GMPPA&#39;, &#39;GNE&#39;, &#39;GNG11&#39;, &#39;GNG3&#39;, &#39;GNLY&#39;, &#39;GNPAT&#39;, &#39;GOLGB1&#39;, &#39;GP9&#39;, &#39;GPATCH4&#39;, &#39;GPKOW&#39;, &#39;GPR171&#39;, &#39;GPR183&#39;, &#39;GPR35&#39;, &#39;GPS1&#39;, &#39;GPX1&#39;, &#39;GRAP&#39;, &#39;GRN&#39;, &#39;GSTP1&#39;, &#39;GTPBP6&#39;, &#39;GUSB&#39;, &#39;GYS1&#39;, &#39;GZMA&#39;, &#39;GZMB&#39;, &#39;GZMH&#39;, &#39;GZMK&#39;, &#39;HAGH&#39;, &#39;HBA1&#39;, &#39;HBP1&#39;, &#39;HCFC2&#39;, &#39;HDAC1&#39;, &#39;HDAC5&#39;, &#39;HDAC9&#39;, &#39;HELQ&#39;, &#39;HEMK1&#39;, &#39;HERPUD2&#39;, &#39;HIST1H1B&#39;, &#39;HIST1H2AC&#39;, &#39;HIST1H2AH&#39;, &#39;HLA-DMA&#39;, &#39;HLA-DMB&#39;, &#39;HLA-DOB&#39;, &#39;HLA-DPA1&#39;, &#39;HLA-DPB1&#39;, &#39;HLA-DQA1&#39;, &#39;HLA-DQB1&#39;, &#39;HLA-DRB1&#39;, &#39;HMBOX1&#39;, &#39;HMGCL&#39;, &#39;HMGXB4&#39;, &#39;HNRNPH3&#39;, &#39;HOOK2&#39;, &#39;HOPX&#39;, &#39;HSPB11&#39;, &#39;HVCN1&#39;, &#39;ICAM2&#39;, &#39;ICOS&#39;, &#39;ICOSLG&#39;, &#39;ID2&#39;, &#39;IDUA&#39;, &#39;IFFO1&#39;, &#39;IFI27&#39;, &#39;IFIT1&#39;, &#39;IFIT2&#39;, &#39;IFITM3&#39;, &#39;IGFBP7&#39;, &#39;IGJ&#39;, &#39;IGLL5&#39;, &#39;IL1B&#39;, &#39;IL1RAP&#39;, &#39;IL23A&#39;, &#39;IL24&#39;, &#39;IL27RA&#39;, &#39;IL32&#39;, &#39;IL6&#39;, &#39;IL8&#39;, &#39;ILF3&#39;, &#39;ILF3-AS1&#39;, &#39;ING5&#39;, &#39;INSL3&#39;, &#39;INTS12&#39;, &#39;INTS2&#39;, &#39;IP6K1&#39;, &#39;IQCE&#39;, &#39;IRF8&#39;, &#39;IRF9&#39;, &#39;ISCA2&#39;, &#39;ISOC1&#39;, &#39;ITGA2B&#39;, &#39;ITGB7&#39;, &#39;ITM2A&#39;, &#39;ITSN2&#39;, &#39;JAKMIP1&#39;, &#39;JUND&#39;, &#39;KARS&#39;, &#39;KCNG1&#39;, &#39;KCNQ1OT1&#39;, &#39;KIAA0040&#39;, &#39;KIAA0125&#39;, &#39;KIAA0196&#39;, &#39;KIAA1430&#39;, &#39;KIF3A&#39;, &#39;KIF3C&#39;, &#39;KIF5B&#39;, &#39;KLHL24&#39;, &#39;KLRB1&#39;, &#39;KLRG1&#39;, &#39;KRBOX4&#39;, &#39;LAMP3&#39;, &#39;LARS&#39;, &#39;LAT2&#39;, &#39;LBR&#39;, &#39;LDLRAP1&#39;, &#39;LGALS1&#39;, &#39;LGALS2&#39;, &#39;LGALS3&#39;, &#39;LILRA4&#39;, &#39;LIN52&#39;, &#39;LINC00494&#39;, &#39;LINC00662&#39;, &#39;LINC00886&#39;, &#39;LINC00926&#39;, &#39;LINC00936&#39;, &#39;LINC01013&#39;, &#39;LIX1L&#39;, &#39;LONRF1&#39;, &#39;LPIN1&#39;, &#39;LRBA&#39;, &#39;LRRIQ3&#39;, &#39;LSM14A&#39;, &#39;LST1&#39;, &#39;LTB&#39;, &#39;LTV1&#39;, &#39;LUC7L&#39;, &#39;LUC7L3&#39;, &#39;LYAR&#39;, &#39;LYPD2&#39;, &#39;LYPLA1&#39;, &#39;LYRM4&#39;, &#39;LYSMD4&#39;, &#39;LZTS2&#39;, &#39;MADD&#39;, &#39;MAEA&#39;, &#39;MAGEH1&#39;, &#39;MAL&#39;, &#39;MALT1&#39;, &#39;MAP2K7&#39;, &#39;MARCKSL1&#39;, &#39;MCF2L&#39;, &#39;MCM3&#39;, &#39;MDS2&#39;, &#39;MED30&#39;, &#39;MED9&#39;, &#39;METTL21A&#39;, &#39;METTL3&#39;, &#39;METTL8&#39;, &#39;MFF&#39;, &#39;MFSD10&#39;, &#39;MIS18A&#39;, &#39;MKKS&#39;, &#39;MLLT11&#39;, &#39;MLLT6&#39;, &#39;MMADHC&#39;, &#39;MMP9&#39;, &#39;MNAT1&#39;, &#39;MOCS2&#39;, &#39;MORF4L2&#39;, &#39;MPHOSPH10&#39;, &#39;MRM1&#39;, &#39;MRPL1&#39;, &#39;MRPL19&#39;, &#39;MRPL42&#39;, &#39;MRPS12&#39;, &#39;MRPS33&#39;, &#39;MS4A1&#39;, &#39;MS4A6A&#39;, &#39;MTERFD2&#39;, &#39;MTIF2&#39;, &#39;MTRF1&#39;, &#39;MUM1&#39;, &#39;MYADM&#39;, &#39;MYCBP2&#39;, &#39;MYL9&#39;, &#39;MYO1E&#39;, &#39;MYOM2&#39;, &#39;MZB1&#39;, &#39;MZT1&#39;, &#39;NAA20&#39;, &#39;NAP1L4&#39;, &#39;NAPA-AS1&#39;, &#39;NARG2&#39;, &#39;NAT9&#39;, &#39;NBR1&#39;, &#39;NCOR2&#39;, &#39;NCR3&#39;, &#39;NDUFA10&#39;, &#39;NDUFA12&#39;, &#39;NECAB3&#39;, &#39;NEFH&#39;, &#39;NEK8&#39;, &#39;NELFB&#39;, &#39;NEMF&#39;, &#39;NFAT5&#39;, &#39;NFE2L2&#39;, &#39;NFIC&#39;, &#39;NFU1&#39;, &#39;NIT2&#39;, &#39;NKAP&#39;, &#39;NKG7&#39;, &#39;NKTR&#39;, &#39;NME3&#39;, &#39;NME6&#39;, &#39;NMNAT3&#39;, &#39;NNT-AS1&#39;, &#39;NOC4L&#39;, &#39;NOG&#39;, &#39;NOL11&#39;, &#39;NONO&#39;, &#39;NOP58&#39;, &#39;NPC2&#39;, &#39;NPHP3&#39;, &#39;NPRL2&#39;, &#39;NR2C1&#39;, &#39;NR3C1&#39;, &#39;NSA2&#39;, &#39;NT5C&#39;, &#39;NT5C3A&#39;, &#39;NUDCD1&#39;, &#39;NUDT16L1&#39;, &#39;NUP54&#39;, &#39;NXT2&#39;, &#39;OARD1&#39;, &#39;OAT&#39;, &#39;OBSCN&#39;, &#39;ODC1&#39;, &#39;ORAI1&#39;, &#39;ORC2&#39;, &#39;OSBPL1A&#39;, &#39;OSBPL7&#39;, &#39;OXLD1&#39;, &#39;P2RX5&#39;, &#39;P2RY10&#39;, &#39;PACS1&#39;, &#39;PACSIN2&#39;, &#39;PAICS&#39;, &#39;PARP1&#39;, &#39;PARS2&#39;, &#39;PASK&#39;, &#39;PAWR&#39;, &#39;PAXIP1-AS1&#39;, &#39;PBLD&#39;, &#39;PBRM1&#39;, &#39;PCNA&#39;, &#39;PCSK7&#39;, &#39;PDCD1&#39;, &#39;PDCD2L&#39;, &#39;PDE6B&#39;, &#39;PDIA3&#39;, &#39;PDIK1L&#39;, &#39;PDK2&#39;, &#39;PDXDC1&#39;, &#39;PDZD4&#39;, &#39;PEMT&#39;, &#39;PEX16&#39;, &#39;PEX26&#39;, &#39;PF4&#39;, &#39;PGM1&#39;, &#39;PGM2L1&#39;, &#39;PHACTR4&#39;, &#39;PHF12&#39;, &#39;PHF14&#39;, &#39;PHF3&#39;, &#39;PIGF&#39;, &#39;PIGU&#39;, &#39;PIGX&#39;, &#39;PIK3R1&#39;, &#39;PITHD1&#39;, &#39;PITPNA-AS1&#39;, &#39;PJA1&#39;, &#39;PKIG&#39;, &#39;PLA2G12A&#39;, &#39;PLCL1&#39;, &#39;PLD6&#39;, &#39;PLEKHA1&#39;, &#39;PLEKHA3&#39;, &#39;PLRG1&#39;, &#39;PMEPA1&#39;, &#39;PNOC&#39;, &#39;POLR2I&#39;, &#39;POLR2K&#39;, &#39;POLR3E&#39;, &#39;POMT1&#39;, &#39;PPA2&#39;, &#39;PPBP&#39;, &#39;PPIE&#39;, &#39;PPIG&#39;, &#39;PPIL2&#39;, &#39;PPIL4&#39;, &#39;PPP1R14A&#39;, &#39;PPP1R2&#39;, &#39;PPP2R1B&#39;, &#39;PPP6C&#39;, &#39;PPT2-EGFL8&#39;, &#39;PQBP1&#39;, &#39;PRAF2&#39;, &#39;PRDX1&#39;, &#39;PRELID2&#39;, &#39;PRF1&#39;, &#39;PRICKLE1&#39;, &#39;PRKACB&#39;, &#39;PRKCB&#39;, &#39;PRKD2&#39;, &#39;PRMT2&#39;, &#39;PRNP&#39;, &#39;PRPF31&#39;, &#39;PRPS2&#39;, &#39;PRR5&#39;, &#39;PSMD14&#39;, &#39;PTCRA&#39;, &#39;PTGDR&#39;, &#39;PTGDS&#39;, &#39;PTGES2&#39;, &#39;PTPN7&#39;, &#39;PURA&#39;, &#39;PWP1&#39;, &#39;PXMP4&#39;, &#39;PYCARD&#39;, &#39;R3HDM1&#39;, &#39;R3HDM2&#39;, &#39;RAB40C&#39;, &#39;RABEP2&#39;, &#39;RABL6&#39;, &#39;RAD51B&#39;, &#39;RALBP1&#39;, &#39;RALY&#39;, &#39;RASD1&#39;, &#39;RASGRP2&#39;, &#39;RBM25&#39;, &#39;RBM26-AS1&#39;, &#39;RBM39&#39;, &#39;RBM4&#39;, &#39;RBM48&#39;, &#39;RBM5&#39;, &#39;RBM7&#39;, &#39;RBPJ&#39;, &#39;RCE1&#39;, &#39;RCHY1&#39;, &#39;RCL1&#39;, &#39;RCN2&#39;, &#39;RDH14&#39;, &#39;RELB&#39;, &#39;REXO2&#39;, &#39;RFC1&#39;, &#39;RFC5&#39;, &#39;RFNG&#39;, &#39;RFPL2&#39;, &#39;RGS14&#39;, &#39;RIC3&#39;, &#39;RIOK1&#39;, &#39;RIOK2&#39;, &#39;RNF113A&#39;, &#39;RNF125&#39;, &#39;RNF139&#39;, &#39;RNF14&#39;, &#39;RNF168&#39;, &#39;RNF187&#39;, &#39;RNF213&#39;, &#39;RNF25&#39;, &#39;RNF26&#39;, &#39;RORA&#39;, &#39;RP1-28O10.1&#39;, &#39;RP11-1055B8.7&#39;, &#39;RP11-138A9.2&#39;, &#39;RP11-141B14.1&#39;, &#39;RP11-142C4.6&#39;, &#39;RP11-162G10.5&#39;, &#39;RP11-164H13.1&#39;, &#39;RP11-178G16.4&#39;, &#39;RP11-18H21.1&#39;, &#39;RP11-211G3.2&#39;, &#39;RP11-219B17.1&#39;, &#39;RP11-219B4.7&#39;, &#39;RP11-252A24.3&#39;, &#39;RP11-291B21.2&#39;, &#39;RP11-314N13.3&#39;, &#39;RP11-324I22.4&#39;, &#39;RP11-349A22.5&#39;, &#39;RP11-378J18.3&#39;, &#39;RP11-390B4.5&#39;, &#39;RP11-398C13.6&#39;, &#39;RP11-400F19.6&#39;, &#39;RP11-421L21.3&#39;, &#39;RP11-428G5.5&#39;, &#39;RP11-432I5.1&#39;, &#39;RP11-468E2.4&#39;, &#39;RP11-488C13.5&#39;, &#39;RP11-493L12.4&#39;, &#39;RP11-527L4.5&#39;, &#39;RP11-545I5.3&#39;, &#39;RP11-589C21.6&#39;, &#39;RP11-5C23.1&#39;, &#39;RP11-701P16.5&#39;, &#39;RP11-706O15.1&#39;, &#39;RP11-70P17.1&#39;, &#39;RP11-727F15.9&#39;, &#39;RP11-798G7.6&#39;, &#39;RP11-879F14.2&#39;, &#39;RP11-950C14.3&#39;, &#39;RP3-325F22.5&#39;, &#39;RP5-1073O3.7&#39;, &#39;RP5-827C21.4&#39;, &#39;RP5-887A10.1&#39;, &#39;RPH3A&#39;, &#39;RPL39L&#39;, &#39;RPL7L1&#39;, &#39;RPN2&#39;, &#39;RPS6KL1&#39;, &#39;RPUSD2&#39;, &#39;RRAGC&#39;, &#39;RRS1&#39;, &#39;RUNDC1&#39;, &#39;S100A11&#39;, &#39;S100A12&#39;, &#39;S100A8&#39;, &#39;S100B&#39;, &#39;SAFB2&#39;, &#39;SAMD1&#39;, &#39;SAMD3&#39;, &#39;SAMSN1&#39;, &#39;SARDH&#39;, &#39;SARS&#39;, &#39;SAT1&#39;, &#39;SCAI&#39;, &#39;SCAPER&#39;, &#39;SCGB3A1&#39;, &#39;SCPEP1&#39;, &#39;SDCCAG8&#39;, &#39;SDPR&#39;, &#39;SEC61A2&#39;, &#39;SELL&#39;, &#39;SEPT11&#39;, &#39;SERAC1&#39;, &#39;SETD1B&#39;, &#39;SF3B1&#39;, &#39;SF3B5&#39;, &#39;SH3GLB1&#39;, &#39;SH3KBP1&#39;, &#39;SHOC2&#39;, &#39;SHPK-1&#39;, &#39;SIAH2&#39;, &#39;SIRPG&#39;, &#39;SIRT1&#39;, &#39;SIVA1&#39;, &#39;SLA&#39;, &#39;SLBP&#39;, &#39;SLC22A4&#39;, &#39;SLC25A11&#39;, &#39;SLC25A12&#39;, &#39;SLC25A14&#39;, &#39;SLC27A1&#39;, &#39;SLC2A13&#39;, &#39;SLC35A2&#39;, &#39;SLC48A1&#39;, &#39;SLFN5&#39;, &#39;SMARCA4&#39;, &#39;SMARCC2&#39;, &#39;SMC2&#39;, &#39;SMCHD1&#39;, &#39;SMDT1&#39;, &#39;SMIM14&#39;, &#39;SMIM7&#39;, &#39;SNAP47&#39;, &#39;SNHG12&#39;, &#39;SNHG8&#39;, &#39;SNTA1&#39;, &#39;SNX29P2&#39;, &#39;SOX13&#39;, &#39;SPARC&#39;, &#39;SPATA7&#39;, &#39;SPG7&#39;, &#39;SPIB&#39;, &#39;SPIN1&#39;, &#39;SPOCD1&#39;, &#39;SPON2&#39;, &#39;SPSB2&#39;, &#39;SREBF1&#39;, &#39;SRM&#39;, &#39;SRP9&#39;, &#39;SRSF6&#39;, &#39;SSBP1&#39;, &#39;ST3GAL2&#39;, &#39;STAMBP&#39;, &#39;STAU2&#39;, &#39;STK17A&#39;, &#39;STK38&#39;, &#39;STMN1&#39;, &#39;STOML2&#39;, &#39;STUB1&#39;, &#39;STX16&#39;, &#39;STX18&#39;, &#39;SUCLG2&#39;, &#39;SUOX&#39;, &#39;SURF1&#39;, &#39;SURF6&#39;, &#39;SWAP70&#39;, &#39;SYCE1&#39;, &#39;SYP&#39;, &#39;SYVN1&#39;, &#39;TACR2&#39;, &#39;TADA2A&#39;, &#39;TAF10&#39;, &#39;TAF12&#39;, &#39;TAF1D&#39;, &#39;TAL1&#39;, &#39;TALDO1&#39;, &#39;TAPBP&#39;, &#39;TARSL2&#39;, &#39;TASP1&#39;, &#39;TBC1D15&#39;, &#39;TBCK&#39;, &#39;TBXA2R&#39;, &#39;TCEAL4&#39;, &#39;TCEAL8&#39;, &#39;TCL1A&#39;, &#39;TCL1B&#39;, &#39;TCP1&#39;, &#39;TDG&#39;, &#39;TERF2IP&#39;, &#39;TGFBRAP1&#39;, &#39;THAP2&#39;, &#39;THEM4&#39;, &#39;THOC7&#39;, &#39;THUMPD3&#39;, &#39;THYN1&#39;, &#39;TIGIT&#39;, &#39;TIMM10B&#39;, &#39;TMEM116&#39;, &#39;TMEM138&#39;, &#39;TMEM140&#39;, &#39;TMEM14B&#39;, &#39;TMEM165&#39;, &#39;TMEM177&#39;, &#39;TMEM194A&#39;, &#39;TMEM219&#39;, &#39;TMEM242&#39;, &#39;TMEM40&#39;, &#39;TMEM60&#39;, &#39;TMEM80&#39;, &#39;TMEM87A&#39;, &#39;TMEM87B&#39;, &#39;TMEM91&#39;, &#39;TMTC2&#39;, &#39;TMX2&#39;, &#39;TMX3&#39;, &#39;TNFRSF17&#39;, &#39;TNFRSF25&#39;, &#39;TNFRSF4&#39;, &#39;TNFRSF9&#39;, &#39;TNFSF10&#39;, &#39;TOP1MT&#39;, &#39;TOP2B&#39;, &#39;TRABD2A&#39;, &#39;TRAF3IP3&#39;, &#39;TRAPPC12-AS1&#39;, &#39;TRAPPC3&#39;, &#39;TREML1&#39;, &#39;TRIM23&#39;, &#39;TRIP12&#39;, &#39;TRIT1&#39;, &#39;TRMT61A&#39;, &#39;TRPM4&#39;, &#39;TSC22D1&#39;, &#39;TSPAN15&#39;, &#39;TSSC1&#39;, &#39;TTC1&#39;, &#39;TTC14&#39;, &#39;TTC3&#39;, &#39;TTC8&#39;, &#39;TTN-AS1&#39;, &#39;TUBB1&#39;, &#39;TUBG2&#39;, &#39;TYMP&#39;, &#39;TYROBP&#39;, &#39;U2SURP&#39;, &#39;UBA5&#39;, &#39;UBAC2&#39;, &#39;UBE2D2&#39;, &#39;UBE2D4&#39;, &#39;UBE2K&#39;, &#39;UBE2Q1&#39;, &#39;UBE3A&#39;, &#39;UBIAD1&#39;, &#39;UBLCP1&#39;, &#39;UBXN4&#39;, &#39;UCK1&#39;, &#39;UNC45A&#39;, &#39;UQCC1&#39;, &#39;URB2&#39;, &#39;URGCP&#39;, &#39;USP30&#39;, &#39;USP33&#39;, &#39;USP36&#39;, &#39;USP38&#39;, &#39;USP5&#39;, &#39;USP7&#39;, &#39;VAMP5&#39;, &#39;VDAC3&#39;, &#39;VIPR1&#39;, &#39;VPS13A&#39;, &#39;VPS13C&#39;, &#39;VPS25&#39;, &#39;VPS26B&#39;, &#39;VPS28&#39;, &#39;VTI1A&#39;, &#39;VTI1B&#39;, &#39;WARS2&#39;, &#39;WBP2NL&#39;, &#39;WDR55&#39;, &#39;WDR91&#39;, &#39;WDYHV1&#39;, &#39;WNK1&#39;, &#39;WTAP&#39;, &#39;XCL2&#39;, &#39;XPOT&#39;, &#39;XRRA1&#39;, &#39;XXbac-BPG299F13.17&#39;, &#39;YEATS2&#39;, &#39;YES1&#39;, &#39;YPEL2&#39;, &#39;YPEL3&#39;, &#39;YTHDF2&#39;, &#39;ZAP70&#39;, &#39;ZBED5-AS1&#39;, &#39;ZBP1&#39;, &#39;ZC3H15&#39;, &#39;ZCCHC11&#39;, &#39;ZCCHC9&#39;, &#39;ZFAND4&#39;, &#39;ZNF175&#39;, &#39;ZNF232&#39;, &#39;ZNF256&#39;, &#39;ZNF263&#39;, &#39;ZNF276&#39;, &#39;ZNF32&#39;, &#39;ZNF350&#39;, &#39;ZNF436&#39;, &#39;ZNF45&#39;, &#39;ZNF493&#39;, &#39;ZNF503&#39;, &#39;ZNF528&#39;, &#39;ZNF559&#39;, &#39;ZNF561&#39;, &#39;ZNF587B&#39;, &#39;ZNF594&#39;, &#39;ZNF653&#39;, &#39;ZNF682&#39;, &#39;ZNF688&#39;, &#39;ZNF718&#39;, &#39;ZNF747&#39;, &#39;ZNF799&#39;, &#39;ZNF836&#39;, &#39;ZNF92&#39;, &#39;ZRANB3&#39;, &#39;ZSWIM6&#39;, &#39;ZUFSP&#39;]
2023-05-26 17:16:31,768 [INFO] probe_length_min = 38
2023-05-26 17:16:31,769 [INFO] probe_length_max = 45
2023-05-26 17:16:31,771 [INFO] min_probes_per_gene = 3
2023-05-26 17:16:31,772 [INFO] n_jobs = 4
2023-05-26 17:56:26,665 [INFO] Genes with &lt;= 3 probes will be removed from the probe database and their names will be stored in &#39;./output/regions_with_insufficient_oligos.txt&#39;.
2023-05-26 17:56:26,862 [INFO] Step - Generate Probes: the database contains 35957196 probes from 887 genes.
</pre></div></div>
</div>
<p>In order to create experiment-specific probes, we have to apply several filter to each probe, e.g. melting temperature or GC content filters.</p>
</div>
<div class="section" id="Parameters-for-Property-Filters">
<h3>Parameters for Property Filters<a class="headerlink" href="#Parameters-for-Property-Filters" title="Permalink to this heading"></a></h3>
<p><em>Parameters for Probe Sequence:</em> - GC_content_min: minimum GC content of probes - GC_content_max: maximum GC content of probes - Tm_min: minimum melting temperature of probes - Tm_max: maximum melting temperature of probes</p>
<p><em>Parameters for Padlock Arms:</em> - min_arm_length: minimum length of each arm - max_arm_Tm_dif: maximum melting temperature difference of both arms - arm_Tm_min: minimum melting temperature of each arm (difference shouldn’t be higher than 5! But range is not super important, the lower the better) - arm_Tm_max: maximum melting temperature of each arm</p>
<p><em>Parameters for Melting Temperature:</em> - Tm_parameters_probe: melting temperature parameters for probe design - Tm_chem_correction_param_pobe: parameters for chemical correction of melting temperature for probe design</p>
<p><em>Note: The melting temperature is used in 2 different stages (probe and detection oligo design), where a few parameters are shared and the others differ. Parameters for melting temperature -&gt; for more information on parameters, see:</em><a class="reference external" href="https://biopython.org/docs/1.75/api/Bio.SeqUtils.MeltingTemp.html#Bio.SeqUtils.MeltingTemp.Tm_NN">here</a></p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[8]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1">####### Load existing database #######</span>
<span class="c1"># file_database = &quot;./output/oligo_database/probe_database_initial.txt&quot;</span>
<span class="c1"># min_probes_per_gene = 3</span>
<span class="c1"># probe_database = probe_designer.load_probe_database(file_database=file_database, min_probes_per_gene=min_probes_per_gene)</span>

<span class="c1">####### Apply Property Filter #######</span>
<span class="n">GC_content_min</span><span class="o">=</span><span class="mi">40</span>
<span class="n">GC_content_max</span><span class="o">=</span><span class="mi">60</span>
<span class="n">Tm_min</span><span class="o">=</span><span class="mi">52</span>
<span class="n">Tm_max</span><span class="o">=</span><span class="mi">67</span>
<span class="n">min_arm_length</span><span class="o">=</span><span class="mi">10</span>
<span class="n">max_arm_Tm_dif</span><span class="o">=</span><span class="mi">2</span>
<span class="n">arm_Tm_min</span><span class="o">=</span><span class="mi">38</span>
<span class="n">arm_Tm_max</span><span class="o">=</span><span class="mi">49</span>

<span class="n">probe_database</span><span class="p">,</span> <span class="n">file_database</span> <span class="o">=</span> <span class="n">probe_designer</span><span class="o">.</span><span class="n">filter_probes_by_property</span><span class="p">(</span><span class="n">probe_database</span><span class="p">,</span> <span class="n">GC_content_min</span><span class="o">=</span><span class="n">GC_content_min</span><span class="p">,</span> <span class="n">GC_content_max</span><span class="o">=</span><span class="n">GC_content_max</span><span class="p">,</span>
                                                                         <span class="n">Tm_min</span><span class="o">=</span><span class="n">Tm_min</span><span class="p">,</span> <span class="n">Tm_max</span><span class="o">=</span><span class="n">Tm_max</span><span class="p">,</span> <span class="n">min_arm_length</span><span class="o">=</span><span class="n">min_arm_length</span><span class="p">,</span> <span class="n">max_arm_Tm_dif</span><span class="o">=</span><span class="n">max_arm_Tm_dif</span><span class="p">,</span> <span class="n">arm_Tm_min</span><span class="o">=</span><span class="n">arm_Tm_min</span><span class="p">,</span> <span class="n">arm_Tm_max</span><span class="o">=</span><span class="n">arm_Tm_max</span><span class="p">,</span> <span class="n">n_jobs</span><span class="o">=</span><span class="mi">4</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
2023-05-26 18:41:05,992 [INFO] Parameters Property Filters:
2023-05-26 18:41:06,004 [INFO] probe_database = &lt;oligo_designer_toolsuite.database._oligos_database.OligoDatabase object at 0x134ce32b0&gt;
2023-05-26 18:41:06,010 [INFO] GC_content_min = 40
2023-05-26 18:41:06,012 [INFO] GC_content_max = 60
2023-05-26 18:41:06,015 [INFO] Tm_min = 52
2023-05-26 18:41:06,016 [INFO] Tm_max = 67
2023-05-26 18:41:06,017 [INFO] min_arm_length = 10
2023-05-26 18:41:06,019 [INFO] max_arm_Tm_dif = 2
2023-05-26 18:41:06,025 [INFO] arm_Tm_min = 38
2023-05-26 18:41:06,026 [INFO] arm_Tm_max = 49
2023-05-26 18:41:06,029 [INFO] Tm_parameters_probe = {&#39;check&#39;: True, &#39;strict&#39;: True, &#39;c_seq&#39;: None, &#39;shift&#39;: 0, &#39;nn_table&#39;: &#39;DNA_NN3&#39;, &#39;tmm_table&#39;: &#39;DNA_TMM1&#39;, &#39;imm_table&#39;: &#39;DNA_IMM1&#39;, &#39;de_table&#39;: &#39;DNA_DE1&#39;, &#39;dnac1&#39;: 50, &#39;dnac2&#39;: 0, &#39;selfcomp&#39;: False, &#39;dNTPs&#39;: 0, &#39;saltcorr&#39;: 7, &#39;Na&#39;: 1.25, &#39;K&#39;: 75, &#39;Tris&#39;: 20, &#39;Mg&#39;: 10}
2023-05-26 18:41:06,030 [INFO] Tm_chem_correction_param_probe = {&#39;DMSO&#39;: 0, &#39;DMSOfactor&#39;: 0.75, &#39;fmdfactor&#39;: 0.65, &#39;fmdmethod&#39;: 1, &#39;GC&#39;: None, &#39;fmd&#39;: 20}
2023-05-26 18:41:06,031 [INFO] n_jobs = 4
2023-05-26 21:18:45,849 [INFO] Step - Filter Probes by Sequence Property: the database contains 3732914 probes from 882 genes, while 32224282 probes and 5 genes have been deleted in this step.
</pre></div></div>
</div>
</div>
<div class="section" id="Parameters-for-Specificity-Filters">
<h3>Parameters for Specificity Filters<a class="headerlink" href="#Parameters-for-Specificity-Filters" title="Permalink to this heading"></a></h3>
<p><em>BlastN Similarity Filter:</em> - blast_word_size: word size for the blastn seed (exact match to target) - blast_percent_identity: maximum similarity between oligos and target sequences, ranging from 0 to 100% (no missmatch) - blast_coverage: minimum coverage between oligos and target sequence, ranging from 0 to 100% (full coverage)</p>
<p><em>Bowtie Ligation Region filter:</em> - ligation_region_size: size of the seed region around the ligation site for bowtie seed region filter</p>
<p><em>Note: Depending on the number of genes, this step might be time and memory consuming. For high number of genes, you might want to run this step on a bigger machine!</em></p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[9]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1">####### Load existing database #######</span>
<span class="c1"># load annotation files for Reference Database</span>
<span class="c1"># source = &quot;custom&quot;</span>
<span class="c1"># custom_params = {</span>
<span class="c1">#     &quot;file_annotation&quot;: &quot;./output/annotation/GCF_000001405.40_GRCh38.p14_genomic.gtf&quot;,</span>
<span class="c1">#     &quot;file_sequence&quot;: &quot;./output/annotation/GCF_000001405.40_GRCh38.p14_genomic.fna&quot;,</span>
<span class="c1">#     &quot;files_source&quot;: &quot;NCBI&quot;,</span>
<span class="c1">#     &quot;species&quot;: &quot;Homo_sapiens&quot;,</span>
<span class="c1">#     &quot;annotation_release&quot;: &quot;110&quot;,</span>
<span class="c1">#     &quot;genome_assembly&quot;: &quot;GRCh38.p14&quot;,</span>
<span class="c1"># }</span>
<span class="c1"># probe_designer.load_annotations(source=source, source_params=custom_params)</span>

<span class="c1"># # load existing database</span>
<span class="c1"># file_database = &quot;./output/oligo_database/probe_database_property_filter.txt&quot;</span>
<span class="c1"># min_probes_per_gene = 3</span>
<span class="c1"># probe_database = probe_designer.load_probe_database(file_database=file_database, min_probes_per_gene=min_probes_per_gene)</span>

<span class="c1">####### Apply Specificity Filter #######</span>
<span class="n">ligation_region_size</span><span class="o">=</span><span class="mi">5</span>
<span class="n">blast_word_size</span><span class="o">=</span><span class="mi">10</span>
<span class="n">blast_percent_identity</span><span class="o">=</span><span class="mi">80</span>
<span class="n">blast_coverage</span><span class="o">=</span><span class="mi">50</span>

<span class="n">probe_database</span><span class="p">,</span> <span class="n">file_database</span> <span class="o">=</span> <span class="n">probe_designer</span><span class="o">.</span><span class="n">filter_probes_by_specificity</span><span class="p">(</span><span class="n">probe_database</span><span class="p">,</span> <span class="n">ligation_region_size</span><span class="o">=</span><span class="n">ligation_region_size</span><span class="p">,</span>
                                                                            <span class="n">blast_word_size</span><span class="o">=</span><span class="n">blast_word_size</span><span class="p">,</span> <span class="n">blast_percent_identity</span><span class="o">=</span><span class="n">blast_percent_identity</span><span class="p">,</span> <span class="n">blast_coverage</span><span class="o">=</span><span class="n">blast_coverage</span><span class="p">,</span> <span class="n">n_jobs</span><span class="o">=</span><span class="mi">2</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
2023-05-26 21:18:46,642 [INFO] Parameters Specificity Filters:
2023-05-26 21:18:46,647 [INFO] probe_database = &lt;oligo_designer_toolsuite.database._oligos_database.OligoDatabase object at 0x134ce32b0&gt;
2023-05-26 21:18:46,651 [INFO] ligation_region_size = 5
2023-05-26 21:18:46,652 [INFO] blast_word_size = 10
2023-05-26 21:18:46,653 [INFO] blast_percent_identity = 80
2023-05-26 21:18:46,654 [INFO] blast_coverage = 50
2023-05-26 21:18:46,654 [INFO] n_jobs = 2
2023-05-27 01:19:33,100 [INFO] Step - Filter Probes by Specificity: the database contains 570410 probes from 796 genes, while 3162504 probes and 86 genes have been deleted in this step.
</pre></div></div>
</div>
<p>After applying different sets of filters to the probe database, we will create probesets for each gene, which are sets of probes that do not overlap and have a high efficiency score (calculated from melting temperature and GC content).</p>
</div>
<div class="section" id="Parameters-for-Oligo-Efficiency-Score">
<h3>Parameters for Oligo Efficiency Score<a class="headerlink" href="#Parameters-for-Oligo-Efficiency-Score" title="Permalink to this heading"></a></h3>
<ul class="simple">
<li><p>Tm_min: minimum melting temperature of probes</p></li>
<li><p>Tm_max: maximum melting temperature of probes</p></li>
<li><p>Tm_opt: optimal melting temperature of probes</p></li>
<li><p>Tm_weight: weight of the Tm of the probe in the efficiency score</p></li>
<li><p>GC_content_min: minimum GC content of probes</p></li>
<li><p>GC_content_max: maximum GC content of probes</p></li>
<li><p>GC_content_opt: optimal GC content of probes</p></li>
<li><p>GC_weight: weight of the GC content of the probe in the efficiency score</p></li>
</ul>
</div>
<div class="section" id="Parameters-for-Oligosets-Generation">
<h3>Parameters for Oligosets Generation<a class="headerlink" href="#Parameters-for-Oligosets-Generation" title="Permalink to this heading"></a></h3>
<ul class="simple">
<li><p>probeset_size_opt: ideal number of oligos per probeset</p></li>
<li><p>probeset_size_min: minimum number of oligos per probeset</p></li>
<li><p>n_sets: maximum number of sets per gene</p></li>
</ul>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[10]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1">####### Load existing database #######</span>
<span class="c1"># file_database = &quot;./output/oligo_database/oligo_database_specificity_filters.txt&quot;</span>
<span class="c1"># min_probes_per_gene = 3</span>
<span class="c1"># probe_database = probe_designer.load_probe_database(file_database=file_database, min_probes_per_gene=min_probes_per_gene)</span>

<span class="c1">####### Apply Probe Set Selection #######</span>
<span class="n">probeset_size_opt</span><span class="o">=</span><span class="mi">5</span>
<span class="n">probeset_size_min</span><span class="o">=</span><span class="mi">2</span>
<span class="n">n_sets</span><span class="o">=</span><span class="mi">100</span>
<span class="n">Tm_min</span><span class="o">=</span><span class="mi">52</span>
<span class="n">Tm_max</span><span class="o">=</span><span class="mi">67</span>
<span class="n">Tm_opt</span><span class="o">=</span><span class="mi">60</span>
<span class="n">Tm_weight</span><span class="o">=</span><span class="mi">1</span>
<span class="n">GC_content_min</span><span class="o">=</span><span class="mi">40</span>
<span class="n">GC_content_max</span><span class="o">=</span><span class="mi">60</span>
<span class="n">GC_content_opt</span><span class="o">=</span><span class="mi">50</span>
<span class="n">GC_weight</span><span class="o">=</span><span class="mi">1</span>

<span class="n">probe_database</span><span class="p">,</span> <span class="n">file_database</span><span class="p">,</span> <span class="n">dir_oligosets</span> <span class="o">=</span> <span class="n">probe_designer</span><span class="o">.</span><span class="n">create_probe_sets</span><span class="p">(</span><span class="n">probe_database</span><span class="p">,</span>
                                                                                <span class="n">probeset_size_opt</span><span class="o">=</span><span class="n">probeset_size_opt</span><span class="p">,</span>
                                                                                <span class="n">probeset_size_min</span><span class="o">=</span><span class="n">probeset_size_min</span><span class="p">,</span>
                                                                                <span class="n">n_sets</span><span class="o">=</span><span class="n">n_sets</span><span class="p">,</span>
                                                                                <span class="n">Tm_min</span><span class="o">=</span><span class="n">Tm_min</span><span class="p">,</span>
                                                                                <span class="n">Tm_max</span><span class="o">=</span><span class="n">Tm_max</span><span class="p">,</span>
                                                                                <span class="n">Tm_opt</span><span class="o">=</span><span class="n">Tm_opt</span><span class="p">,</span>
                                                                                <span class="n">Tm_weight</span><span class="o">=</span><span class="n">Tm_weight</span><span class="p">,</span>
                                                                                <span class="n">GC_content_min</span><span class="o">=</span><span class="n">GC_content_min</span><span class="p">,</span>
                                                                                <span class="n">GC_content_max</span><span class="o">=</span><span class="n">GC_content_max</span><span class="p">,</span>
                                                                                <span class="n">GC_content_opt</span><span class="o">=</span><span class="n">GC_content_opt</span><span class="p">,</span>
                                                                                <span class="n">GC_weight</span><span class="o">=</span><span class="n">GC_weight</span><span class="p">,</span>
                                                                                <span class="n">n_jobs</span><span class="o">=</span><span class="mi">2</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
2023-05-27 01:19:33,698 [INFO] Parameters Probesets:
2023-05-27 01:19:33,702 [INFO] probe_database = &lt;oligo_designer_toolsuite.database._oligos_database.OligoDatabase object at 0x134ce32b0&gt;
2023-05-27 01:19:33,703 [INFO] probeset_size_opt = 5
2023-05-27 01:19:33,705 [INFO] probeset_size_min = 2
2023-05-27 01:19:33,706 [INFO] n_sets = 100
2023-05-27 01:19:33,706 [INFO] Tm_min = 52
2023-05-27 01:19:33,707 [INFO] Tm_max = 67
2023-05-27 01:19:33,708 [INFO] Tm_opt = 60
2023-05-27 01:19:33,709 [INFO] Tm_weight = 1
2023-05-27 01:19:33,710 [INFO] GC_content_min = 40
2023-05-27 01:19:33,710 [INFO] GC_content_max = 60
2023-05-27 01:19:33,711 [INFO] GC_content_opt = 50
2023-05-27 01:19:33,712 [INFO] GC_weight = 1
2023-05-27 01:19:33,713 [INFO] n_jobs = 2
2023-05-27 03:19:32,175 [INFO] Step - Generate Oligosets: the database contains 11525 probes from 756 genes, while 558885 probes and 40 genes have been deleted in this step.
</pre></div></div>
</div>
<div class="line-block">
<div class="line">In the probe database, the gene names are the keys of the database. All genes that do not have sufficient probes were removed from the database.</div>
<div class="line">Once we hve all genes with sufficient probes, we can run the gene set selection step. Therefore, we include additional metadata information to our adata object, i.e. the genes that have sufficient probes and the genes that fulfill both constraint (<em>highly variable</em> and <em>sufficient probes</em>).</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[11]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># get gene names of genes with sufficient number of probes to proceed with next step</span>
<span class="n">genes_with_sufficient_probes</span> <span class="o">=</span> <span class="n">probe_database</span><span class="o">.</span><span class="n">database</span><span class="o">.</span><span class="n">keys</span><span class="p">()</span>

<span class="c1"># add extra clolumn to anndata to mark genes with sufficient probes</span>
<span class="n">pbmc_data</span><span class="o">.</span><span class="n">var</span><span class="p">[</span><span class="s2">&quot;sufficient_probes&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="kc">False</span>
<span class="k">for</span> <span class="n">gene</span> <span class="ow">in</span> <span class="n">highly_variable_genes</span><span class="p">:</span>
    <span class="k">if</span> <span class="n">gene</span> <span class="ow">in</span> <span class="n">genes_with_sufficient_probes</span><span class="p">:</span>
        <span class="n">pbmc_data</span><span class="o">.</span><span class="n">var</span><span class="p">[</span><span class="s2">&quot;sufficient_probes&quot;</span><span class="p">][</span><span class="n">gene</span><span class="p">]</span> <span class="o">=</span> <span class="kc">True</span>

<span class="c1"># create a new variable that indicates if the gene passes the first constraint filter</span>
<span class="n">pbmc_data</span><span class="o">.</span><span class="n">var</span><span class="p">[</span><span class="s2">&quot;pass_constraints&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="n">su_p</span> <span class="ow">and</span> <span class="n">hi_v</span> <span class="k">for</span> <span class="n">su_p</span><span class="p">,</span> <span class="n">hi_v</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span><span class="n">pbmc_data</span><span class="o">.</span><span class="n">var</span><span class="p">[</span><span class="s2">&quot;sufficient_probes&quot;</span><span class="p">],</span> <span class="n">pbmc_data</span><span class="o">.</span><span class="n">var</span><span class="p">[</span><span class="s2">&quot;highly_variable&quot;</span><span class="p">])]</span>
</pre></div>
</div>
</div>
<div class="line-block">
<div class="line">To run the geneset selection step, we create an instance of the <code class="docutils literal notranslate"><span class="pre">ProbesetSelector</span></code> class. We specify the number of genes <code class="docutils literal notranslate"><span class="pre">n</span></code> (20) and the keys in <code class="docutils literal notranslate"><span class="pre">adata.obs</span></code> and <code class="docutils literal notranslate"><span class="pre">adata.var</span></code> where we find the cell type annotations (<code class="docutils literal notranslate"><span class="pre">celltype_key=&quot;celltype&quot;</span></code>) and selected genes (<code class="docutils literal notranslate"><span class="pre">celltype_key=&quot;celltype&quot;</span></code>), respectively.</div>
<div class="line"><em>Note: that you can specify a ``save_dir`` to save results during the selection step and reload them next time a ``ProbesetSelector`` with the given ``save_dir`` is instantiated.</em></div>
</div>
<p>Executing the cell below will give us a warning that the cell type clusters for dendritic cells and megakaroycytes are quite small and therefore the genes that are selected to identify these cell types potentially don’t generalize very well. The method will not exclude these cell types automatically, but it can be done manually by setting the parameter <code class="docutils literal notranslate"><span class="pre">celltypes</span></code> to a subset of cell types instead of <code class="docutils literal notranslate"><span class="pre">celltypes=&quot;all&quot;</span></code>.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[12]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1">##### Select genes for gene panel #####</span>
<span class="n">selector</span> <span class="o">=</span> <span class="n">sp</span><span class="o">.</span><span class="n">se</span><span class="o">.</span><span class="n">ProbesetSelector</span><span class="p">(</span><span class="n">pbmc_data</span><span class="p">,</span> <span class="n">n</span><span class="o">=</span><span class="mi">20</span><span class="p">,</span> <span class="n">genes_key</span><span class="o">=</span><span class="s2">&quot;pass_constraints&quot;</span><span class="p">,</span> <span class="n">celltype_key</span><span class="o">=</span><span class="s2">&quot;celltype&quot;</span><span class="p">,</span> <span class="n">verbosity</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">save_dir</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span>
<span class="n">selector</span><span class="o">.</span><span class="n">select_probeset</span><span class="p">()</span>
<span class="n">selected_genes</span> <span class="o">=</span> <span class="n">selector</span><span class="o">.</span><span class="n">probeset</span><span class="o">.</span><span class="n">index</span><span class="p">[</span><span class="n">selector</span><span class="o">.</span><span class="n">probeset</span><span class="o">.</span><span class="n">selection</span><span class="p">]</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<script type="application/vnd.jupyter.widget-view+json">{"model_id": "8670f070af094472be824ecf093d55a3", "version_major": 2, "version_minor": 0}</script></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Note: The following celltypes&#39; test set sizes for forest training are below min_test_n (=20):
         Dendritic cells : 9
         Megakaryocytes  : 3
The genes selected for those cell types potentially don&#39;t generalize well. Find the genes for each of those cell types in self.genes_of_primary_trees after running self.select_probeset().
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area rendered_html docutils container">
<pre style="white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace"></pre></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area rendered_html docutils container">
<pre style="white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace">
</pre></div>
</div>
<p>Once we have all selected genes, we create the final “read to order” probe sequences. Calling the fuction below will produce two files, *[padlock, merfish, seqfish]_probes* and *[padlock, merfish, seqfish]_probes_order*. The latter file contains the ready to order probe sequences for each gene.</p>
</div>
<div class="section" id="Parameters-for-Padlock-Final-Sequence-Design">
<h3>Parameters for Padlock Final Sequence Design<a class="headerlink" href="#Parameters-for-Padlock-Final-Sequence-Design" title="Permalink to this heading"></a></h3>
<ul class="simple">
<li><p>detect_oligo_length_min: minimum length of detection oligo</p></li>
<li><p>detect_oligo_length_max: maximum length of detection oligo</p></li>
<li><p>detect_oligo_Tm_opt: optimal melting temperature of detection oligo</p></li>
<li><p>Tm_parameters_detection_oligo: melting temperature parameters for detection oligo design</p></li>
<li><p>Tm_chem_correction_param_detection_oligo: parameters for chemical correction of melting temperature for detection oligo design</p></li>
</ul>
<p><em>Note: The melting temperature is used in 2 different stages (probe and detection oligo design), where a few parameters are shared and the others differ. Parameters for melting temperature -&gt; for more information on parameters, see:</em><a class="reference external" href="https://biopython.org/docs/1.75/api/Bio.SeqUtils.MeltingTemp.html#Bio.SeqUtils.MeltingTemp.Tm_NN">here</a></p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[13]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1">##### Remove all genes from the database that are not selected for the gene panel ####</span>
<span class="n">probe_database</span><span class="o">.</span><span class="n">database</span> <span class="o">=</span> <span class="p">{</span><span class="n">key</span><span class="p">:</span> <span class="n">value</span> <span class="k">for</span> <span class="n">key</span><span class="p">,</span> <span class="n">value</span> <span class="ow">in</span> <span class="n">probe_database</span><span class="o">.</span><span class="n">database</span><span class="o">.</span><span class="n">items</span><span class="p">()</span> <span class="k">if</span> <span class="n">key</span> <span class="ow">in</span> <span class="n">selected_genes</span><span class="p">}</span>
<span class="n">probe_database</span><span class="o">.</span><span class="n">oligosets</span> <span class="o">=</span> <span class="p">{</span><span class="n">key</span><span class="p">:</span> <span class="n">value</span> <span class="k">for</span> <span class="n">key</span><span class="p">,</span> <span class="n">value</span> <span class="ow">in</span> <span class="n">probe_database</span><span class="o">.</span><span class="n">oligosets</span><span class="o">.</span><span class="n">items</span><span class="p">()</span> <span class="k">if</span> <span class="n">key</span> <span class="ow">in</span> <span class="n">selected_genes</span><span class="p">}</span>
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[14]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1">##### Design final sequences #####</span>
<span class="n">detect_oligo_length_min</span> <span class="o">=</span> <span class="mi">18</span>
<span class="n">detect_oligo_length_max</span> <span class="o">=</span> <span class="mi">25</span>
<span class="n">detect_oligo_Tm_opt</span> <span class="o">=</span> <span class="mi">32</span>

<span class="n">probe_designer</span><span class="o">.</span><span class="n">create_final_sequences</span><span class="p">(</span><span class="n">probe_database</span><span class="p">,</span> <span class="n">detect_oligo_length_min</span><span class="p">,</span> <span class="n">detect_oligo_length_max</span><span class="p">,</span> <span class="n">detect_oligo_Tm_opt</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
2023-05-27 03:23:40,007 [DEBUG] handle_msg[8670f070af094472be824ecf093d55a3]({&#39;header&#39;: {&#39;date&#39;: datetime.datetime(2023, 5, 27, 1, 23, 39, 934000, tzinfo=tzutc()), &#39;msg_id&#39;: &#39;5a26b004-efe4-482a-a30b-7bed1068077b&#39;, &#39;msg_type&#39;: &#39;comm_msg&#39;, &#39;session&#39;: &#39;f80bbb6a-f083-4dbc-9c24-00aa83c84915&#39;, &#39;username&#39;: &#39;16c5b64d-425a-4c50-b860-381f99d9c018&#39;, &#39;version&#39;: &#39;5.2&#39;}, &#39;msg_id&#39;: &#39;5a26b004-efe4-482a-a30b-7bed1068077b&#39;, &#39;msg_type&#39;: &#39;comm_msg&#39;, &#39;parent_header&#39;: {}, &#39;metadata&#39;: {}, &#39;content&#39;: {&#39;comm_id&#39;: &#39;8670f070af094472be824ecf093d55a3&#39;, &#39;data&#39;: {&#39;method&#39;: &#39;update&#39;, &#39;state&#39;: {&#39;outputs&#39;: [{&#39;output_type&#39;: &#39;display_data&#39;, &#39;data&#39;: {&#39;text/plain&#39;: &#39;\x1b[1;30mSPAPROS PROBESET SELECTION:                                                                     \x1b[0m \x1b[33m0:04:06\x1b[0m\n\x1b[1;34mSelect pca genes..........................................\x1b[0m \x1b[38;2;114;156;31m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\x1b[0m  \x1b[35m100%\x1b[0m \x1b[33m0:00:00\x1b[0m\n\x1b[1;34mTrain baseline forest based on DE genes...................\x1b[0m \x1b[38;2;114;156;31m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\x1b[0m  \x1b[35m 4/4\x1b[0m \x1b[33m0:03:12\x1b[0m\n  \x1b[1;2;36mSelect DE genes.........................................\x1b[0m \x1b[38;2;114;156;31m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\x1b[0m  \x1b[35m 8/8\x1b[0m \x1b[33m0:00:00\x1b[0m\n  \x1b[1;2;36mTrain prior forest for DE_baseline forest...............\x1b[0m \x1b[38;2;114;156;31m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\x1b[0m  \x1b[35m 3/3\x1b[0m \x1b[33m0:00:40\x1b[0m\n  \x1b[1;2;36mIteratively add DE genes to DE_baseline forest..........\x1b[0m \x1b[38;2;114;156;31m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\x1b[0m  \x1b[35m 3/3\x1b[0m \x1b[33m0:01:41\x1b[0m\n  \x1b[1;2;36mTrain final baseline forest on all celltypes............\x1b[0m \x1b[38;2;114;156;31m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\x1b[0m  \x1b[35m 3/3\x1b[0m \x1b[33m0:00:49\x1b[0m\n\x1b[1;34mTrain final forests.......................................\x1b[0m \x1b[38;2;114;156;31m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\x1b[0m  \x1b[35m 3/3\x1b[0m \x1b[33m0:00:53\x1b[0m\n  \x1b[1;2;36mTrain forest on pre/prior/pca selected genes............\x1b[0m \x1b[38;2;114;156;31m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\x1b[0m  \x1b[35m 3/3\x1b[0m \x1b[33m0:00:53\x1b[0m\n  \x1b[1;2;36mInitial results are good enough. No genes are added.......................................\x1b[0m  \n\x1b[1;34mCompile probeset list.....................................\x1b[0m \x1b[38;2;114;156;31m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\x1b[0m  \x1b[35m100%\x1b[0m \x1b[33m0:00:00\x1b[0m\n\x1b[1;30mFINISHED\x1b[0m  \n          \n&#39;, &#39;text/html&#39;: &#39;&lt;pre style=&#34;white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,\&#39;DejaVu Sans Mono\&#39;,consolas,\&#39;Courier New\&#39;,monospace&#34;&gt;&lt;span style=&#34;color: #000000; text-decoration-color: #000000; font-weight: bold&#34;&gt;SPAPROS PROBESET SELECTION:                                                                     &lt;/span&gt; &lt;span style=&#34;color: #808000; text-decoration-color: #808000&#34;&gt;0:04:06&lt;/span&gt;\n&lt;span style=&#34;color: #000080; text-decoration-color: #000080; font-weight: bold&#34;&gt;Select pca genes..........................................&lt;/span&gt; &lt;span style=&#34;color: #729c1f; text-decoration-color: #729c1f&#34;&gt;━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━&lt;/span&gt;  &lt;span style=&#34;color: #800080; text-decoration-color: #800080&#34;&gt;100%&lt;/span&gt; &lt;span style=&#34;color: #808000; text-decoration-color: #808000&#34;&gt;0:00:00&lt;/span&gt;\n&lt;span style=&#34;color: #000080; text-decoration-color: #000080; font-weight: bold&#34;&gt;Train baseline forest based on DE genes...................&lt;/span&gt; &lt;span style=&#34;color: #729c1f; text-decoration-color: #729c1f&#34;&gt;━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━&lt;/span&gt;  &lt;span style=&#34;color: #800080; text-decoration-color: #800080&#34;&gt; 4/4&lt;/span&gt; &lt;span style=&#34;color: #808000; text-decoration-color: #808000&#34;&gt;0:03:12&lt;/span&gt;\n  &lt;span style=&#34;color: #7fbfbf; text-decoration-color: #7fbfbf; font-weight: bold&#34;&gt;Select DE genes.........................................&lt;/span&gt; &lt;span style=&#34;color: #729c1f; text-decoration-color: #729c1f&#34;&gt;━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━&lt;/span&gt;  &lt;span style=&#34;color: #800080; text-decoration-color: #800080&#34;&gt; 8/8&lt;/span&gt; &lt;span style=&#34;color: #808000; text-decoration-color: #808000&#34;&gt;0:00:00&lt;/span&gt;\n  &lt;span style=&#34;color: #7fbfbf; text-decoration-color: #7fbfbf; font-weight: bold&#34;&gt;Train prior forest for DE_baseline forest...............&lt;/span&gt; &lt;span style=&#34;color: #729c1f; text-decoration-color: #729c1f&#34;&gt;━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━&lt;/span&gt;  &lt;span style=&#34;color: #800080; text-decoration-color: #800080&#34;&gt; 3/3&lt;/span&gt; &lt;span style=&#34;color: #808000; text-decoration-color: #808000&#34;&gt;0:00:40&lt;/span&gt;\n  &lt;span style=&#34;color: #7fbfbf; text-decoration-color: #7fbfbf; font-weight: bold&#34;&gt;Iteratively add DE genes to DE_baseline forest..........&lt;/span&gt; &lt;span style=&#34;color: #729c1f; text-decoration-color: #729c1f&#34;&gt;━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━&lt;/span&gt;  &lt;span style=&#34;color: #800080; text-decoration-color: #800080&#34;&gt; 3/3&lt;/span&gt; &lt;span style=&#34;color: #808000; text-decoration-color: #808000&#34;&gt;0:01:41&lt;/span&gt;\n  &lt;span style=&#34;color: #7fbfbf; text-decoration-color: #7fbfbf; font-weight: bold&#34;&gt;Train final baseline forest on all celltypes............&lt;/span&gt; &lt;span style=&#34;color: #729c1f; text-decoration-color: #729c1f&#34;&gt;━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━&lt;/span&gt;  &lt;span style=&#34;color: #800080; text-decoration-color: #800080&#34;&gt; 3/3&lt;/span&gt; &lt;span style=&#34;color: #808000; text-decoration-color: #808000&#34;&gt;0:00:49&lt;/span&gt;\n&lt;span style=&#34;color: #000080; text-decoration-color: #000080; font-weight: bold&#34;&gt;Train final forests.......................................&lt;/span&gt; &lt;span style=&#34;color: #729c1f; text-decoration-color: #729c1f&#34;&gt;━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━&lt;/span&gt;  &lt;span style=&#34;color: #800080; text-decoration-color: #800080&#34;&gt; 3/3&lt;/span&gt; &lt;span style=&#34;color: #808000; text-decoration-color: #808000&#34;&gt;0:00:53&lt;/span&gt;\n  &lt;span style=&#34;color: #7fbfbf; text-decoration-color: #7fbfbf; font-weight: bold&#34;&gt;Train forest on pre/prior/pca selected genes............&lt;/span&gt; &lt;span style=&#34;color: #729c1f; text-decoration-color: #729c1f&#34;&gt;━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━&lt;/span&gt;  &lt;span style=&#34;color: #800080; text-decoration-color: #800080&#34;&gt; 3/3&lt;/span&gt; &lt;span style=&#34;color: #808000; text-decoration-color: #808000&#34;&gt;0:00:53&lt;/span&gt;\n  &lt;span style=&#34;color: #7fbfbf; text-decoration-color: #7fbfbf; font-weight: bold&#34;&gt;Initial results are good enough. No genes are added.......................................&lt;/span&gt;  \n&lt;span style=&#34;color: #000080; text-decoration-color: #000080; font-weight: bold&#34;&gt;Compile probeset list.....................................&lt;/span&gt; &lt;span style=&#34;color: #729c1f; text-decoration-color: #729c1f&#34;&gt;━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━&lt;/span&gt;  &lt;span style=&#34;color: #800080; text-decoration-color: #800080&#34;&gt;100%&lt;/span&gt; &lt;span style=&#34;color: #808000; text-decoration-color: #808000&#34;&gt;0:00:00&lt;/span&gt;\n&lt;span style=&#34;color: #000000; text-decoration-color: #000000; font-weight: bold&#34;&gt;FINISHED&lt;/span&gt;  \n          \n&lt;/pre&gt;\n&#39;}, &#39;metadata&#39;: {}}]}, &#39;buffer_paths&#39;: []}}, &#39;buffers&#39;: []})
2023-05-27 03:23:40,069 [INFO] Parameters Final Sequence Design:
2023-05-27 03:23:40,070 [INFO] probe_database = &lt;oligo_designer_toolsuite.database._oligos_database.OligoDatabase object at 0x134ce32b0&gt;
2023-05-27 03:23:40,072 [INFO] detect_oligo_length_min = 18
2023-05-27 03:23:40,073 [INFO] detect_oligo_length_max = 25
2023-05-27 03:23:40,074 [INFO] detect_oligo_Tm_opt = 32
2023-05-27 03:23:40,076 [INFO] Tm_parameters_detection_oligo = {&#39;check&#39;: True, &#39;strict&#39;: True, &#39;c_seq&#39;: None, &#39;shift&#39;: 0, &#39;nn_table&#39;: &#39;DNA_NN3&#39;, &#39;tmm_table&#39;: &#39;DNA_TMM1&#39;, &#39;imm_table&#39;: &#39;DNA_IMM1&#39;, &#39;de_table&#39;: &#39;DNA_DE1&#39;, &#39;dnac1&#39;: 50, &#39;dnac2&#39;: 0, &#39;selfcomp&#39;: False, &#39;dNTPs&#39;: 0, &#39;saltcorr&#39;: 7, &#39;Na&#39;: 39, &#39;K&#39;: 0, &#39;Tris&#39;: 0, &#39;Mg&#39;: 0}
2023-05-27 03:23:40,077 [INFO] Tm_chem_correction_param_detection_oligo = {&#39;DMSO&#39;: 0, &#39;DMSOfactor&#39;: 0.75, &#39;fmdfactor&#39;: 0.65, &#39;fmdmethod&#39;: 1, &#39;GC&#39;: None, &#39;fmd&#39;: 30}
2023-05-27 03:23:40,347 [INFO] Step - Design Final Padlock Sequences: padlock sequences are stored in &#39;./output/padlock_sequences/padlock_sequences&#39; directory.
</pre></div></div>
</div>
</div>
</div>
</div>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="spapros_tutorial_end_to_end_selection_short.html" class="btn btn-neutral float-left" title="End to end selection (short version)" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="../authors.html" class="btn btn-neutral float-right" title="Credits" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2022, Lukas Heumos.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>